
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://biocypher.org/0.11.0/learn/tutorials/tutorial_basics_neo4j_offline/tutorial_004_neo4j_offline/">
      
      
        <link rel="prev" href="../../agent_api_guide/">
      
      
        <link rel="next" href="../../../guides/">
      
      
      <link rel="icon" href="../../../../assets/img/favicon_biocypher_32px.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.16">
    
    
      
        <title>Hands-on Building Graphs with BioCypher and Neo4j - BioCypher</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.7e37652d.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Lato";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../../assets/css/css-v1.1.0.css">
    
      <link rel="stylesheet" href="../../../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#hands-on-building-graphs-with-biocypher-offline-mode-and-neo4j" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="BioCypher" class="md-header__button md-logo" aria-label="BioCypher" data-md-component="logo">
      
  <img src="../../../../assets/img/logo-biocypher.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            BioCypher
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Hands-on Building Graphs with BioCypher and Neo4j
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/biocypher/biocypher" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    biocypher/biocypher
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../../biocypher-project/project/" class="md-tabs__link">
          
  
  
  About

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../../installation/" class="md-tabs__link">
          
  
  
  Get Started

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../tutorial001_basics/" class="md-tabs__link">
          
  
  
  Learn

        </a>
      </li>
    
  

    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../../reference/" class="md-tabs__link">
          
  
  
  Reference

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../../community/" class="md-tabs__link">
          
  
  
  Community

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="BioCypher" class="md-nav__button md-logo" aria-label="BioCypher" data-md-component="logo">
      
  <img src="../../../../assets/img/logo-biocypher.png" alt="logo">

    </a>
    BioCypher
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/biocypher/biocypher" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    biocypher/biocypher
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    About
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            About
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../biocypher-project/project/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Project
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../biocypher-project/design-philosophy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Design philosophy
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../biocypher-project/use-cases/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Use Cases
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../biocypher-project/biochatter-integration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    BioCypher + LLMs
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../biocypher-project/r-bioc/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    R and Bioconductor
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Get Started
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Get Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../quickstart/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Quickstart
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../llms/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    LLM Integration
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Learn
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Learn
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial001_basics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Basics
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial002_handling_ontologies/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Handling Ontologies
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial003_adapters/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Adapters
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../pandas_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Example Notebook: BioCypher and Pandas
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../agent_api_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Agent API Guide
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Hands-on Building Graphs with BioCypher and Neo4j
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Hands-on Building Graphs with BioCypher and Neo4j
    
  </span>
  

      </a>
      
        

  

<nav class="md-nav md-nav--secondary" aria-label="On this page">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      On this page
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    <span class="md-ellipsis">
      Pre-requisites
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    <span class="md-ellipsis">
      Setup
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#setup-python-project" class="md-nav__link">
    <span class="md-ellipsis">
      Setup Python project
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#setup-neo4j" class="md-nav__link">
    <span class="md-ellipsis">
      Setup Neo4j
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-1-exploratory-data-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Section 1. Exploratory Data Analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-2-graph-modeling" class="md-nav__link">
    <span class="md-ellipsis">
      Section 2. Graph Modeling
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 2. Graph Modeling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#graph-modeling" class="md-nav__link">
    <span class="md-ellipsis">
      Graph Modeling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exercise-1-example-of-a-graph-we-expect-with-our-data" class="md-nav__link">
    <span class="md-ellipsis">
      Exercise 1. Example of a graph we expect with our data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-3-graph-creation-with-biocypher" class="md-nav__link">
    <span class="md-ellipsis">
      Section 3. Graph creation with BioCypher
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 3. Graph creation with BioCypher">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Step 1. Configuration
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 1. Configuration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#create-a-schema-for-your-graph" class="md-nav__link">
    <span class="md-ellipsis">
      Create a schema for your graph
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Create a schema for your graph">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#nodes" class="md-nav__link">
    <span class="md-ellipsis">
      Nodes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#edges-relationships" class="md-nav__link">
    <span class="md-ellipsis">
      Edges (relationships)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#a-comment-about-the-connection-between-biocypher-and-ontologies" class="md-nav__link">
    <span class="md-ellipsis">
      A comment about the connection between BioCypher and Ontologies
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#configure-biocypher-behavior" class="md-nav__link">
    <span class="md-ellipsis">
      Configure BioCypher behavior
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-create-an-adapter" class="md-nav__link">
    <span class="md-ellipsis">
      Step 2. Create an adapter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 2. Create an adapter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#nodes-convention" class="md-nav__link">
    <span class="md-ellipsis">
      Nodes Convention
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#edges-convention" class="md-nav__link">
    <span class="md-ellipsis">
      Edges Convention
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-create-a-knowledge-graph-script" class="md-nav__link">
    <span class="md-ellipsis">
      Step 3. Create a knowledge graph script
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 3. Create a knowledge graph script">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Run the script
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-4-interacting-with-your-graph-using-neo4j" class="md-nav__link">
    <span class="md-ellipsis">
      Section 4. Interacting with your graph using Neo4j
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 4. Interacting with your graph using Neo4j">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#load-the-graph-using-an-import-script" class="md-nav__link">
    <span class="md-ellipsis">
      Load the graph using an import script
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualize-the-graph" class="md-nav__link">
    <span class="md-ellipsis">
      Visualize the graph
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#execute-cypher-queries" class="md-nav__link">
    <span class="md-ellipsis">
      Execute cypher queries
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Next Steps
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#feedback-contributions" class="md-nav__link">
    <span class="md-ellipsis">
      Feedback &amp; Contributions
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    HowTo / FAQ
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            HowTo / FAQ
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Index
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/htg001_standalone_docker_biocypher/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Standalone Docker Image
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Explanations
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Explanations
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../explanation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Index
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../explanation/adapters/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    About Adapters
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../explanation/ontologies/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    About Ontologies
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../explanation/graph_class_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Graph Class Guide
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../explanation/architecture-migration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Architecture Migration Guide
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Reference Index
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2" >
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    API Documentation
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            API Documentation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/biocypher/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    BioCypher
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/biocypher-workflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    BioCypherWorkflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/graph-class/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Graph Class
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/output-write/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Output Writing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/output-in-memory/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Output In-Memory
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/output-driver/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Output Driver
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/download-cache/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Download and Cache
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/ontology/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Ontology Handling
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/graph-handling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Graph Handling
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/translation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Mapping and Translation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/logging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Logging
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/source/utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Miscellaneous Utility Functions
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/biocypher-config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    BioCypher Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" >
        
          
          <label class="md-nav__link" for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Output Configuration
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            Output Configuration
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/outputs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/outputs/neo4j-output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Neo4j
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/outputs/sqlite-output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    SQLite
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/outputs/postgresql-output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    PostgreSQL
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/outputs/networkx-output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    NetworkX
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/outputs/tabular-output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Tabular Format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/outputs/rdf-output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RDF
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/outputs/owl-output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    OWL
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/outputs/arangodb-output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    ArangoDB
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../reference/schema-config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Schema Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Community
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Community
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../community/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Join Us
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../community/contribute/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Where to Start
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../community/contribute-docs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contribute to the Documentation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../community/contribute-codebase/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contribute to the Code Base
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  

<nav class="md-nav md-nav--secondary" aria-label="On this page">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      On this page
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    <span class="md-ellipsis">
      Pre-requisites
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    <span class="md-ellipsis">
      Setup
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#setup-python-project" class="md-nav__link">
    <span class="md-ellipsis">
      Setup Python project
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#setup-neo4j" class="md-nav__link">
    <span class="md-ellipsis">
      Setup Neo4j
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-1-exploratory-data-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Section 1. Exploratory Data Analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-2-graph-modeling" class="md-nav__link">
    <span class="md-ellipsis">
      Section 2. Graph Modeling
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 2. Graph Modeling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#graph-modeling" class="md-nav__link">
    <span class="md-ellipsis">
      Graph Modeling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exercise-1-example-of-a-graph-we-expect-with-our-data" class="md-nav__link">
    <span class="md-ellipsis">
      Exercise 1. Example of a graph we expect with our data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-3-graph-creation-with-biocypher" class="md-nav__link">
    <span class="md-ellipsis">
      Section 3. Graph creation with BioCypher
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 3. Graph creation with BioCypher">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Step 1. Configuration
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 1. Configuration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#create-a-schema-for-your-graph" class="md-nav__link">
    <span class="md-ellipsis">
      Create a schema for your graph
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Create a schema for your graph">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#nodes" class="md-nav__link">
    <span class="md-ellipsis">
      Nodes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#edges-relationships" class="md-nav__link">
    <span class="md-ellipsis">
      Edges (relationships)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#a-comment-about-the-connection-between-biocypher-and-ontologies" class="md-nav__link">
    <span class="md-ellipsis">
      A comment about the connection between BioCypher and Ontologies
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#configure-biocypher-behavior" class="md-nav__link">
    <span class="md-ellipsis">
      Configure BioCypher behavior
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-create-an-adapter" class="md-nav__link">
    <span class="md-ellipsis">
      Step 2. Create an adapter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 2. Create an adapter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#nodes-convention" class="md-nav__link">
    <span class="md-ellipsis">
      Nodes Convention
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#edges-convention" class="md-nav__link">
    <span class="md-ellipsis">
      Edges Convention
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-create-a-knowledge-graph-script" class="md-nav__link">
    <span class="md-ellipsis">
      Step 3. Create a knowledge graph script
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 3. Create a knowledge graph script">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-the-script" class="md-nav__link">
    <span class="md-ellipsis">
      Run the script
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#section-4-interacting-with-your-graph-using-neo4j" class="md-nav__link">
    <span class="md-ellipsis">
      Section 4. Interacting with your graph using Neo4j
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Section 4. Interacting with your graph using Neo4j">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#load-the-graph-using-an-import-script" class="md-nav__link">
    <span class="md-ellipsis">
      Load the graph using an import script
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualize-the-graph" class="md-nav__link">
    <span class="md-ellipsis">
      Visualize the graph
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#execute-cypher-queries" class="md-nav__link">
    <span class="md-ellipsis">
      Execute cypher queries
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Next Steps
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#feedback-contributions" class="md-nav__link">
    <span class="md-ellipsis">
      Feedback &amp; Contributions
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/biocypher/biocypher/edit/main/docs/learn/tutorials/tutorial_basics_neo4j_offline/tutorial_004_neo4j_offline.md" title="Edit this page" class="md-content__button md-icon" rel="edit">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


<h1 id="hands-on-building-graphs-with-biocypher-offline-mode-and-neo4j">🧑‍💻 Hands-on Building Graphs with BioCypher (offline mode) and Neo4j</h1>
<h2 id="overview">Overview</h2>
<p>This tutorial will help you get started with BioCypher in offline mode. You will learn how to create a simple knowledge graph with a synthetic dataset that contains information about proteins and its interactions.</p>
<p>By the end of this tutorial, you will be able to:</p>
<ul>
<li>Set up BioCypher for a basic project.</li>
<li>Explore a synthetic dataset and how to obtain a graph model from it.</li>
<li>Build a small knowledge graph from the data.</li>
<li>View and query the graph using Neo4j.</li>
</ul>
<h2 id="pre-requisites">Pre-requisites</h2>
<blockquote>
<p><strong>Note:</strong> Ensure you have the following prerequisites before continue with the tutorial.</p>
</blockquote>
<table>
<thead>
<tr>
<th>Tool</th>
<th>Version/Requirement</th>
<th>Installation Link</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td>Git</td>
<td>Any</td>
<td><a href="https://git-scm.com/downloads">Git Docs</a></td>
<td>For version control</td>
</tr>
<tr>
<td>Neo4j</td>
<td>&gt;=1.6</td>
<td><a href="https://neo4j.com/download/">Neo4j Desktop</a></td>
<td>For querying graphs</td>
</tr>
<tr>
<td>uv</td>
<td>&gt;=0.7.x</td>
<td><a href="https://docs.astral.sh/uv/getting-started/installation/">uv Docs</a></td>
<td>For dependency management</td>
</tr>
<tr>
<td>Python</td>
<td>&gt;= 3.11</td>
<td><a href="https://www.python.org/downloads/">Python.org</a></td>
<td>Required for BioCypher</td>
</tr>
<tr>
<td>Jupyter (optional)</td>
<td>Any</td>
<td><a href="https://jupyter.org/">Jupter</a></td>
<td>Required for exploring the sample data</td>
</tr>
</tbody>
</table>
<h2 id="setup">Setup</h2>
<h3 id="setup-python-project">Setup Python project</h3>
<p>In this section, you will set up your working environment using the BioCypher Project Template. This template provides a ready-to-use folder structure and example files, so you can focus on building your application.</p>
<p><strong>Steps:</strong></p>
<ol>
<li>
<p>Go to the <a href="https://github.com/biocypher/project-template">BioCypher Project Template</a> and click on "Use this template". Name your repository as <code>tutorial-basics-biocypher</code>. Then clone your repository to obtain a local copy:
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>git<span class="w"> </span>clone<span class="w"> </span>&lt;your-repository-url&gt;
</span></code></pre></div></p>
</li>
<li>
<p>Check the current project structure. Below, we list the most important folders and files for this tutorial.</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>/tutorial-basics-biocypher
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>    ├── config
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>    │   ├── biocypher_config.yaml
</span><span id="__span-1-4"><a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>    │   ├── biocypher_docker_config.yaml
</span><span id="__span-1-5"><a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>    │   └── schema_config.yaml
</span><span id="__span-1-6"><a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>    ├── create_knowledge_graph.py
</span><span id="__span-1-7"><a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>    ├── pyproject.toml
</span><span id="__span-1-8"><a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>    ├── template_package
</span><span id="__span-1-9"><a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>    │   └── adapters
</span><span id="__span-1-10"><a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a>    │       └── example_adapter.py
</span><span id="__span-1-11"><a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a>    ...
</span></code></pre></div>
</li>
<li>
<p>Install the dependencies using your preferred package manager (e.g. uv, Poetry or pip):</p>
</li>
</ol>
<p>You should always first create a dedicated Python environment for your project, and then install the dependencies into the environment. Environments can be managed by <a href="https://docs.conda.io/projects/conda/en/stable/user-guide/tasks/manage-environments.html">conda</a>, <a href="https://docs.astral.sh/uv/pip/environments/">uv</a> ,<a href="https://python-poetry.org/docs/managing-environments/">poetry</a> or <a href="https://docs.python.org/3/library/venv.html">venv</a>, for example.</p>
<p>After you have created your environment, activate the environment and install the required packages using your preferred package manager.</p>
<p><strong>Using uv: (recommended)</strong>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>uv<span class="w"> </span>sync
</span></code></pre></div></p>
<p><strong>Using Poetry:</strong>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>poetry<span class="w"> </span>install<span class="w"> </span>--no-root
</span></code></pre></div></p>
<p><strong>Using pip:</strong>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>pip<span class="w"> </span>install<span class="w"> </span>.
</span></code></pre></div></p>
<p>You also need to install Jupyter into your environment, i.e. <code>pip install jupyter</code>, if later you want to explore the sample data in a Jupyter notebook.</p>
<h3 id="setup-neo4j">Setup Neo4j</h3>
<p>In this section, we will create a Neo4j instance to use later in the tutorial. It is important to set this up now.</p>
<ol>
<li>
<p>Execute Neo4j Desktop, if this the first time you should see a window like this one.</p>
<p><figure markdown="span">
<img alt="Image title" src="../assets/neo4j_desktop_homepage.png" width="800" />
<figcaption>Figure 1. Neo4j Desktop start screen.</figcaption>
</figure></p>
</li>
<li>
<p>Create a new instance in Neo4j. For this tutorial, name it <code>neo4j-tutorial-instance</code> and choose a password you can remember.</p>
<p><figure markdown="span">
<img alt="Image title" src="../assets/neo4j_instance_creation.png" width="800" />
<figcaption>Figure 2. Create Instance window. This may vary depending on your Neo4j version.</figcaption>
</figure></p>
</li>
<li>
<p>Access details in the option <em>Overview</em>.</p>
<p><figure markdown="span">
<img alt="Image title" src="../assets/neo4j_overview_option.png" width="800" />
<figcaption>Figure 3. <em>Overview</em> option to check details related to your Neo4j instance.</figcaption>
</figure></p>
</li>
<li>
<p>Save the path to your Neo4j instance, we are going to use this path later in this tutorial.</p>
<p><figure markdown="span">
<img alt="Image title" src="../assets/neo4j_folder_details.png" width="800" />
<figcaption>Figure 4. Neo4j instance with its path location highlighted.</figcaption>
</figure></p>
</li>
</ol>
<h2 id="section-1-exploratory-data-analysis">Section 1. Exploratory Data Analysis</h2>
<p>For this tutorial we are going to use a <a href="https://zenodo.org/records/16902349">synthetic dataset</a> that contains information about the interaction between proteins. The dataset is contained in a <code>tsv</code> file, similar to a <code>csv</code> file but using tabs instead of commas as delimiters.</p>
<ul>
<li>
<p>First, download the dataset:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>./data/in/
</span><span id="__span-5-2"><a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>curl<span class="w"> </span>-o<span class="w"> </span>./data/in/synthetic_protein_interactions.tsv<span class="w"> </span><span class="se">\</span>
</span><span id="__span-5-3"><a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>https://zenodo.org/records/16902349/files/synthetic_protein_interactions.tsv
</span></code></pre></div>
</li>
<li>
<p>Create a folder called <code>notebooks</code> under <code>tutorial-basics-biocypher</code>
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>./notebooks/
</span></code></pre></div></p>
</li>
<li>
<p>Create and run either a Python file or a Jupyter notebook containing the following code.</p>
<details class="example">
<summary><strong>File: <code>notebooks/eda_synthetic_data.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
</span><span id="__span-7-2"><a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>
</span><span id="__span-7-3"><a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="c1"># Load the dataset</span>
</span><span id="__span-7-4"><a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s1">&#39;../data/in/synthetic_protein_interactions.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</span><span id="__span-7-5"><a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a>
</span><span id="__span-7-6"><a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a><span class="c1"># Show the first few rows</span>
</span><span id="__span-7-7"><a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">---- First 10 rows in the dataset&quot;</span><span class="p">)</span>
</span><span id="__span-7-8"><a id="__codelineno-7-8" name="__codelineno-7-8" href="#__codelineno-7-8"></a><span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
</span><span id="__span-7-9"><a id="__codelineno-7-9" name="__codelineno-7-9" href="#__codelineno-7-9"></a>
</span><span id="__span-7-10"><a id="__codelineno-7-10" name="__codelineno-7-10" href="#__codelineno-7-10"></a><span class="c1"># List the columns in the dataset</span>
</span><span id="__span-7-11"><a id="__codelineno-7-11" name="__codelineno-7-11" href="#__codelineno-7-11"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">---- Columns in the dataset&quot;</span><span class="p">)</span>
</span><span id="__span-7-12"><a id="__codelineno-7-12" name="__codelineno-7-12" href="#__codelineno-7-12"></a><span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
</span><span id="__span-7-13"><a id="__codelineno-7-13" name="__codelineno-7-13" href="#__codelineno-7-13"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-7-14"><a id="__codelineno-7-14" name="__codelineno-7-14" href="#__codelineno-7-14"></a>
</span><span id="__span-7-15"><a id="__codelineno-7-15" name="__codelineno-7-15" href="#__codelineno-7-15"></a><span class="c1"># Get basic info about the datasets</span>
</span><span id="__span-7-16"><a id="__codelineno-7-16" name="__codelineno-7-16" href="#__codelineno-7-16"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">---- Summary Dataframe&quot;</span><span class="p">)</span>
</span><span id="__span-7-17"><a id="__codelineno-7-17" name="__codelineno-7-17" href="#__codelineno-7-17"></a><span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">info</span><span class="p">())</span>
</span><span id="__span-7-18"><a id="__codelineno-7-18" name="__codelineno-7-18" href="#__codelineno-7-18"></a>
</span><span id="__span-7-19"><a id="__codelineno-7-19" name="__codelineno-7-19" href="#__codelineno-7-19"></a><span class="c1"># Check for missing values</span>
</span><span id="__span-7-20"><a id="__codelineno-7-20" name="__codelineno-7-20" href="#__codelineno-7-20"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">---- Check missing values&quot;</span><span class="p">)</span>
</span><span id="__span-7-21"><a id="__codelineno-7-21" name="__codelineno-7-21" href="#__codelineno-7-21"></a><span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
</span><span id="__span-7-22"><a id="__codelineno-7-22" name="__codelineno-7-22" href="#__codelineno-7-22"></a>
</span><span id="__span-7-23"><a id="__codelineno-7-23" name="__codelineno-7-23" href="#__codelineno-7-23"></a><span class="c1"># Show summary statistics for numeric columns</span>
</span><span id="__span-7-24"><a id="__codelineno-7-24" name="__codelineno-7-24" href="#__codelineno-7-24"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">---- Describe Dataframe statistics&quot;</span><span class="p">)</span>
</span><span id="__span-7-25"><a id="__codelineno-7-25" name="__codelineno-7-25" href="#__codelineno-7-25"></a><span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">describe</span><span class="p">())</span>
</span><span id="__span-7-26"><a id="__codelineno-7-26" name="__codelineno-7-26" href="#__codelineno-7-26"></a>
</span><span id="__span-7-27"><a id="__codelineno-7-27" name="__codelineno-7-27" href="#__codelineno-7-27"></a><span class="c1"># Count the unique number of proteins in each column</span>
</span><span id="__span-7-28"><a id="__codelineno-7-28" name="__codelineno-7-28" href="#__codelineno-7-28"></a><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">---- Number of unique proteins per column.&quot;</span><span class="p">)</span>
</span><span id="__span-7-29"><a id="__codelineno-7-29" name="__codelineno-7-29" href="#__codelineno-7-29"></a><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Unique proteins in column protein_a:&#39;</span><span class="p">,</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">())</span>
</span><span id="__span-7-30"><a id="__codelineno-7-30" name="__codelineno-7-30" href="#__codelineno-7-30"></a><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Unique proteins in column protein_b:&#39;</span><span class="p">,</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;target&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">())</span>
</span></code></pre></div>
</details>
</li>
</ul>
<blockquote>
<p>📝 <strong>Exercise:</strong></p>
<p>a. How many unique proteins do we have in the dataset? Hint: count unique proteins in the source and target columns.</p>
<p>b. How many interactions exist in our dataset? Hint: count unique interactions between sources and targets.</p>
<p>c. Some columns contain boolean values represented as "1" and "0". Can you detect which ones?</p>
</blockquote>
<details class="success">
<summary>Answer:</summary>
<p>a. Number of unique proteins: 15.</p>
<p>b. Number of unique interactions: there are 22 unique interactions. One interaction is repeated, with a difference in one of its properties.</p>
<p>c. <code>is_directed</code>, <code>is_stimulation</code>, <code>is_inhibition</code>, <code>consensus_direction</code>, <code>consensus_stimulation</code>,<code>consensus_inhibition</code>.</p>
</details>
<h2 id="section-2-graph-modeling">Section 2. Graph Modeling</h2>
<h3 id="graph-modeling">Graph Modeling</h3>
<p>By looking at the <code>tsv</code> file, we can see that there are two columns called <code>source</code> and <code>target</code>, which represent proteins. This means that each row represents an interaction between a source protein and a target protein. For now, our graph could look like this.</p>
<figure>
<img alt="Image title" src="../assets/model_graph_1.png" width="400" />

<figcaption>Figure 5. Simple graph model for representing interactions between proteins.</figcaption>
</figure>
<p>Can we improve the graph? Absolutely! Understanding the data is essential for building an effective graph. By examining the other columns in the table, we can identify additional details:</p>
<ul>
<li>
<p>The <code>source</code> and <code>target</code> columns represent <strong>nodes</strong> in the graph, with each node corresponding to a protein.</p>
</li>
<li>
<p>Each protein listed in the <code>source</code> column has associated properties found in other columns:</p>
<ul>
<li><code>source_genesymbol</code>: the gene symbol of the source protein.</li>
<li><code>ncbi_tax_id_source</code>: the NCBI taxonomy identifier of the source protein.</li>
<li><code>entity_type_source</code>: the type of entity for the source protein.</li>
</ul>
</li>
<li>
<p>Each protein in the <code>target</code> column has associated properties found in other columns:</p>
<ul>
<li><code>target_genesymbol</code></li>
<li><code>ncbi_tax_id_target</code></li>
<li><code>entity_type_target</code></li>
</ul>
</li>
</ul>
<figure>
<img alt="Image title" src="../assets/model_graph_2.png" width="400" />

<figcaption>Figure 6. Simple protein interaction graph with properties in nodes.</figcaption>
</figure>
<p>We know that a <code>source</code> protein interacts with a <code>target</code> protein, but do we know <strong>how</strong>?</p>
<p>Remaining columns in the table describe properties of these protein-protein interactions:</p>
<p><strong>Interaction properties</strong></p>
<ul>
<li><code>is_directed</code></li>
<li><code>is_stimulation</code></li>
<li><code>is_inhibition</code></li>
<li><code>consensus direction</code></li>
<li><code>consensus stimulation</code></li>
<li><code>consensus inhibition</code></li>
<li><code>type</code></li>
</ul>
<p>It is these protein-protein interactions that form the <strong>edges</strong> in the graph. Here, <code>is_directed</code>, <code>is_stimulation</code>, and <code>is_inhibition</code> describe properties that characterize each interaction <code>type</code>, while <code>consensus direction</code>, <code>consensus stimulation</code>, and <code>consensus inhibition</code> indicate the aggregated or consensus value derived from multiple sources in OmniPath for each property.</p>
<p>We are ready to model our second version of our graph. It is like follows:</p>
<figure>
<img alt="Image title" src="../assets/model_graph_3.png" width="400" />

<figcaption>Figure 7. Protein interaction graph showing node and edge properties.</figcaption>
</figure>
<p>Finally, we can model a more detailed graph using our dataset. Rather than representing all interactions in a generic way, we can use the <code>type</code> field to show the specific type of interaction occurring between each pair of proteins.</p>
<figure>
<img alt="Image title" src="../assets/model_graph_4.png" width="550" />

<figcaption>Figure 8. Graph model for representing different interactions between proteins.</figcaption>
</figure>
<h3 id="exercise-1-example-of-a-graph-we-expect-with-our-data">Exercise 1. Example of a graph we expect with our data</h3>
<blockquote>
<p>📝 <strong>Exercise:</strong>
Sketch a portion of the knowledge graph using the provided dataset.</p>
</blockquote>
<details class="success">
<summary>Answer:</summary>
<p>If you include all the nodes and edges from your TSV file, your sketch should look like the following example:</p>
<pre class="mermaid"><code>graph LR
    SOD1((SOD1))   -- binding --&gt; EGFR((EGFR))
    CDK1((CDK1))   -- activation --&gt; TP53((TP53))
    MYC((MYC))     -- phosporylation --&gt; GAPDH((GAPDH))
    MTOR((MTOR))   -- ubiquitination --&gt; NFKB1((NFKB1))
    NFKB1((NFKB1)) -- activation --&gt; CDK1((CDK1))
    MAPK1((MAPK1)) -- ubiquitination --&gt; HSP90((HSP90))
    TP53((TP53))     -- ubiquitination --&gt; CREB1((CREB1))
    HSP90((HSP90)) -- activation --&gt; APP((APP))
    HSP90((HSP90)) -- ubiquitination --&gt; RHOA((RHOA))
    SOD1((SOD1))   -- inhibition --&gt; TP53((TP53))
    AKT1((AKT1))   -- ubiquitination --&gt; HSP90((HSP90))
    HSP90((HSP90)) -- ubiquitination --&gt; MYC((MYC))
    MAPK1((MAPK1)) -- ubiquitination --&gt; BRCA1((BRCA1))
    NFKB1((NFKB1)) -- inhibition --&gt; RHOA((RHOA))
    NFKB1((NFKB1)) -- phosphorylation --&gt; APP((APP))
    HSP90((HSP90)) -- binding --&gt; GAPDH((GAPDH))
    GAPDH((GAPDH)) -- activation --&gt; TP53((TP53))
    AKT1((AKT1))   -- phosphorylation --&gt; TP53((TP53))
    GAPDH((GAPDH)) -- activation --&gt; APP((APP))
    TP53((TP53))     -- activation --&gt; MAPK1((MAPK1))
    TP53((TP53))     -- ubiquitination --&gt; CREB1((CREB1))
    MYC((MYC))     -- phosphorylation --&gt; GAPDH((GAPDH))
    EGFR((EGFR))   -- binding --&gt; SOD1((SOD1))</code></pre>
</details>
<h2 id="section-3-graph-creation-with-biocypher">Section 3. Graph creation with BioCypher</h2>
<p>We aim to create a knowledge graph using the data we found in the <code>tsv</code> file. Let's recap our exercise:</p>
<ul>
<li>
<p>Create a <strong>graph</strong> with the following characteristics:</p>
<ul>
<li>One node type: <code>Protein</code>.</li>
<li>Five edge types: <code>activation</code>, <code>binding</code>, <code>inhibition</code>, <code>phosphorylation</code>, <code>ubiquitination</code>.</li>
</ul>
</li>
<li>
<p>Each <strong>node</strong> has properties:</p>
<ul>
<li><em>genesymbol</em></li>
<li><em>ncbi_tax_id</em></li>
<li><em>entity_type</em></li>
</ul>
</li>
<li>
<p>Each <strong>edge</strong> has properties:</p>
<ul>
<li><em>is_directed</em></li>
<li><em>is_stimulation</em></li>
<li><em>is_inhibition</em></li>
<li><em>consensus_direction</em></li>
<li><em>consensus_stimulation</em></li>
<li><em>consensus_inhibition</em></li>
</ul>
</li>
<li>
<p>We must export the knowledge graph to Neo4j.</p>
</li>
</ul>
<p>To achieve this, we can divide the process into three sections:</p>
<ol>
<li>
<p><a href="#step-1-configuration">Configuration</a>.</p>
<ul>
<li><a href="#create-a-schema-for-your-graph">Schema configuration</a></li>
<li><a href="#configure-biocypher-behavior">BioCypher configuration</a></li>
</ul>
</li>
<li>
<p><a href="#step-2-create-an-adapter">Adapter creation</a>.</p>
<ul>
<li>Read/connect to input data</li>
<li>Process data</li>
<li>Stream processed data</li>
</ul>
</li>
<li>
<p><a href="#step-3-create-a-knowledge-graph-script">Knowledge Graph script</a></p>
</li>
</ol>
<h3 id="step-1-configuration">Step 1. Configuration</h3>
<figure>
<img alt="Image title" src="../assets/biocypher_section_conf.png" width="1000" />

<figcaption>Figure 9. Configuration step in the BioCypher pipeline.</figcaption>
</figure>
<h4 id="create-a-schema-for-your-graph">Create a schema for your graph</h4>
<p><strong>Rationale:</strong> the schema file allows us to define the skeleton for our knowledge graph. Nodes, edges, properties are defined here.</p>
<p>The following is an example of how our schema file should look like, all of this is based on how we defined the graph structure (nodes, edges and their properties).</p>
<details class="example">
<summary><strong>File: <code>config/schema_config.yaml</code></strong></summary>
<div class="language-yaml highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-8-2"><a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a><span class="c1">#-------------------------      NODES      -------------------------</span>
</span><span id="__span-8-3"><a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-8-4"><a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a><span class="c1">#=========    PARENT NODES</span>
</span><span id="__span-8-5"><a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a><span class="nt">protein</span><span class="p">:</span>
</span><span id="__span-8-6"><a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">node</span>
</span><span id="__span-8-7"><a id="__codelineno-8-7" name="__codelineno-8-7" href="#__codelineno-8-7"></a><span class="w">    </span><span class="nt">preferred_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">uniprot</span>
</span><span id="__span-8-8"><a id="__codelineno-8-8" name="__codelineno-8-8" href="#__codelineno-8-8"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">uniprot_protein</span>
</span><span id="__span-8-9"><a id="__codelineno-8-9" name="__codelineno-8-9" href="#__codelineno-8-9"></a>
</span><span id="__span-8-10"><a id="__codelineno-8-10" name="__codelineno-8-10" href="#__codelineno-8-10"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-8-11"><a id="__codelineno-8-11" name="__codelineno-8-11" href="#__codelineno-8-11"></a><span class="c1">#------------------      RELATIONSHIPS (EDGES)     -----------------</span>
</span><span id="__span-8-12"><a id="__codelineno-8-12" name="__codelineno-8-12" href="#__codelineno-8-12"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-8-13"><a id="__codelineno-8-13" name="__codelineno-8-13" href="#__codelineno-8-13"></a><span class="c1">#=========    PARENT EDGES</span>
</span><span id="__span-8-14"><a id="__codelineno-8-14" name="__codelineno-8-14" href="#__codelineno-8-14"></a><span class="nt">protein protein interaction</span><span class="p">:</span>
</span><span id="__span-8-15"><a id="__codelineno-8-15" name="__codelineno-8-15" href="#__codelineno-8-15"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pairwise molecular interaction</span>
</span><span id="__span-8-16"><a id="__codelineno-8-16" name="__codelineno-8-16" href="#__codelineno-8-16"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-8-17"><a id="__codelineno-8-17" name="__codelineno-8-17" href="#__codelineno-8-17"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein_protein_interaction</span>
</span><span id="__span-8-18"><a id="__codelineno-8-18" name="__codelineno-8-18" href="#__codelineno-8-18"></a><span class="w">    </span><span class="nt">properties</span><span class="p">:</span>
</span><span id="__span-8-19"><a id="__codelineno-8-19" name="__codelineno-8-19" href="#__codelineno-8-19"></a><span class="w">        </span><span class="nt">is_directed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-8-20"><a id="__codelineno-8-20" name="__codelineno-8-20" href="#__codelineno-8-20"></a><span class="w">        </span><span class="nt">is_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-8-21"><a id="__codelineno-8-21" name="__codelineno-8-21" href="#__codelineno-8-21"></a><span class="w">        </span><span class="nt">is_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-8-22"><a id="__codelineno-8-22" name="__codelineno-8-22" href="#__codelineno-8-22"></a><span class="w">        </span><span class="nt">consensus_direction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-8-23"><a id="__codelineno-8-23" name="__codelineno-8-23" href="#__codelineno-8-23"></a><span class="w">        </span><span class="nt">consensus_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-8-24"><a id="__codelineno-8-24" name="__codelineno-8-24" href="#__codelineno-8-24"></a><span class="w">        </span><span class="nt">consensus_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-8-25"><a id="__codelineno-8-25" name="__codelineno-8-25" href="#__codelineno-8-25"></a>
</span><span id="__span-8-26"><a id="__codelineno-8-26" name="__codelineno-8-26" href="#__codelineno-8-26"></a><span class="c1">#=========    INHERITED EDGES</span>
</span><span id="__span-8-27"><a id="__codelineno-8-27" name="__codelineno-8-27" href="#__codelineno-8-27"></a><span class="nt">binding</span><span class="p">:</span>
</span><span id="__span-8-28"><a id="__codelineno-8-28" name="__codelineno-8-28" href="#__codelineno-8-28"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein protein interaction</span>
</span><span id="__span-8-29"><a id="__codelineno-8-29" name="__codelineno-8-29" href="#__codelineno-8-29"></a><span class="w">    </span><span class="nt">inherit_properties</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-8-30"><a id="__codelineno-8-30" name="__codelineno-8-30" href="#__codelineno-8-30"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-8-31"><a id="__codelineno-8-31" name="__codelineno-8-31" href="#__codelineno-8-31"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">binding</span>
</span><span id="__span-8-32"><a id="__codelineno-8-32" name="__codelineno-8-32" href="#__codelineno-8-32"></a>
</span><span id="__span-8-33"><a id="__codelineno-8-33" name="__codelineno-8-33" href="#__codelineno-8-33"></a><span class="c1"># ...rest of schema_config.yaml omitted for brevity...</span>
</span></code></pre></div>
</details>
<h5 id="nodes">Nodes</h5>
<p>The <code>protein</code> top-level key in the YAML snippet identifies our entity and connects it to the ontological backbone.</p>
<table>
<thead>
<tr>
<th>Key</th>
<th>Value</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>represented_as</code></td>
<td><code>node</code></td>
<td>Specifies how BioCypher should represent each entity in the graph; in this case, as a node.</td>
</tr>
<tr>
<td><code>preferred_id</code></td>
<td><code>uniprot</code></td>
<td>Defines a namespace for our proteins. In this example, all proteins follow the UniProt convention—a 5-character alphanumeric string (e.g., P00533).</td>
</tr>
<tr>
<td><code>input_label</code></td>
<td><code>uniprot_protein</code></td>
<td>Indicates the expected label in the node tuple. All other input nodes without this label are ignored unless they are defined in the schema configuration.</td>
</tr>
</tbody>
</table>
<p>For more information about which other keywords you can use to configure your nodes in the schema file consult <a href="https://biocypher.org/BioCypher/reference/schema-config/#fields-reference">Fields reference</a>.</p>
<h5 id="edges-relationships">Edges (relationships)</h5>
<p>As shown in <a href="#graph-modeling">Figure 7</a>, each edge has the same set of properties (<code>is_directed</code>, <code>consensus_direction</code>, etc.). At this stage, we have two options for defining the edges:</p>
<ul>
<li>Option 1: Create each edge and explicitly define the same set of property fields for every edge.</li>
</ul>
<details class="example">
<summary><strong>File: <code>config/schema_config.yaml</code></strong></summary>
<div class="language-yaml highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-9-2"><a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a><span class="c1">#------------------      RELATIONSHIPS (EDGES)     -----------------</span>
</span><span id="__span-9-3"><a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-9-4"><a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a><span class="nt">activation</span><span class="p">:</span>
</span><span id="__span-9-5"><a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pairwise molecular interaction</span>
</span><span id="__span-9-6"><a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-9-7"><a id="__codelineno-9-7" name="__codelineno-9-7" href="#__codelineno-9-7"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein_protein_interaction</span>
</span><span id="__span-9-8"><a id="__codelineno-9-8" name="__codelineno-9-8" href="#__codelineno-9-8"></a><span class="w">    </span><span class="nt">properties</span><span class="p">:</span>
</span><span id="__span-9-9"><a id="__codelineno-9-9" name="__codelineno-9-9" href="#__codelineno-9-9"></a><span class="w">        </span><span class="nt">is_directed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-10"><a id="__codelineno-9-10" name="__codelineno-9-10" href="#__codelineno-9-10"></a><span class="w">        </span><span class="nt">is_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-11"><a id="__codelineno-9-11" name="__codelineno-9-11" href="#__codelineno-9-11"></a><span class="w">        </span><span class="nt">is_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-12"><a id="__codelineno-9-12" name="__codelineno-9-12" href="#__codelineno-9-12"></a><span class="w">        </span><span class="nt">consensus_direction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-13"><a id="__codelineno-9-13" name="__codelineno-9-13" href="#__codelineno-9-13"></a><span class="w">        </span><span class="nt">consensus_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-14"><a id="__codelineno-9-14" name="__codelineno-9-14" href="#__codelineno-9-14"></a><span class="w">        </span><span class="nt">consensus_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-15"><a id="__codelineno-9-15" name="__codelineno-9-15" href="#__codelineno-9-15"></a>
</span><span id="__span-9-16"><a id="__codelineno-9-16" name="__codelineno-9-16" href="#__codelineno-9-16"></a><span class="nt">binding</span><span class="p">:</span>
</span><span id="__span-9-17"><a id="__codelineno-9-17" name="__codelineno-9-17" href="#__codelineno-9-17"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pairwise molecular interaction</span>
</span><span id="__span-9-18"><a id="__codelineno-9-18" name="__codelineno-9-18" href="#__codelineno-9-18"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-9-19"><a id="__codelineno-9-19" name="__codelineno-9-19" href="#__codelineno-9-19"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein_protein_interaction</span>
</span><span id="__span-9-20"><a id="__codelineno-9-20" name="__codelineno-9-20" href="#__codelineno-9-20"></a><span class="w">    </span><span class="nt">properties</span><span class="p">:</span>
</span><span id="__span-9-21"><a id="__codelineno-9-21" name="__codelineno-9-21" href="#__codelineno-9-21"></a><span class="w">        </span><span class="nt">is_directed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-22"><a id="__codelineno-9-22" name="__codelineno-9-22" href="#__codelineno-9-22"></a><span class="w">        </span><span class="nt">is_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-23"><a id="__codelineno-9-23" name="__codelineno-9-23" href="#__codelineno-9-23"></a><span class="w">        </span><span class="nt">is_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-24"><a id="__codelineno-9-24" name="__codelineno-9-24" href="#__codelineno-9-24"></a><span class="w">        </span><span class="nt">consensus_direction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-25"><a id="__codelineno-9-25" name="__codelineno-9-25" href="#__codelineno-9-25"></a><span class="w">        </span><span class="nt">consensus_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-26"><a id="__codelineno-9-26" name="__codelineno-9-26" href="#__codelineno-9-26"></a><span class="w">        </span><span class="nt">consensus_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-9-27"><a id="__codelineno-9-27" name="__codelineno-9-27" href="#__codelineno-9-27"></a>
</span><span id="__span-9-28"><a id="__codelineno-9-28" name="__codelineno-9-28" href="#__codelineno-9-28"></a><span class="c1"># ...rest of schema_config.yaml omitted for brevity...</span>
</span></code></pre></div>
</details>
<ul>
<li>Option 2 (<strong>recommended</strong>): Create a base edge with the properties, and then create edges that inherit the behavior of this base edge. This approach reduces lines of code and avoids repetition. For example, if you have more than 20 edges, Option 1 would likely not be practical.</li>
</ul>
<details class="example">
<summary><strong>File: <code>config/schema_config.yaml</code></strong></summary>
<div class="language-yaml highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-10-2"><a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a><span class="c1">#------------------      RELATIONSHIPS (EDGES)     -----------------</span>
</span><span id="__span-10-3"><a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-10-4"><a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a><span class="c1">#====   BASE EDGE or PARENT EDGE</span>
</span><span id="__span-10-5"><a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a><span class="nt">protein protein interaction</span><span class="p">:</span>
</span><span id="__span-10-6"><a id="__codelineno-10-6" name="__codelineno-10-6" href="#__codelineno-10-6"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pairwise molecular interaction</span>
</span><span id="__span-10-7"><a id="__codelineno-10-7" name="__codelineno-10-7" href="#__codelineno-10-7"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-10-8"><a id="__codelineno-10-8" name="__codelineno-10-8" href="#__codelineno-10-8"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein_protein_interaction</span>
</span><span id="__span-10-9"><a id="__codelineno-10-9" name="__codelineno-10-9" href="#__codelineno-10-9"></a><span class="w">    </span><span class="nt">properties</span><span class="p">:</span>
</span><span id="__span-10-10"><a id="__codelineno-10-10" name="__codelineno-10-10" href="#__codelineno-10-10"></a><span class="w">        </span><span class="nt">is_directed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-10-11"><a id="__codelineno-10-11" name="__codelineno-10-11" href="#__codelineno-10-11"></a><span class="w">        </span><span class="nt">is_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-10-12"><a id="__codelineno-10-12" name="__codelineno-10-12" href="#__codelineno-10-12"></a><span class="w">        </span><span class="nt">is_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-10-13"><a id="__codelineno-10-13" name="__codelineno-10-13" href="#__codelineno-10-13"></a><span class="w">        </span><span class="nt">consensus_direction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-10-14"><a id="__codelineno-10-14" name="__codelineno-10-14" href="#__codelineno-10-14"></a><span class="w">        </span><span class="nt">consensus_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-10-15"><a id="__codelineno-10-15" name="__codelineno-10-15" href="#__codelineno-10-15"></a><span class="w">        </span><span class="nt">consensus_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-10-16"><a id="__codelineno-10-16" name="__codelineno-10-16" href="#__codelineno-10-16"></a>
</span><span id="__span-10-17"><a id="__codelineno-10-17" name="__codelineno-10-17" href="#__codelineno-10-17"></a><span class="c1">#====   INHERITED EDGES</span>
</span><span id="__span-10-18"><a id="__codelineno-10-18" name="__codelineno-10-18" href="#__codelineno-10-18"></a><span class="nt">activation</span><span class="p">:</span>
</span><span id="__span-10-19"><a id="__codelineno-10-19" name="__codelineno-10-19" href="#__codelineno-10-19"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein protein interaction</span>
</span><span id="__span-10-20"><a id="__codelineno-10-20" name="__codelineno-10-20" href="#__codelineno-10-20"></a><span class="w">    </span><span class="nt">inherit_properties</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-10-21"><a id="__codelineno-10-21" name="__codelineno-10-21" href="#__codelineno-10-21"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-10-22"><a id="__codelineno-10-22" name="__codelineno-10-22" href="#__codelineno-10-22"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">activation</span>
</span><span id="__span-10-23"><a id="__codelineno-10-23" name="__codelineno-10-23" href="#__codelineno-10-23"></a>
</span><span id="__span-10-24"><a id="__codelineno-10-24" name="__codelineno-10-24" href="#__codelineno-10-24"></a><span class="nt">binding</span><span class="p">:</span>
</span><span id="__span-10-25"><a id="__codelineno-10-25" name="__codelineno-10-25" href="#__codelineno-10-25"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein protein interaction</span>
</span><span id="__span-10-26"><a id="__codelineno-10-26" name="__codelineno-10-26" href="#__codelineno-10-26"></a><span class="w">    </span><span class="nt">inherit_properties</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-10-27"><a id="__codelineno-10-27" name="__codelineno-10-27" href="#__codelineno-10-27"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-10-28"><a id="__codelineno-10-28" name="__codelineno-10-28" href="#__codelineno-10-28"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">binding</span>
</span><span id="__span-10-29"><a id="__codelineno-10-29" name="__codelineno-10-29" href="#__codelineno-10-29"></a>
</span><span id="__span-10-30"><a id="__codelineno-10-30" name="__codelineno-10-30" href="#__codelineno-10-30"></a><span class="c1"># ...rest of schema_config.yaml omitted for brevity...</span>
</span></code></pre></div>
</details>
<p>Let's explain the keys and values for the second case (Option 2), because we are going to use the second option approach.</p>
<p><strong>Base Edge</strong>
The <code>protein protein interaction</code> top-level key in the YAML snippet identifies our edge entity.</p>
<table>
<thead>
<tr>
<th>Key</th>
<th>Value</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>is_a</code></td>
<td><code>pairwise molecular interaction</code></td>
<td>Defines the type of entity based on the ontology.</td>
</tr>
<tr>
<td><code>represented_as</code></td>
<td><code>edge</code></td>
<td>Explicitly specifies that this entity is an edge.</td>
</tr>
<tr>
<td><code>input_label</code></td>
<td><code>protein_protein_interaction</code></td>
<td>Defines a namespace for our relationships.</td>
</tr>
<tr>
<td><code>properties</code></td>
<td><em>property</em>: <em>datatype</em> (i.e. <code>is_directed: bool</code>)</td>
<td>Contains all properties associated with this edge; each property has a name and an associated datatype.</td>
</tr>
</tbody>
</table>
<p><strong>Inherited Edges</strong>
The <code>activation:</code> top-level key in the YAML snippet identifies our edge entity.</p>
<table>
<thead>
<tr>
<th>Key</th>
<th>Value</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>is_a</code></td>
<td><code>protein protein interaction</code></td>
<td>Defines the type of entity; in this case, it is a child of the base edge we defined previously.</td>
</tr>
<tr>
<td><code>inherit_properties</code></td>
<td><code>true</code></td>
<td>Indicates whether all properties defined in the base edge should be inherited.</td>
</tr>
<tr>
<td><code>represented_as</code></td>
<td><code>edge</code></td>
<td>Specifies that BioCypher will treat this entity (<code>activation</code>) as an edge.</td>
</tr>
<tr>
<td><code>input_label</code></td>
<td><code>binding</code></td>
<td>Specifies the expected edge label; edges without this label are ignored unless defined in the schema.</td>
</tr>
</tbody>
</table>
<h4 id="a-comment-about-the-connection-between-biocypher-and-ontologies">A comment about the connection between BioCypher and Ontologies</h4>
<p>In BioCypher, ontologies are integrated through the schema configuration file. This YAML file defines the structure of the graph by specifying which entities and relationships should be included. At the same time, it links those entities to the biomedical domain by aligning them with an ontological hierarchy. In this tutorial, we use the Biolink Model as the backbone of that hierarchy. The guiding principle is simple: only entities that are defined in the schema configuration <strong>and</strong> present in the input data are incorporated into the final knowledge graph.</p>
<p>Figure 10 illustrates the Biolink Model and some of its components organized in a hierarchy. Notice that entities such as <em>protein</em> (nodes) and <em>pairwise molecular interaction</em> (edges) appear both in the schema configuration and in the ontology. This alignment ensures that BioCypher graphs are not only structured consistently but also grounded in standardized biomedical concepts. For a deeper exploration of ontologies in BioCypher, see our <a href="https://biocypher.org/BioCypher/learn/tutorials/tutorial002_handling_ontologies/">ontology tutorial</a>.</p>
<figure>
<img alt="Image title" src="../assets/biolink_ontology.png" width="1000" />

<figcaption>Figure 10. The Biolink Model as an ontology backbone. On the right, <b>protein</b> is represented as an entity; on the left, <b>pairwise molecular interaction</b> is defined as an association. Together, these demonstrate how the schema anchors graph components to standardized biomedical concepts.</figcaption>
</figure>
<blockquote>
<p>📝 <strong>Exercise:</strong>
Revise and complete the <code>schema_config.yaml</code> file, and make sure it is located in the <code>config</code> folder.</p>
</blockquote>
<details class="success">
<summary>Answer:</summary>
<p>See the example below for a completed <code>schema_config.yaml</code>.</p>
<p><strong>File: <code>config/schema_config.yaml</code></strong></p>
<div class="language-yaml highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-11-2"><a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a><span class="c1">#-------------------------      NODES      -------------------------</span>
</span><span id="__span-11-3"><a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-11-4"><a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a><span class="c1">#====   PARENT NODES</span>
</span><span id="__span-11-5"><a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a><span class="nt">protein</span><span class="p">:</span>
</span><span id="__span-11-6"><a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">node</span>
</span><span id="__span-11-7"><a id="__codelineno-11-7" name="__codelineno-11-7" href="#__codelineno-11-7"></a><span class="w">    </span><span class="nt">preferred_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">uniprot</span>
</span><span id="__span-11-8"><a id="__codelineno-11-8" name="__codelineno-11-8" href="#__codelineno-11-8"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">uniprot_protein</span>
</span><span id="__span-11-9"><a id="__codelineno-11-9" name="__codelineno-11-9" href="#__codelineno-11-9"></a>
</span><span id="__span-11-10"><a id="__codelineno-11-10" name="__codelineno-11-10" href="#__codelineno-11-10"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-11-11"><a id="__codelineno-11-11" name="__codelineno-11-11" href="#__codelineno-11-11"></a><span class="c1">#------------------      RELATIONSHIPS (EDGES)     -----------------</span>
</span><span id="__span-11-12"><a id="__codelineno-11-12" name="__codelineno-11-12" href="#__codelineno-11-12"></a><span class="c1">#-------------------------------------------------------------------</span>
</span><span id="__span-11-13"><a id="__codelineno-11-13" name="__codelineno-11-13" href="#__codelineno-11-13"></a><span class="c1">#====   PARENT EDGES</span>
</span><span id="__span-11-14"><a id="__codelineno-11-14" name="__codelineno-11-14" href="#__codelineno-11-14"></a><span class="nt">protein protein interaction</span><span class="p">:</span>
</span><span id="__span-11-15"><a id="__codelineno-11-15" name="__codelineno-11-15" href="#__codelineno-11-15"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pairwise molecular interaction</span>
</span><span id="__span-11-16"><a id="__codelineno-11-16" name="__codelineno-11-16" href="#__codelineno-11-16"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-11-17"><a id="__codelineno-11-17" name="__codelineno-11-17" href="#__codelineno-11-17"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein_protein_interaction</span>
</span><span id="__span-11-18"><a id="__codelineno-11-18" name="__codelineno-11-18" href="#__codelineno-11-18"></a><span class="w">    </span><span class="nt">properties</span><span class="p">:</span>
</span><span id="__span-11-19"><a id="__codelineno-11-19" name="__codelineno-11-19" href="#__codelineno-11-19"></a><span class="w">        </span><span class="nt">is_directed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-11-20"><a id="__codelineno-11-20" name="__codelineno-11-20" href="#__codelineno-11-20"></a><span class="w">        </span><span class="nt">is_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-11-21"><a id="__codelineno-11-21" name="__codelineno-11-21" href="#__codelineno-11-21"></a><span class="w">        </span><span class="nt">is_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-11-22"><a id="__codelineno-11-22" name="__codelineno-11-22" href="#__codelineno-11-22"></a><span class="w">        </span><span class="nt">consensus_direction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-11-23"><a id="__codelineno-11-23" name="__codelineno-11-23" href="#__codelineno-11-23"></a><span class="w">        </span><span class="nt">consensus_stimulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-11-24"><a id="__codelineno-11-24" name="__codelineno-11-24" href="#__codelineno-11-24"></a><span class="w">        </span><span class="nt">consensus_inhibition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">bool</span>
</span><span id="__span-11-25"><a id="__codelineno-11-25" name="__codelineno-11-25" href="#__codelineno-11-25"></a>
</span><span id="__span-11-26"><a id="__codelineno-11-26" name="__codelineno-11-26" href="#__codelineno-11-26"></a><span class="c1">#====   INHERITED EDGES</span>
</span><span id="__span-11-27"><a id="__codelineno-11-27" name="__codelineno-11-27" href="#__codelineno-11-27"></a><span class="nt">activation</span><span class="p">:</span>
</span><span id="__span-11-28"><a id="__codelineno-11-28" name="__codelineno-11-28" href="#__codelineno-11-28"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein protein interaction</span>
</span><span id="__span-11-29"><a id="__codelineno-11-29" name="__codelineno-11-29" href="#__codelineno-11-29"></a><span class="w">    </span><span class="nt">inherit_properties</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-11-30"><a id="__codelineno-11-30" name="__codelineno-11-30" href="#__codelineno-11-30"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-11-31"><a id="__codelineno-11-31" name="__codelineno-11-31" href="#__codelineno-11-31"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">activation</span>
</span><span id="__span-11-32"><a id="__codelineno-11-32" name="__codelineno-11-32" href="#__codelineno-11-32"></a>
</span><span id="__span-11-33"><a id="__codelineno-11-33" name="__codelineno-11-33" href="#__codelineno-11-33"></a><span class="nt">binding</span><span class="p">:</span>
</span><span id="__span-11-34"><a id="__codelineno-11-34" name="__codelineno-11-34" href="#__codelineno-11-34"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein protein interaction</span>
</span><span id="__span-11-35"><a id="__codelineno-11-35" name="__codelineno-11-35" href="#__codelineno-11-35"></a><span class="w">    </span><span class="nt">inherit_properties</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-11-36"><a id="__codelineno-11-36" name="__codelineno-11-36" href="#__codelineno-11-36"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-11-37"><a id="__codelineno-11-37" name="__codelineno-11-37" href="#__codelineno-11-37"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">binding</span>
</span><span id="__span-11-38"><a id="__codelineno-11-38" name="__codelineno-11-38" href="#__codelineno-11-38"></a>
</span><span id="__span-11-39"><a id="__codelineno-11-39" name="__codelineno-11-39" href="#__codelineno-11-39"></a><span class="nt">inhibition</span><span class="p">:</span>
</span><span id="__span-11-40"><a id="__codelineno-11-40" name="__codelineno-11-40" href="#__codelineno-11-40"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein protein interaction</span>
</span><span id="__span-11-41"><a id="__codelineno-11-41" name="__codelineno-11-41" href="#__codelineno-11-41"></a><span class="w">    </span><span class="nt">inherit_properties</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-11-42"><a id="__codelineno-11-42" name="__codelineno-11-42" href="#__codelineno-11-42"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-11-43"><a id="__codelineno-11-43" name="__codelineno-11-43" href="#__codelineno-11-43"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">inhibition</span>
</span><span id="__span-11-44"><a id="__codelineno-11-44" name="__codelineno-11-44" href="#__codelineno-11-44"></a>
</span><span id="__span-11-45"><a id="__codelineno-11-45" name="__codelineno-11-45" href="#__codelineno-11-45"></a><span class="nt">phosphorylation</span><span class="p">:</span>
</span><span id="__span-11-46"><a id="__codelineno-11-46" name="__codelineno-11-46" href="#__codelineno-11-46"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein protein interaction</span>
</span><span id="__span-11-47"><a id="__codelineno-11-47" name="__codelineno-11-47" href="#__codelineno-11-47"></a><span class="w">    </span><span class="nt">inherit_properties</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-11-48"><a id="__codelineno-11-48" name="__codelineno-11-48" href="#__codelineno-11-48"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-11-49"><a id="__codelineno-11-49" name="__codelineno-11-49" href="#__codelineno-11-49"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">phosphorylation</span>
</span><span id="__span-11-50"><a id="__codelineno-11-50" name="__codelineno-11-50" href="#__codelineno-11-50"></a>
</span><span id="__span-11-51"><a id="__codelineno-11-51" name="__codelineno-11-51" href="#__codelineno-11-51"></a><span class="nt">ubiquitination</span><span class="p">:</span>
</span><span id="__span-11-52"><a id="__codelineno-11-52" name="__codelineno-11-52" href="#__codelineno-11-52"></a><span class="w">    </span><span class="nt">is_a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">protein protein interaction</span>
</span><span id="__span-11-53"><a id="__codelineno-11-53" name="__codelineno-11-53" href="#__codelineno-11-53"></a><span class="w">    </span><span class="nt">inherit_properties</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-11-54"><a id="__codelineno-11-54" name="__codelineno-11-54" href="#__codelineno-11-54"></a><span class="w">    </span><span class="nt">represented_as</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edge</span>
</span><span id="__span-11-55"><a id="__codelineno-11-55" name="__codelineno-11-55" href="#__codelineno-11-55"></a><span class="w">    </span><span class="nt">input_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ubiquitination</span>
</span></code></pre></div>
</details>
<h4 id="configure-biocypher-behavior">Configure BioCypher behavior</h4>
<p><strong>Rationale:</strong> The purpose of writing a <code>biocypher_config.yaml</code> is to define how BioCypher should operate for your project—specifying settings for data import, graph creation, and database interaction—all in one place for clarity and easy customization.</p>
<details class="example">
<summary><strong>File: <code>config/biocypher_config.yaml</code></strong></summary>
<div class="language-yaml highlight"><pre><span></span><code><span id="__span-12-1"><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a><span class="c1">#---------------------------------------------------------------</span>
</span><span id="__span-12-2"><a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a><span class="c1">#--------        BIOCYPHER GENERAL CONFIGURATION        --------</span>
</span><span id="__span-12-3"><a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a><span class="c1">#---------------------------------------------------------------</span>
</span><span id="__span-12-4"><a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a><span class="nt">biocypher</span><span class="p">:</span>
</span><span id="__span-12-5"><a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a><span class="w">    </span><span class="nt">offline</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-12-6"><a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a><span class="w">    </span><span class="nt">debug</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
</span><span id="__span-12-7"><a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a><span class="w">    </span><span class="nt">schema_config_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/schema_config.yaml</span>
</span><span id="__span-12-8"><a id="__codelineno-12-8" name="__codelineno-12-8" href="#__codelineno-12-8"></a><span class="w">    </span><span class="nt">cache_directory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">.cache</span>
</span><span id="__span-12-9"><a id="__codelineno-12-9" name="__codelineno-12-9" href="#__codelineno-12-9"></a>
</span><span id="__span-12-10"><a id="__codelineno-12-10" name="__codelineno-12-10" href="#__codelineno-12-10"></a><span class="c1">#----------------------------------------------------</span>
</span><span id="__span-12-11"><a id="__codelineno-12-11" name="__codelineno-12-11" href="#__codelineno-12-11"></a><span class="c1">#--------        OUTPUT CONFIGURATION        --------</span>
</span><span id="__span-12-12"><a id="__codelineno-12-12" name="__codelineno-12-12" href="#__codelineno-12-12"></a><span class="c1">#----------------------------------------------------</span>
</span><span id="__span-12-13"><a id="__codelineno-12-13" name="__codelineno-12-13" href="#__codelineno-12-13"></a><span class="nt">neo4j</span><span class="p">:</span>
</span><span id="__span-12-14"><a id="__codelineno-12-14" name="__codelineno-12-14" href="#__codelineno-12-14"></a><span class="w">    </span><span class="nt">database_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">neo4j</span>
</span><span id="__span-12-15"><a id="__codelineno-12-15" name="__codelineno-12-15" href="#__codelineno-12-15"></a><span class="w">    </span><span class="nt">delimiter</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;\t&#39;</span>
</span><span id="__span-12-16"><a id="__codelineno-12-16" name="__codelineno-12-16" href="#__codelineno-12-16"></a><span class="w">    </span><span class="nt">array_delimiter</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;|&#39;</span>
</span><span id="__span-12-17"><a id="__codelineno-12-17" name="__codelineno-12-17" href="#__codelineno-12-17"></a><span class="w">    </span><span class="nt">skip_duplicate_nodes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-12-18"><a id="__codelineno-12-18" name="__codelineno-12-18" href="#__codelineno-12-18"></a><span class="w">    </span><span class="nt">skip_bad_relationships</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-12-19"><a id="__codelineno-12-19" name="__codelineno-12-19" href="#__codelineno-12-19"></a><span class="w">    </span><span class="nt">import_call_bin_prefix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path to your Neo4j instance from Setup Neo4j section&gt;/bin/</span>
</span></code></pre></div>
</details>
<p>The first block is the BioCypher Core Settings, which starts with <code>biocypher:</code></p>
<table>
<thead>
<tr>
<th>key</th>
<th>value</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>offline</code></td>
<td><code>true</code></td>
<td>Whether to run in offline mode (no running DBMS or in-memory object)</td>
</tr>
<tr>
<td><code>debug</code></td>
<td><code>false</code></td>
<td>Whether to enable debug logging</td>
</tr>
<tr>
<td><code>schema_config_path</code></td>
<td><code>config/schema_config.yaml</code></td>
<td>Path to the schema configuration file</td>
</tr>
<tr>
<td><code>cache_directory</code></td>
<td><code>.cache</code></td>
<td>Path to the schema configuration file</td>
</tr>
</tbody>
</table>
<p>The second block is the Database Management System Settings, which starts with the name of the DBMS, in this case it's <code>neo4j:</code></p>
<table>
<thead>
<tr>
<th>key</th>
<th>value</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>delimiter</code></td>
<td><code>'\t'</code></td>
<td>Field delimiter for TSV import files</td>
</tr>
<tr>
<td><code>array_delimiter</code></td>
<td><code>';'</code></td>
<td>Delimiter for array values</td>
</tr>
<tr>
<td><code>skip_duplicate_nodes</code></td>
<td><code>true</code></td>
<td>Whether to skip duplicate nodes during import</td>
</tr>
<tr>
<td><code>skip_bad_relationships</code></td>
<td><code>true</code></td>
<td>Whether to skip relationships with missing endpoints</td>
</tr>
<tr>
<td><code>import_call_bin_prefix</code></td>
<td>i.e., <code>/usr/bin/</code></td>
<td>Prefix for the import command binary (optional)</td>
</tr>
</tbody>
</table>
<p>The <code>import_call_bin_prefix</code> is the path to your Neo4j instance that you looked up in <a href="#setup-neo4j">section Setup Neo4j</a> together with the prefix <code>/bin</code>.</p>
<p>The default configuration that comes with BioCypher and more configuration parameters for the Settings are listed in <a href="https://biocypher.org/BioCypher/reference/biocypher-config/">BioCypher Configuration Reference</a>.</p>
<blockquote>
<p>📝 <strong>Exercise:</strong>
Revise and complete the <code>biocypher_config.yaml</code> file, and make sure it is located in the <code>config</code> folder.</p>
</blockquote>
<details class="success">
<summary>Answer:</summary>
<p>See the example below for a completed <code>biocypher_config.yaml</code>. Note, the path in the  <code>import_call_bin_prefix</code> correspond to my personal instance, <strong>you MUST update this path with yours, do not forget to add /bin/</strong> as in my example</p>
<p><strong>File: <code>biocypher_config.yaml</code></strong></p>
<div class="language-yaml highlight"><pre><span></span><code><span id="__span-13-1"><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="c1">#---------------------------------------------------------------</span>
</span><span id="__span-13-2"><a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a><span class="c1">#--------        BIOCYPHER GENERAL CONFIGURATION        --------</span>
</span><span id="__span-13-3"><a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a><span class="c1">#---------------------------------------------------------------</span>
</span><span id="__span-13-4"><a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a><span class="nt">biocypher</span><span class="p">:</span>
</span><span id="__span-13-5"><a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a><span class="w">    </span><span class="nt">offline</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-13-6"><a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a><span class="w">    </span><span class="nt">debug</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
</span><span id="__span-13-7"><a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a><span class="w">    </span><span class="nt">schema_config_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/schema_config.yaml</span>
</span><span id="__span-13-8"><a id="__codelineno-13-8" name="__codelineno-13-8" href="#__codelineno-13-8"></a><span class="w">    </span><span class="nt">cache_directory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">.cache</span>
</span><span id="__span-13-9"><a id="__codelineno-13-9" name="__codelineno-13-9" href="#__codelineno-13-9"></a>
</span><span id="__span-13-10"><a id="__codelineno-13-10" name="__codelineno-13-10" href="#__codelineno-13-10"></a><span class="c1">#----------------------------------------------------</span>
</span><span id="__span-13-11"><a id="__codelineno-13-11" name="__codelineno-13-11" href="#__codelineno-13-11"></a><span class="c1">#--------        OUTPUT CONFIGURATION        --------</span>
</span><span id="__span-13-12"><a id="__codelineno-13-12" name="__codelineno-13-12" href="#__codelineno-13-12"></a><span class="c1">#----------------------------------------------------</span>
</span><span id="__span-13-13"><a id="__codelineno-13-13" name="__codelineno-13-13" href="#__codelineno-13-13"></a><span class="nt">neo4j</span><span class="p">:</span>
</span><span id="__span-13-14"><a id="__codelineno-13-14" name="__codelineno-13-14" href="#__codelineno-13-14"></a><span class="w">    </span><span class="nt">database_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">neo4j</span>
</span><span id="__span-13-15"><a id="__codelineno-13-15" name="__codelineno-13-15" href="#__codelineno-13-15"></a><span class="w">    </span><span class="nt">delimiter</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;\t&#39;</span>
</span><span id="__span-13-16"><a id="__codelineno-13-16" name="__codelineno-13-16" href="#__codelineno-13-16"></a><span class="w">    </span><span class="nt">array_delimiter</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;|&#39;</span>
</span><span id="__span-13-17"><a id="__codelineno-13-17" name="__codelineno-13-17" href="#__codelineno-13-17"></a><span class="w">    </span><span class="nt">skip_duplicate_nodes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-13-18"><a id="__codelineno-13-18" name="__codelineno-13-18" href="#__codelineno-13-18"></a><span class="w">    </span><span class="nt">skip_bad_relationships</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</span><span id="__span-13-19"><a id="__codelineno-13-19" name="__codelineno-13-19" href="#__codelineno-13-19"></a><span class="w">    </span><span class="nt">import_call_bin_prefix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/home/egcarren/.config/neo4j-desktop/Application/Data/dbmss/dbms-08155706-b96e-4e74-a965-7d6d27b78db8/bin/</span>
</span></code></pre></div>
</details>
<h3 id="step-2-create-an-adapter">Step 2. Create an adapter</h3>
<figure>
<img alt="Image title" src="../assets/biocypher_section_adapter.png" width="1000" />

<figcaption>Figure 11. Adapter creation in the BioCypher pipeline.</figcaption>
</figure>
<p><strong>Rationale:</strong> An adapter allows you to efficiently transform, integrate, combine data from different sources ensuring compatibility with BioCypher's schema and streamlining the import process.</p>
<p>a. Create a file called <code>adapter_synthetic_proteins.py</code> under the folder <code>/template_package/adapters/</code>, in this file we are going to create our adapter.</p>
<p>b. Define Enums for the different types of nodes in our graph and their properties. An Enum is simply a list of allowed values for a category — for example, all possible node types in the graph. Using Enums helps ensure that only valid types are used, makes the code easier to read, and avoids typos. For now, just focus on analyzing the snippet and compare its values with the node elements expected in the schema file(<code>schema_config.yaml</code>).</p>
<details class="example">
<summary><strong>File: <code>/template_package/adapters/adapter_synthetic_proteins.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-14-1"><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="k">class</span><span class="w"> </span><span class="nc">AdapterNodeType</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="__span-14-2"><a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-14-3"><a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a><span class="sd">    Define types of nodes the adapter can provide.</span>
</span><span id="__span-14-4"><a id="__codelineno-14-4" name="__codelineno-14-4" href="#__codelineno-14-4"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-14-5"><a id="__codelineno-14-5" name="__codelineno-14-5" href="#__codelineno-14-5"></a>
</span><span id="__span-14-6"><a id="__codelineno-14-6" name="__codelineno-14-6" href="#__codelineno-14-6"></a>    <span class="n">PROTEIN</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span><span id="__span-14-7"><a id="__codelineno-14-7" name="__codelineno-14-7" href="#__codelineno-14-7"></a>
</span><span id="__span-14-8"><a id="__codelineno-14-8" name="__codelineno-14-8" href="#__codelineno-14-8"></a><span class="k">class</span><span class="w"> </span><span class="nc">AdapterProteinField</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="__span-14-9"><a id="__codelineno-14-9" name="__codelineno-14-9" href="#__codelineno-14-9"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-14-10"><a id="__codelineno-14-10" name="__codelineno-14-10" href="#__codelineno-14-10"></a><span class="sd">    Define possible fields the adapter can provide for proteins.</span>
</span><span id="__span-14-11"><a id="__codelineno-14-11" name="__codelineno-14-11" href="#__codelineno-14-11"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-14-12"><a id="__codelineno-14-12" name="__codelineno-14-12" href="#__codelineno-14-12"></a>
</span><span id="__span-14-13"><a id="__codelineno-14-13" name="__codelineno-14-13" href="#__codelineno-14-13"></a>    <span class="n">ID</span> <span class="o">=</span> <span class="s2">&quot;id&quot;</span>
</span><span id="__span-14-14"><a id="__codelineno-14-14" name="__codelineno-14-14" href="#__codelineno-14-14"></a>    <span class="n">PREFERRED_ID</span> <span class="o">=</span> <span class="s2">&quot;preferred_id&quot;</span>
</span><span id="__span-14-15"><a id="__codelineno-14-15" name="__codelineno-14-15" href="#__codelineno-14-15"></a>    <span class="n">GENE_SYMBOL</span> <span class="o">=</span> <span class="s2">&quot;genesymbol&quot;</span>
</span><span id="__span-14-16"><a id="__codelineno-14-16" name="__codelineno-14-16" href="#__codelineno-14-16"></a>    <span class="n">NCBI_TAX_ID</span> <span class="o">=</span> <span class="s2">&quot;ncbi_tax_id&quot;</span>
</span></code></pre></div>
</details>
<p>c. Similarly, define Enums for the different types of edges in our graph and their properties. For now, analyze the following snippet and compare the listed types with the edge elements expected in our schema file (<code>schema_config.yaml</code>).</p>
<details class="example">
<summary><strong>File: <code>/template_package/adapters/adapter_synthetic_proteins.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-15-1"><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a><span class="k">class</span><span class="w"> </span><span class="nc">AdapterEdgeType</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="__span-15-2"><a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-15-3"><a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a><span class="sd">    Enum for the types of the protein adapter.</span>
</span><span id="__span-15-4"><a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-15-5"><a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a>
</span><span id="__span-15-6"><a id="__codelineno-15-6" name="__codelineno-15-6" href="#__codelineno-15-6"></a>    <span class="n">PROTEIN_PROTEIN_INTERACTION</span> <span class="o">=</span> <span class="s2">&quot;protein_protein_interaction&quot;</span>
</span><span id="__span-15-7"><a id="__codelineno-15-7" name="__codelineno-15-7" href="#__codelineno-15-7"></a>    <span class="n">BINDING</span> <span class="o">=</span> <span class="s2">&quot;binding&quot;</span>
</span><span id="__span-15-8"><a id="__codelineno-15-8" name="__codelineno-15-8" href="#__codelineno-15-8"></a>    <span class="n">ACTIVATION</span> <span class="o">=</span> <span class="s2">&quot;activation&quot;</span>
</span><span id="__span-15-9"><a id="__codelineno-15-9" name="__codelineno-15-9" href="#__codelineno-15-9"></a>    <span class="n">PHOSPHORYLATION</span> <span class="o">=</span> <span class="s2">&quot;phosphorylation&quot;</span>
</span><span id="__span-15-10"><a id="__codelineno-15-10" name="__codelineno-15-10" href="#__codelineno-15-10"></a>    <span class="n">UBIQUITINATION</span> <span class="o">=</span> <span class="s2">&quot;ubiquitination&quot;</span>
</span><span id="__span-15-11"><a id="__codelineno-15-11" name="__codelineno-15-11" href="#__codelineno-15-11"></a>    <span class="n">INHIBITION</span> <span class="o">=</span> <span class="s2">&quot;inhibition&quot;</span>
</span><span id="__span-15-12"><a id="__codelineno-15-12" name="__codelineno-15-12" href="#__codelineno-15-12"></a>
</span><span id="__span-15-13"><a id="__codelineno-15-13" name="__codelineno-15-13" href="#__codelineno-15-13"></a><span class="k">class</span><span class="w"> </span><span class="nc">AdapterProteinProteinEdgeField</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="__span-15-14"><a id="__codelineno-15-14" name="__codelineno-15-14" href="#__codelineno-15-14"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-15-15"><a id="__codelineno-15-15" name="__codelineno-15-15" href="#__codelineno-15-15"></a><span class="sd">    Define possible fields the adapter can provide for protein-protein edges.</span>
</span><span id="__span-15-16"><a id="__codelineno-15-16" name="__codelineno-15-16" href="#__codelineno-15-16"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-15-17"><a id="__codelineno-15-17" name="__codelineno-15-17" href="#__codelineno-15-17"></a>
</span><span id="__span-15-18"><a id="__codelineno-15-18" name="__codelineno-15-18" href="#__codelineno-15-18"></a>    <span class="n">INTERACTION_TYPE</span> <span class="o">=</span> <span class="s2">&quot;interaction_type&quot;</span>
</span><span id="__span-15-19"><a id="__codelineno-15-19" name="__codelineno-15-19" href="#__codelineno-15-19"></a>    <span class="n">INTERACTION_SOURCE</span> <span class="o">=</span> <span class="s2">&quot;interaction_source&quot;</span>
</span><span id="__span-15-20"><a id="__codelineno-15-20" name="__codelineno-15-20" href="#__codelineno-15-20"></a>    <span class="n">IS_DIRECTED</span> <span class="o">=</span> <span class="s2">&quot;is_directed&quot;</span>
</span><span id="__span-15-21"><a id="__codelineno-15-21" name="__codelineno-15-21" href="#__codelineno-15-21"></a>    <span class="n">IS_STIMULATION</span> <span class="o">=</span> <span class="s2">&quot;is_stimulation&quot;</span>
</span><span id="__span-15-22"><a id="__codelineno-15-22" name="__codelineno-15-22" href="#__codelineno-15-22"></a>    <span class="n">IS_INHIBITION</span> <span class="o">=</span> <span class="s2">&quot;is_inhibition&quot;</span>
</span><span id="__span-15-23"><a id="__codelineno-15-23" name="__codelineno-15-23" href="#__codelineno-15-23"></a>    <span class="n">CONSENSUS_DIRECTION</span> <span class="o">=</span> <span class="s2">&quot;consensus_direction&quot;</span>
</span><span id="__span-15-24"><a id="__codelineno-15-24" name="__codelineno-15-24" href="#__codelineno-15-24"></a>    <span class="n">CONSENSUS_STIMULATION</span> <span class="o">=</span> <span class="s2">&quot;consensus_stimulation&quot;</span>
</span><span id="__span-15-25"><a id="__codelineno-15-25" name="__codelineno-15-25" href="#__codelineno-15-25"></a>    <span class="n">CONSENSUS_INHIBITION</span> <span class="o">=</span> <span class="s2">&quot;consensus_inhibition&quot;</span>
</span></code></pre></div>
</details>
<p>Now you are ready to write functions that load the data into a dataframe and extract nodes and edges. Before doing this you need to understand that BioCypher expect a certain format. Let's discuss this:</p>
<p>BioCypher expects each node or edge being a tuple (datastructure) as explained in the next section.</p>
<h4 id="nodes-convention">Nodes Convention</h4>
<p>BioCypher expects each node being a <strong>3-element tuple</strong>, with elements in the following order:
1. [<em>mandatory</em>]  <code>ID</code>: this is a unique identifier for the node (i.e., <code>P53</code>)
2. [<em>mandatory</em>]  <code>LABEL</code>: this is a namespace for the node (i.e, <code>uniprot</code>)
3. [<em>mandatory</em>]   <code>PROPERTIES</code>: a dictionary containing properties for the node, which may be empty (i.e., have no entries).</p>
<p>For example:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-16-1"><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a><span class="c1">#  ✅ This is a 3-element tuple</span>
</span><span id="__span-16-2"><a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a><span class="c1"># 1     2          3</span>
</span><span id="__span-16-3"><a id="__codelineno-16-3" name="__codelineno-16-3" href="#__codelineno-16-3"></a><span class="p">(</span><span class="s2">&quot;TP53&quot;</span><span class="p">,</span> <span class="s2">&quot;uniprot&quot;</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;entity_type&quot;</span><span class="p">:</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="s2">&quot;ncbi_tax_id&quot;</span><span class="p">:</span><span class="mi">9096</span><span class="p">})</span>
</span></code></pre></div>
<div class="language-python highlight"><pre><span></span><code><span id="__span-17-1"><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a><span class="c1"># ✖️ This is NOT a 3-element tuple</span>
</span><span id="__span-17-2"><a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a><span class="c1">#1      2          3              4          5              6</span>
</span><span id="__span-17-3"><a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a><span class="p">(</span><span class="s2">&quot;TP53&quot;</span><span class="p">,</span> <span class="s2">&quot;uniprot&quot;</span><span class="p">,</span> <span class="s2">&quot;entity_type&quot;</span><span class="p">,</span> <span class="s2">&quot;protein&quot;</span><span class="p">,</span> <span class="s2">&quot;ncbi_tax_id&quot;</span><span class="p">,</span> <span class="mi">9096</span><span class="p">)</span>
</span></code></pre></div>
<h4 id="edges-convention">Edges Convention</h4>
<p>BioCypher expects each edge being a <strong>5-element tuple</strong>, with elements in the following order:</p>
<ol>
<li>[<em>optional</em>]  <code>ID</code>: this is a unique identifier for the edge (i.e., <code>P53CREB1</code>)</li>
<li>[<em>mandatory</em>]  <code>SOURCE</code>: this is the ID for the source node (i.e, <code>TP53</code>)</li>
<li>[<em>mandatory</em>]  <code>TARGET</code>: this is the ID for the source node (i.e, <code>CREB1</code>)</li>
<li>[<em>mandatory</em>]  <code>LABEL</code>: this is a namespace for the edge (i.e, <code>activation</code>, <code>inhibition</code>)</li>
<li>[<em>mandatory</em>] <code>PROPERTIES</code>: a dictionary containing properties for the edge, which may be empty (i.e., have no entries).</li>
</ol>
<div class="language-python highlight"><pre><span></span><code><span id="__span-18-1"><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a><span class="c1">#  ✅ This is a 5-element tuple for BioCypher</span>
</span><span id="__span-18-2"><a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a><span class="c1">#1   2      3        4                 5</span>
</span><span id="__span-18-3"><a id="__codelineno-18-3" name="__codelineno-18-3" href="#__codelineno-18-3"></a><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;TP53&quot;</span><span class="p">,</span> <span class="s2">&quot;CREB1&quot;</span><span class="p">,</span> <span class="s2">&quot;ubiquitination&quot;</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;is_directed&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">})</span>
</span></code></pre></div>
<div class="language-python highlight"><pre><span></span><code><span id="__span-19-1"><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a><span class="c1"># ✖️ This is NOT the 5-element tuple format that BioCypher expects. Can you explain why?</span>
</span><span id="__span-19-2"><a id="__codelineno-19-2" name="__codelineno-19-2" href="#__codelineno-19-2"></a><span class="c1">#1           2      3                 4              5</span>
</span><span id="__span-19-3"><a id="__codelineno-19-3" name="__codelineno-19-3" href="#__codelineno-19-3"></a><span class="p">(</span><span class="s2">&quot;TP53CREB1&quot;</span><span class="p">,</span> <span class="s2">&quot;TP53&quot;</span><span class="p">,</span> <span class="s2">&quot;ubiquitination&quot;</span><span class="p">,</span> <span class="s2">&quot;is_directed&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
</span></code></pre></div>
<p>Finally, write the functions that read the data as a DataFrame and override the functions to extract nodes and edges in the formats expected by BioCypher. This is illustrated in the next snippet.</p>
<details class="example">
<summary><strong>File: <code>/template_package/adapters/adapter_synthetic_proteins.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-20-1"><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a><span class="k">def</span><span class="w"> </span><span class="nf">_read_tsv</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
</span><span id="__span-20-2"><a id="__codelineno-20-2" name="__codelineno-20-2" href="#__codelineno-20-2"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-20-3"><a id="__codelineno-20-3" name="__codelineno-20-3" href="#__codelineno-20-3"></a><span class="sd">    Reads and validates the TSV file.</span>
</span><span id="__span-20-4"><a id="__codelineno-20-4" name="__codelineno-20-4" href="#__codelineno-20-4"></a><span class="sd">    Returns:</span>
</span><span id="__span-20-5"><a id="__codelineno-20-5" name="__codelineno-20-5" href="#__codelineno-20-5"></a><span class="sd">        pd.DataFrame: DataFrame containing the TSV data.</span>
</span><span id="__span-20-6"><a id="__codelineno-20-6" name="__codelineno-20-6" href="#__codelineno-20-6"></a><span class="sd">    Raises:</span>
</span><span id="__span-20-7"><a id="__codelineno-20-7" name="__codelineno-20-7" href="#__codelineno-20-7"></a><span class="sd">        FileNotFoundError: If the file does not exist.</span>
</span><span id="__span-20-8"><a id="__codelineno-20-8" name="__codelineno-20-8" href="#__codelineno-20-8"></a><span class="sd">        ValueError: If required columns are missing.</span>
</span><span id="__span-20-9"><a id="__codelineno-20-9" name="__codelineno-20-9" href="#__codelineno-20-9"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-20-10"><a id="__codelineno-20-10" name="__codelineno-20-10" href="#__codelineno-20-10"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv_path</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
</span><span id="__span-20-11"><a id="__codelineno-20-11" name="__codelineno-20-11" href="#__codelineno-20-11"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;TSV file not found: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-20-12"><a id="__codelineno-20-12" name="__codelineno-20-12" href="#__codelineno-20-12"></a>        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;TSV file not found: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-20-13"><a id="__codelineno-20-13" name="__codelineno-20-13" href="#__codelineno-20-13"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</span><span id="__span-20-14"><a id="__codelineno-20-14" name="__codelineno-20-14" href="#__codelineno-20-14"></a>    <span class="n">required_columns</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-20-15"><a id="__codelineno-20-15" name="__codelineno-20-15" href="#__codelineno-20-15"></a>        <span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="s1">&#39;source_genesymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;target_genesymbol&#39;</span><span class="p">,</span>
</span><span id="__span-20-16"><a id="__codelineno-20-16" name="__codelineno-20-16" href="#__codelineno-20-16"></a>        <span class="s1">&#39;ncbi_tax_id_source&#39;</span><span class="p">,</span> <span class="s1">&#39;ncbi_tax_id_target&#39;</span><span class="p">,</span> <span class="s1">&#39;type&#39;</span><span class="p">,</span>
</span><span id="__span-20-17"><a id="__codelineno-20-17" name="__codelineno-20-17" href="#__codelineno-20-17"></a>        <span class="s1">&#39;is_directed&#39;</span><span class="p">,</span> <span class="s1">&#39;is_stimulation&#39;</span><span class="p">,</span> <span class="s1">&#39;is_inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;consensus_direction&#39;</span><span class="p">,</span>
</span><span id="__span-20-18"><a id="__codelineno-20-18" name="__codelineno-20-18" href="#__codelineno-20-18"></a>        <span class="s1">&#39;consensus_stimulation&#39;</span><span class="p">,</span> <span class="s1">&#39;consensus_inhibition&#39;</span>
</span><span id="__span-20-19"><a id="__codelineno-20-19" name="__codelineno-20-19" href="#__codelineno-20-19"></a>    <span class="p">]</span>
</span><span id="__span-20-20"><a id="__codelineno-20-20" name="__codelineno-20-20" href="#__codelineno-20-20"></a>    <span class="n">missing</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">required_columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
</span><span id="__span-20-21"><a id="__codelineno-20-21" name="__codelineno-20-21" href="#__codelineno-20-21"></a>    <span class="k">if</span> <span class="n">missing</span><span class="p">:</span>
</span><span id="__span-20-22"><a id="__codelineno-20-22" name="__codelineno-20-22" href="#__codelineno-20-22"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing columns in TSV: </span><span class="si">{</span><span class="n">missing</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-20-23"><a id="__codelineno-20-23" name="__codelineno-20-23" href="#__codelineno-20-23"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;TSV must contain columns: </span><span class="si">{</span><span class="n">missing</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-20-24"><a id="__codelineno-20-24" name="__codelineno-20-24" href="#__codelineno-20-24"></a>    <span class="k">return</span> <span class="n">df</span>
</span><span id="__span-20-25"><a id="__codelineno-20-25" name="__codelineno-20-25" href="#__codelineno-20-25"></a>
</span><span id="__span-20-26"><a id="__codelineno-20-26" name="__codelineno-20-26" href="#__codelineno-20-26"></a><span class="k">def</span><span class="w"> </span><span class="nf">get_nodes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;Generator[tuple[str, str, dict], None, None]&#39;</span><span class="p">:</span>
</span><span id="__span-20-27"><a id="__codelineno-20-27" name="__codelineno-20-27" href="#__codelineno-20-27"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-20-28"><a id="__codelineno-20-28" name="__codelineno-20-28" href="#__codelineno-20-28"></a><span class="sd">    Yields node tuples for node types specified in the adapter constructor.</span>
</span><span id="__span-20-29"><a id="__codelineno-20-29" name="__codelineno-20-29" href="#__codelineno-20-29"></a>
</span><span id="__span-20-30"><a id="__codelineno-20-30" name="__codelineno-20-30" href="#__codelineno-20-30"></a><span class="sd">    Returns:</span>
</span><span id="__span-20-31"><a id="__codelineno-20-31" name="__codelineno-20-31" href="#__codelineno-20-31"></a><span class="sd">        Generator[tuple[str, str, dict], None, None]:</span>
</span><span id="__span-20-32"><a id="__codelineno-20-32" name="__codelineno-20-32" href="#__codelineno-20-32"></a><span class="sd">            Each tuple is (id, label, properties).</span>
</span><span id="__span-20-33"><a id="__codelineno-20-33" name="__codelineno-20-33" href="#__codelineno-20-33"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-20-34"><a id="__codelineno-20-34" name="__codelineno-20-34" href="#__codelineno-20-34"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading nodes.&quot;</span><span class="p">)</span>
</span><span id="__span-20-35"><a id="__codelineno-20-35" name="__codelineno-20-35" href="#__codelineno-20-35"></a>    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_tsv</span><span class="p">()</span>
</span><span id="__span-20-36"><a id="__codelineno-20-36" name="__codelineno-20-36" href="#__codelineno-20-36"></a>
</span><span id="__span-20-37"><a id="__codelineno-20-37" name="__codelineno-20-37" href="#__codelineno-20-37"></a>    <span class="c1"># Generator for nodes in the `source` column</span>
</span><span id="__span-20-38"><a id="__codelineno-20-38" name="__codelineno-20-38" href="#__codelineno-20-38"></a>    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-20-39"><a id="__codelineno-20-39" name="__codelineno-20-39" href="#__codelineno-20-39"></a>        <span class="nb">id</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">source</span>
</span><span id="__span-20-40"><a id="__codelineno-20-40" name="__codelineno-20-40" href="#__codelineno-20-40"></a>        <span class="n">input_label</span> <span class="o">=</span> <span class="s2">&quot;uniprot_protein&quot;</span>
</span><span id="__span-20-41"><a id="__codelineno-20-41" name="__codelineno-20-41" href="#__codelineno-20-41"></a>
</span><span id="__span-20-42"><a id="__codelineno-20-42" name="__codelineno-20-42" href="#__codelineno-20-42"></a>        <span class="n">properties</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-20-43"><a id="__codelineno-20-43" name="__codelineno-20-43" href="#__codelineno-20-43"></a>            <span class="s1">&#39;genesymbol&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">source_genesymbol</span><span class="p">,</span>
</span><span id="__span-20-44"><a id="__codelineno-20-44" name="__codelineno-20-44" href="#__codelineno-20-44"></a>            <span class="s1">&#39;ncbi_tax_id&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">ncbi_tax_id_source</span><span class="p">,</span>
</span><span id="__span-20-45"><a id="__codelineno-20-45" name="__codelineno-20-45" href="#__codelineno-20-45"></a>            <span class="s1">&#39;entity_type&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">entity_type_source</span><span class="p">,</span>
</span><span id="__span-20-46"><a id="__codelineno-20-46" name="__codelineno-20-46" href="#__codelineno-20-46"></a>        <span class="p">}</span>
</span><span id="__span-20-47"><a id="__codelineno-20-47" name="__codelineno-20-47" href="#__codelineno-20-47"></a>
</span><span id="__span-20-48"><a id="__codelineno-20-48" name="__codelineno-20-48" href="#__codelineno-20-48"></a>        <span class="k">yield</span><span class="p">(</span>
</span><span id="__span-20-49"><a id="__codelineno-20-49" name="__codelineno-20-49" href="#__codelineno-20-49"></a>            <span class="nb">id</span><span class="p">,</span>
</span><span id="__span-20-50"><a id="__codelineno-20-50" name="__codelineno-20-50" href="#__codelineno-20-50"></a>            <span class="n">input_label</span><span class="p">,</span>
</span><span id="__span-20-51"><a id="__codelineno-20-51" name="__codelineno-20-51" href="#__codelineno-20-51"></a>            <span class="n">properties</span>
</span><span id="__span-20-52"><a id="__codelineno-20-52" name="__codelineno-20-52" href="#__codelineno-20-52"></a>        <span class="p">)</span>
</span><span id="__span-20-53"><a id="__codelineno-20-53" name="__codelineno-20-53" href="#__codelineno-20-53"></a>
</span><span id="__span-20-54"><a id="__codelineno-20-54" name="__codelineno-20-54" href="#__codelineno-20-54"></a>    <span class="c1"># Generator for nodes in the `target` column</span>
</span><span id="__span-20-55"><a id="__codelineno-20-55" name="__codelineno-20-55" href="#__codelineno-20-55"></a>    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-20-56"><a id="__codelineno-20-56" name="__codelineno-20-56" href="#__codelineno-20-56"></a>        <span class="nb">id</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">target</span>
</span><span id="__span-20-57"><a id="__codelineno-20-57" name="__codelineno-20-57" href="#__codelineno-20-57"></a>        <span class="n">input_label</span> <span class="o">=</span> <span class="s2">&quot;uniprot_protein&quot;</span>
</span><span id="__span-20-58"><a id="__codelineno-20-58" name="__codelineno-20-58" href="#__codelineno-20-58"></a>
</span><span id="__span-20-59"><a id="__codelineno-20-59" name="__codelineno-20-59" href="#__codelineno-20-59"></a>        <span class="n">properties</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-20-60"><a id="__codelineno-20-60" name="__codelineno-20-60" href="#__codelineno-20-60"></a>            <span class="s1">&#39;genesymbol&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">target_genesymbol</span><span class="p">,</span>
</span><span id="__span-20-61"><a id="__codelineno-20-61" name="__codelineno-20-61" href="#__codelineno-20-61"></a>            <span class="s1">&#39;ncbi_tax_id&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">ncbi_tax_id_target</span><span class="p">,</span>
</span><span id="__span-20-62"><a id="__codelineno-20-62" name="__codelineno-20-62" href="#__codelineno-20-62"></a>            <span class="s1">&#39;entity_type&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">entity_type_target</span><span class="p">,</span>
</span><span id="__span-20-63"><a id="__codelineno-20-63" name="__codelineno-20-63" href="#__codelineno-20-63"></a>        <span class="p">}</span>
</span><span id="__span-20-64"><a id="__codelineno-20-64" name="__codelineno-20-64" href="#__codelineno-20-64"></a>
</span><span id="__span-20-65"><a id="__codelineno-20-65" name="__codelineno-20-65" href="#__codelineno-20-65"></a>        <span class="k">yield</span><span class="p">(</span>
</span><span id="__span-20-66"><a id="__codelineno-20-66" name="__codelineno-20-66" href="#__codelineno-20-66"></a>            <span class="nb">id</span><span class="p">,</span>
</span><span id="__span-20-67"><a id="__codelineno-20-67" name="__codelineno-20-67" href="#__codelineno-20-67"></a>            <span class="n">input_label</span><span class="p">,</span>
</span><span id="__span-20-68"><a id="__codelineno-20-68" name="__codelineno-20-68" href="#__codelineno-20-68"></a>            <span class="n">properties</span>
</span><span id="__span-20-69"><a id="__codelineno-20-69" name="__codelineno-20-69" href="#__codelineno-20-69"></a>        <span class="p">)</span>
</span><span id="__span-20-70"><a id="__codelineno-20-70" name="__codelineno-20-70" href="#__codelineno-20-70"></a>
</span><span id="__span-20-71"><a id="__codelineno-20-71" name="__codelineno-20-71" href="#__codelineno-20-71"></a><span class="k">def</span><span class="w"> </span><span class="nf">get_edges</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;Generator[tuple[str, str, str, str, dict], None, None]&#39;</span><span class="p">:</span>
</span><span id="__span-20-72"><a id="__codelineno-20-72" name="__codelineno-20-72" href="#__codelineno-20-72"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-20-73"><a id="__codelineno-20-73" name="__codelineno-20-73" href="#__codelineno-20-73"></a><span class="sd">    Yields edge tuples for edge types specified in the adapter constructor.</span>
</span><span id="__span-20-74"><a id="__codelineno-20-74" name="__codelineno-20-74" href="#__codelineno-20-74"></a>
</span><span id="__span-20-75"><a id="__codelineno-20-75" name="__codelineno-20-75" href="#__codelineno-20-75"></a><span class="sd">    Returns:</span>
</span><span id="__span-20-76"><a id="__codelineno-20-76" name="__codelineno-20-76" href="#__codelineno-20-76"></a><span class="sd">        Generator[tuple[str, str, str, str, dict], None, None]:</span>
</span><span id="__span-20-77"><a id="__codelineno-20-77" name="__codelineno-20-77" href="#__codelineno-20-77"></a><span class="sd">            Each tuple is (id, source, target, type, properties).</span>
</span><span id="__span-20-78"><a id="__codelineno-20-78" name="__codelineno-20-78" href="#__codelineno-20-78"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-20-79"><a id="__codelineno-20-79" name="__codelineno-20-79" href="#__codelineno-20-79"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Generating edges.&quot;</span><span class="p">)</span>
</span><span id="__span-20-80"><a id="__codelineno-20-80" name="__codelineno-20-80" href="#__codelineno-20-80"></a>    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_tsv</span><span class="p">()</span>
</span><span id="__span-20-81"><a id="__codelineno-20-81" name="__codelineno-20-81" href="#__codelineno-20-81"></a>
</span><span id="__span-20-82"><a id="__codelineno-20-82" name="__codelineno-20-82" href="#__codelineno-20-82"></a>    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-20-83"><a id="__codelineno-20-83" name="__codelineno-20-83" href="#__codelineno-20-83"></a>        <span class="c1"># Concatenate source and target, i.e., &quot;SOD1EGFR&quot;</span>
</span><span id="__span-20-84"><a id="__codelineno-20-84" name="__codelineno-20-84" href="#__codelineno-20-84"></a>        <span class="nb">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">source</span><span class="si">}{</span><span class="n">row</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s2">&quot;</span>
</span><span id="__span-20-85"><a id="__codelineno-20-85" name="__codelineno-20-85" href="#__codelineno-20-85"></a>
</span><span id="__span-20-86"><a id="__codelineno-20-86" name="__codelineno-20-86" href="#__codelineno-20-86"></a>        <span class="n">source</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">source</span>
</span><span id="__span-20-87"><a id="__codelineno-20-87" name="__codelineno-20-87" href="#__codelineno-20-87"></a>
</span><span id="__span-20-88"><a id="__codelineno-20-88" name="__codelineno-20-88" href="#__codelineno-20-88"></a>        <span class="n">target</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">target</span>
</span><span id="__span-20-89"><a id="__codelineno-20-89" name="__codelineno-20-89" href="#__codelineno-20-89"></a>
</span><span id="__span-20-90"><a id="__codelineno-20-90" name="__codelineno-20-90" href="#__codelineno-20-90"></a>        <span class="nb">type</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">type</span>
</span><span id="__span-20-91"><a id="__codelineno-20-91" name="__codelineno-20-91" href="#__codelineno-20-91"></a>
</span><span id="__span-20-92"><a id="__codelineno-20-92" name="__codelineno-20-92" href="#__codelineno-20-92"></a>        <span class="n">properties</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-20-93"><a id="__codelineno-20-93" name="__codelineno-20-93" href="#__codelineno-20-93"></a>            <span class="s1">&#39;is_directed&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">is_directed</span><span class="p">,</span>
</span><span id="__span-20-94"><a id="__codelineno-20-94" name="__codelineno-20-94" href="#__codelineno-20-94"></a>            <span class="s1">&#39;is_stimulation&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">is_stimulation</span><span class="p">,</span>
</span><span id="__span-20-95"><a id="__codelineno-20-95" name="__codelineno-20-95" href="#__codelineno-20-95"></a>            <span class="s1">&#39;is_inhibition&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">is_inhibition</span><span class="p">,</span>
</span><span id="__span-20-96"><a id="__codelineno-20-96" name="__codelineno-20-96" href="#__codelineno-20-96"></a>            <span class="s1">&#39;consensus_direction&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">consensus_direction</span><span class="p">,</span>
</span><span id="__span-20-97"><a id="__codelineno-20-97" name="__codelineno-20-97" href="#__codelineno-20-97"></a>            <span class="s1">&#39;consensus_stimulation&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">consensus_stimulation</span><span class="p">,</span>
</span><span id="__span-20-98"><a id="__codelineno-20-98" name="__codelineno-20-98" href="#__codelineno-20-98"></a>            <span class="s1">&#39;consensus_inhibition&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">consensus_inhibition</span>
</span><span id="__span-20-99"><a id="__codelineno-20-99" name="__codelineno-20-99" href="#__codelineno-20-99"></a>        <span class="p">}</span>
</span><span id="__span-20-100"><a id="__codelineno-20-100" name="__codelineno-20-100" href="#__codelineno-20-100"></a>
</span><span id="__span-20-101"><a id="__codelineno-20-101" name="__codelineno-20-101" href="#__codelineno-20-101"></a>        <span class="k">yield</span> <span class="p">(</span>
</span><span id="__span-20-102"><a id="__codelineno-20-102" name="__codelineno-20-102" href="#__codelineno-20-102"></a>            <span class="nb">id</span><span class="p">,</span>
</span><span id="__span-20-103"><a id="__codelineno-20-103" name="__codelineno-20-103" href="#__codelineno-20-103"></a>            <span class="n">source</span><span class="p">,</span>
</span><span id="__span-20-104"><a id="__codelineno-20-104" name="__codelineno-20-104" href="#__codelineno-20-104"></a>            <span class="n">target</span><span class="p">,</span>
</span><span id="__span-20-105"><a id="__codelineno-20-105" name="__codelineno-20-105" href="#__codelineno-20-105"></a>            <span class="nb">type</span><span class="p">,</span>
</span><span id="__span-20-106"><a id="__codelineno-20-106" name="__codelineno-20-106" href="#__codelineno-20-106"></a>            <span class="n">properties</span>
</span><span id="__span-20-107"><a id="__codelineno-20-107" name="__codelineno-20-107" href="#__codelineno-20-107"></a>        <span class="p">)</span>
</span></code></pre></div>
</details>
<blockquote>
<p>📝 <strong>Exercise:</strong>
Integrate the aforementioned snippets in a single file call <code>adapter_synthetic_proteins.py</code>.</p>
</blockquote>
<details class="success">
<summary>Answer:</summary>
<p>See the example below for a completed <code>adapter_synthetic_proteins.py</code>.</p>
<p><strong>File: <code>/template_package/adapters/adapter_synthetic_proteins.py</code></strong></p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-21-1"><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
</span><span id="__span-21-2"><a id="__codelineno-21-2" name="__codelineno-21-2" href="#__codelineno-21-2"></a><span class="kn">from</span><span class="w"> </span><span class="nn">enum</span><span class="w"> </span><span class="kn">import</span> <span class="n">Enum</span><span class="p">,</span> <span class="n">auto</span>
</span><span id="__span-21-3"><a id="__codelineno-21-3" name="__codelineno-21-3" href="#__codelineno-21-3"></a><span class="kn">from</span><span class="w"> </span><span class="nn">itertools</span><span class="w"> </span><span class="kn">import</span> <span class="n">chain</span>
</span><span id="__span-21-4"><a id="__codelineno-21-4" name="__codelineno-21-4" href="#__codelineno-21-4"></a><span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Generator</span>
</span><span id="__span-21-5"><a id="__codelineno-21-5" name="__codelineno-21-5" href="#__codelineno-21-5"></a><span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
</span><span id="__span-21-6"><a id="__codelineno-21-6" name="__codelineno-21-6" href="#__codelineno-21-6"></a>
</span><span id="__span-21-7"><a id="__codelineno-21-7" name="__codelineno-21-7" href="#__codelineno-21-7"></a>
</span><span id="__span-21-8"><a id="__codelineno-21-8" name="__codelineno-21-8" href="#__codelineno-21-8"></a><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
</span><span id="__span-21-9"><a id="__codelineno-21-9" name="__codelineno-21-9" href="#__codelineno-21-9"></a><span class="kn">from</span><span class="w"> </span><span class="nn">biocypher</span><span class="w"> </span><span class="kn">import</span> <span class="n">BioCypher</span>
</span><span id="__span-21-10"><a id="__codelineno-21-10" name="__codelineno-21-10" href="#__codelineno-21-10"></a><span class="kn">from</span><span class="w"> </span><span class="nn">biocypher._logger</span><span class="w"> </span><span class="kn">import</span> <span class="n">logger</span>
</span><span id="__span-21-11"><a id="__codelineno-21-11" name="__codelineno-21-11" href="#__codelineno-21-11"></a>
</span><span id="__span-21-12"><a id="__codelineno-21-12" name="__codelineno-21-12" href="#__codelineno-21-12"></a>
</span><span id="__span-21-13"><a id="__codelineno-21-13" name="__codelineno-21-13" href="#__codelineno-21-13"></a><span class="n">TSV_FILE_PATH_SYNTHETIC_PROTEINS</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;./cache/synthetic_protein_interactions.tsv&quot;</span><span class="p">)</span>
</span><span id="__span-21-14"><a id="__codelineno-21-14" name="__codelineno-21-14" href="#__codelineno-21-14"></a>
</span><span id="__span-21-15"><a id="__codelineno-21-15" name="__codelineno-21-15" href="#__codelineno-21-15"></a><span class="k">class</span><span class="w"> </span><span class="nc">AdapterNodeType</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="__span-21-16"><a id="__codelineno-21-16" name="__codelineno-21-16" href="#__codelineno-21-16"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-21-17"><a id="__codelineno-21-17" name="__codelineno-21-17" href="#__codelineno-21-17"></a><span class="sd">    Define types of nodes the adapter can provide.</span>
</span><span id="__span-21-18"><a id="__codelineno-21-18" name="__codelineno-21-18" href="#__codelineno-21-18"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-21-19"><a id="__codelineno-21-19" name="__codelineno-21-19" href="#__codelineno-21-19"></a>
</span><span id="__span-21-20"><a id="__codelineno-21-20" name="__codelineno-21-20" href="#__codelineno-21-20"></a>    <span class="n">PROTEIN</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span><span id="__span-21-21"><a id="__codelineno-21-21" name="__codelineno-21-21" href="#__codelineno-21-21"></a>
</span><span id="__span-21-22"><a id="__codelineno-21-22" name="__codelineno-21-22" href="#__codelineno-21-22"></a><span class="k">class</span><span class="w"> </span><span class="nc">AdapterProteinField</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="__span-21-23"><a id="__codelineno-21-23" name="__codelineno-21-23" href="#__codelineno-21-23"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-21-24"><a id="__codelineno-21-24" name="__codelineno-21-24" href="#__codelineno-21-24"></a><span class="sd">    Define possible fields the adapter can provide for proteins.</span>
</span><span id="__span-21-25"><a id="__codelineno-21-25" name="__codelineno-21-25" href="#__codelineno-21-25"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-21-26"><a id="__codelineno-21-26" name="__codelineno-21-26" href="#__codelineno-21-26"></a>
</span><span id="__span-21-27"><a id="__codelineno-21-27" name="__codelineno-21-27" href="#__codelineno-21-27"></a>    <span class="n">ID</span> <span class="o">=</span> <span class="s2">&quot;id&quot;</span>
</span><span id="__span-21-28"><a id="__codelineno-21-28" name="__codelineno-21-28" href="#__codelineno-21-28"></a>    <span class="n">PREFERRED_ID</span> <span class="o">=</span> <span class="s2">&quot;preferred_id&quot;</span>
</span><span id="__span-21-29"><a id="__codelineno-21-29" name="__codelineno-21-29" href="#__codelineno-21-29"></a>    <span class="n">GENE_SYMBOL</span> <span class="o">=</span> <span class="s2">&quot;genesymbol&quot;</span>
</span><span id="__span-21-30"><a id="__codelineno-21-30" name="__codelineno-21-30" href="#__codelineno-21-30"></a>    <span class="n">NCBI_TAX_ID</span> <span class="o">=</span> <span class="s2">&quot;ncbi_tax_id&quot;</span>
</span><span id="__span-21-31"><a id="__codelineno-21-31" name="__codelineno-21-31" href="#__codelineno-21-31"></a>
</span><span id="__span-21-32"><a id="__codelineno-21-32" name="__codelineno-21-32" href="#__codelineno-21-32"></a><span class="k">class</span><span class="w"> </span><span class="nc">AdapterEdgeType</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="__span-21-33"><a id="__codelineno-21-33" name="__codelineno-21-33" href="#__codelineno-21-33"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-21-34"><a id="__codelineno-21-34" name="__codelineno-21-34" href="#__codelineno-21-34"></a><span class="sd">    Enum for the types of the protein adapter.</span>
</span><span id="__span-21-35"><a id="__codelineno-21-35" name="__codelineno-21-35" href="#__codelineno-21-35"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-21-36"><a id="__codelineno-21-36" name="__codelineno-21-36" href="#__codelineno-21-36"></a>
</span><span id="__span-21-37"><a id="__codelineno-21-37" name="__codelineno-21-37" href="#__codelineno-21-37"></a>    <span class="n">PROTEIN_PROTEIN_INTERACTION</span> <span class="o">=</span> <span class="s2">&quot;protein_protein_interaction&quot;</span>
</span><span id="__span-21-38"><a id="__codelineno-21-38" name="__codelineno-21-38" href="#__codelineno-21-38"></a>    <span class="n">BINDING</span> <span class="o">=</span> <span class="s2">&quot;binding&quot;</span>
</span><span id="__span-21-39"><a id="__codelineno-21-39" name="__codelineno-21-39" href="#__codelineno-21-39"></a>    <span class="n">ACTIVATION</span> <span class="o">=</span> <span class="s2">&quot;activation&quot;</span>
</span><span id="__span-21-40"><a id="__codelineno-21-40" name="__codelineno-21-40" href="#__codelineno-21-40"></a>    <span class="n">PHOSPHORYLATION</span> <span class="o">=</span> <span class="s2">&quot;phosphorylation&quot;</span>
</span><span id="__span-21-41"><a id="__codelineno-21-41" name="__codelineno-21-41" href="#__codelineno-21-41"></a>    <span class="n">UBIQUITINATION</span> <span class="o">=</span> <span class="s2">&quot;ubiquitination&quot;</span>
</span><span id="__span-21-42"><a id="__codelineno-21-42" name="__codelineno-21-42" href="#__codelineno-21-42"></a>    <span class="n">INHIBITION</span> <span class="o">=</span> <span class="s2">&quot;inhibition&quot;</span>
</span><span id="__span-21-43"><a id="__codelineno-21-43" name="__codelineno-21-43" href="#__codelineno-21-43"></a>
</span><span id="__span-21-44"><a id="__codelineno-21-44" name="__codelineno-21-44" href="#__codelineno-21-44"></a><span class="k">class</span><span class="w"> </span><span class="nc">AdapterProteinProteinEdgeField</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="__span-21-45"><a id="__codelineno-21-45" name="__codelineno-21-45" href="#__codelineno-21-45"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-21-46"><a id="__codelineno-21-46" name="__codelineno-21-46" href="#__codelineno-21-46"></a><span class="sd">    Define possible fields the adapter can provide for protein-protein edges.</span>
</span><span id="__span-21-47"><a id="__codelineno-21-47" name="__codelineno-21-47" href="#__codelineno-21-47"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-21-48"><a id="__codelineno-21-48" name="__codelineno-21-48" href="#__codelineno-21-48"></a>
</span><span id="__span-21-49"><a id="__codelineno-21-49" name="__codelineno-21-49" href="#__codelineno-21-49"></a>    <span class="n">INTERACTION_TYPE</span> <span class="o">=</span> <span class="s2">&quot;interaction_type&quot;</span>
</span><span id="__span-21-50"><a id="__codelineno-21-50" name="__codelineno-21-50" href="#__codelineno-21-50"></a>    <span class="n">INTERACTION_SOURCE</span> <span class="o">=</span> <span class="s2">&quot;interaction_source&quot;</span>
</span><span id="__span-21-51"><a id="__codelineno-21-51" name="__codelineno-21-51" href="#__codelineno-21-51"></a>    <span class="n">IS_DIRECTED</span> <span class="o">=</span> <span class="s2">&quot;is_directed&quot;</span>
</span><span id="__span-21-52"><a id="__codelineno-21-52" name="__codelineno-21-52" href="#__codelineno-21-52"></a>    <span class="n">IS_STIMULATION</span> <span class="o">=</span> <span class="s2">&quot;is_stimulation&quot;</span>
</span><span id="__span-21-53"><a id="__codelineno-21-53" name="__codelineno-21-53" href="#__codelineno-21-53"></a>    <span class="n">IS_INHIBITION</span> <span class="o">=</span> <span class="s2">&quot;is_inhibition&quot;</span>
</span><span id="__span-21-54"><a id="__codelineno-21-54" name="__codelineno-21-54" href="#__codelineno-21-54"></a>    <span class="n">CONSENSUS_DIRECTION</span> <span class="o">=</span> <span class="s2">&quot;consensus_direction&quot;</span>
</span><span id="__span-21-55"><a id="__codelineno-21-55" name="__codelineno-21-55" href="#__codelineno-21-55"></a>    <span class="n">CONSENSUS_STIMULATION</span> <span class="o">=</span> <span class="s2">&quot;consensus_stimulation&quot;</span>
</span><span id="__span-21-56"><a id="__codelineno-21-56" name="__codelineno-21-56" href="#__codelineno-21-56"></a>    <span class="n">CONSENSUS_INHIBITION</span> <span class="o">=</span> <span class="s2">&quot;consensus_inhibition&quot;</span>
</span><span id="__span-21-57"><a id="__codelineno-21-57" name="__codelineno-21-57" href="#__codelineno-21-57"></a>
</span><span id="__span-21-58"><a id="__codelineno-21-58" name="__codelineno-21-58" href="#__codelineno-21-58"></a><span class="k">class</span><span class="w"> </span><span class="nc">Adapter</span><span class="p">:</span>
</span><span id="__span-21-59"><a id="__codelineno-21-59" name="__codelineno-21-59" href="#__codelineno-21-59"></a>    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
</span><span id="__span-21-60"><a id="__codelineno-21-60" name="__codelineno-21-60" href="#__codelineno-21-60"></a>        <span class="bp">self</span><span class="p">,</span>
</span><span id="__span-21-61"><a id="__codelineno-21-61" name="__codelineno-21-61" href="#__codelineno-21-61"></a>        <span class="n">tsv_path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">TSV_FILE_PATH_SYNTHETIC_PROTEINS</span><span class="p">,</span>
</span><span id="__span-21-62"><a id="__codelineno-21-62" name="__codelineno-21-62" href="#__codelineno-21-62"></a>        <span class="n">node_types</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">list</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
</span><span id="__span-21-63"><a id="__codelineno-21-63" name="__codelineno-21-63" href="#__codelineno-21-63"></a>        <span class="n">node_fields</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">list</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
</span><span id="__span-21-64"><a id="__codelineno-21-64" name="__codelineno-21-64" href="#__codelineno-21-64"></a>        <span class="n">edge_types</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">list</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
</span><span id="__span-21-65"><a id="__codelineno-21-65" name="__codelineno-21-65" href="#__codelineno-21-65"></a>        <span class="n">edge_fields</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">list</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
</span><span id="__span-21-66"><a id="__codelineno-21-66" name="__codelineno-21-66" href="#__codelineno-21-66"></a>    <span class="p">):</span>
</span><span id="__span-21-67"><a id="__codelineno-21-67" name="__codelineno-21-67" href="#__codelineno-21-67"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">tsv_path</span> <span class="o">=</span> <span class="n">tsv_path</span>
</span><span id="__span-21-68"><a id="__codelineno-21-68" name="__codelineno-21-68" href="#__codelineno-21-68"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">_set_types_and_fields</span><span class="p">(</span><span class="n">node_types</span><span class="p">,</span> <span class="n">node_fields</span><span class="p">,</span> <span class="n">edge_types</span><span class="p">,</span> <span class="n">edge_fields</span><span class="p">)</span>
</span><span id="__span-21-69"><a id="__codelineno-21-69" name="__codelineno-21-69" href="#__codelineno-21-69"></a>
</span><span id="__span-21-70"><a id="__codelineno-21-70" name="__codelineno-21-70" href="#__codelineno-21-70"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">_read_tsv</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
</span><span id="__span-21-71"><a id="__codelineno-21-71" name="__codelineno-21-71" href="#__codelineno-21-71"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-21-72"><a id="__codelineno-21-72" name="__codelineno-21-72" href="#__codelineno-21-72"></a><span class="sd">        Reads and validates the TSV file.</span>
</span><span id="__span-21-73"><a id="__codelineno-21-73" name="__codelineno-21-73" href="#__codelineno-21-73"></a><span class="sd">        Returns:</span>
</span><span id="__span-21-74"><a id="__codelineno-21-74" name="__codelineno-21-74" href="#__codelineno-21-74"></a><span class="sd">            pd.DataFrame: DataFrame containing the TSV data.</span>
</span><span id="__span-21-75"><a id="__codelineno-21-75" name="__codelineno-21-75" href="#__codelineno-21-75"></a><span class="sd">        Raises:</span>
</span><span id="__span-21-76"><a id="__codelineno-21-76" name="__codelineno-21-76" href="#__codelineno-21-76"></a><span class="sd">            FileNotFoundError: If the file does not exist.</span>
</span><span id="__span-21-77"><a id="__codelineno-21-77" name="__codelineno-21-77" href="#__codelineno-21-77"></a><span class="sd">            ValueError: If required columns are missing.</span>
</span><span id="__span-21-78"><a id="__codelineno-21-78" name="__codelineno-21-78" href="#__codelineno-21-78"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-21-79"><a id="__codelineno-21-79" name="__codelineno-21-79" href="#__codelineno-21-79"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv_path</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
</span><span id="__span-21-80"><a id="__codelineno-21-80" name="__codelineno-21-80" href="#__codelineno-21-80"></a>            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;TSV file not found: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-21-81"><a id="__codelineno-21-81" name="__codelineno-21-81" href="#__codelineno-21-81"></a>            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;TSV file not found: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-21-82"><a id="__codelineno-21-82" name="__codelineno-21-82" href="#__codelineno-21-82"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</span><span id="__span-21-83"><a id="__codelineno-21-83" name="__codelineno-21-83" href="#__codelineno-21-83"></a>        <span class="n">required_columns</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-21-84"><a id="__codelineno-21-84" name="__codelineno-21-84" href="#__codelineno-21-84"></a>            <span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="s1">&#39;source_genesymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;target_genesymbol&#39;</span><span class="p">,</span>
</span><span id="__span-21-85"><a id="__codelineno-21-85" name="__codelineno-21-85" href="#__codelineno-21-85"></a>            <span class="s1">&#39;ncbi_tax_id_source&#39;</span><span class="p">,</span> <span class="s1">&#39;ncbi_tax_id_target&#39;</span><span class="p">,</span> <span class="s1">&#39;type&#39;</span><span class="p">,</span>
</span><span id="__span-21-86"><a id="__codelineno-21-86" name="__codelineno-21-86" href="#__codelineno-21-86"></a>            <span class="s1">&#39;is_directed&#39;</span><span class="p">,</span> <span class="s1">&#39;is_stimulation&#39;</span><span class="p">,</span> <span class="s1">&#39;is_inhibition&#39;</span><span class="p">,</span> <span class="s1">&#39;consensus_direction&#39;</span><span class="p">,</span>
</span><span id="__span-21-87"><a id="__codelineno-21-87" name="__codelineno-21-87" href="#__codelineno-21-87"></a>            <span class="s1">&#39;consensus_stimulation&#39;</span><span class="p">,</span> <span class="s1">&#39;consensus_inhibition&#39;</span>
</span><span id="__span-21-88"><a id="__codelineno-21-88" name="__codelineno-21-88" href="#__codelineno-21-88"></a>        <span class="p">]</span>
</span><span id="__span-21-89"><a id="__codelineno-21-89" name="__codelineno-21-89" href="#__codelineno-21-89"></a>        <span class="n">missing</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">required_columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
</span><span id="__span-21-90"><a id="__codelineno-21-90" name="__codelineno-21-90" href="#__codelineno-21-90"></a>        <span class="k">if</span> <span class="n">missing</span><span class="p">:</span>
</span><span id="__span-21-91"><a id="__codelineno-21-91" name="__codelineno-21-91" href="#__codelineno-21-91"></a>            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing columns in TSV: </span><span class="si">{</span><span class="n">missing</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-21-92"><a id="__codelineno-21-92" name="__codelineno-21-92" href="#__codelineno-21-92"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;TSV must contain columns: </span><span class="si">{</span><span class="n">missing</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-21-93"><a id="__codelineno-21-93" name="__codelineno-21-93" href="#__codelineno-21-93"></a>        <span class="k">return</span> <span class="n">df</span>
</span><span id="__span-21-94"><a id="__codelineno-21-94" name="__codelineno-21-94" href="#__codelineno-21-94"></a>
</span><span id="__span-21-95"><a id="__codelineno-21-95" name="__codelineno-21-95" href="#__codelineno-21-95"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">get_nodes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;Generator[tuple[str, str, dict], None, None]&#39;</span><span class="p">:</span>
</span><span id="__span-21-96"><a id="__codelineno-21-96" name="__codelineno-21-96" href="#__codelineno-21-96"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-21-97"><a id="__codelineno-21-97" name="__codelineno-21-97" href="#__codelineno-21-97"></a><span class="sd">        Yields node tuples for node types specified in the adapter constructor.</span>
</span><span id="__span-21-98"><a id="__codelineno-21-98" name="__codelineno-21-98" href="#__codelineno-21-98"></a>
</span><span id="__span-21-99"><a id="__codelineno-21-99" name="__codelineno-21-99" href="#__codelineno-21-99"></a><span class="sd">        Returns:</span>
</span><span id="__span-21-100"><a id="__codelineno-21-100" name="__codelineno-21-100" href="#__codelineno-21-100"></a><span class="sd">            Generator[tuple[str, str, dict], None, None]:</span>
</span><span id="__span-21-101"><a id="__codelineno-21-101" name="__codelineno-21-101" href="#__codelineno-21-101"></a><span class="sd">                Each tuple is (id, label, properties).</span>
</span><span id="__span-21-102"><a id="__codelineno-21-102" name="__codelineno-21-102" href="#__codelineno-21-102"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-21-103"><a id="__codelineno-21-103" name="__codelineno-21-103" href="#__codelineno-21-103"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading nodes.&quot;</span><span class="p">)</span>
</span><span id="__span-21-104"><a id="__codelineno-21-104" name="__codelineno-21-104" href="#__codelineno-21-104"></a>        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_tsv</span><span class="p">()</span>
</span><span id="__span-21-105"><a id="__codelineno-21-105" name="__codelineno-21-105" href="#__codelineno-21-105"></a>
</span><span id="__span-21-106"><a id="__codelineno-21-106" name="__codelineno-21-106" href="#__codelineno-21-106"></a>        <span class="c1"># Generator for nodes in the `source` column</span>
</span><span id="__span-21-107"><a id="__codelineno-21-107" name="__codelineno-21-107" href="#__codelineno-21-107"></a>        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-21-108"><a id="__codelineno-21-108" name="__codelineno-21-108" href="#__codelineno-21-108"></a>            <span class="nb">id</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">source</span>
</span><span id="__span-21-109"><a id="__codelineno-21-109" name="__codelineno-21-109" href="#__codelineno-21-109"></a>            <span class="n">input_label</span> <span class="o">=</span> <span class="s2">&quot;uniprot_protein&quot;</span>
</span><span id="__span-21-110"><a id="__codelineno-21-110" name="__codelineno-21-110" href="#__codelineno-21-110"></a>
</span><span id="__span-21-111"><a id="__codelineno-21-111" name="__codelineno-21-111" href="#__codelineno-21-111"></a>            <span class="n">properties</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-21-112"><a id="__codelineno-21-112" name="__codelineno-21-112" href="#__codelineno-21-112"></a>                <span class="s1">&#39;genesymbol&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">source_genesymbol</span><span class="p">,</span>
</span><span id="__span-21-113"><a id="__codelineno-21-113" name="__codelineno-21-113" href="#__codelineno-21-113"></a>                <span class="s1">&#39;ncbi_tax_id&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">ncbi_tax_id_source</span><span class="p">,</span>
</span><span id="__span-21-114"><a id="__codelineno-21-114" name="__codelineno-21-114" href="#__codelineno-21-114"></a>                <span class="s1">&#39;entity_type&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">entity_type_source</span><span class="p">,</span>
</span><span id="__span-21-115"><a id="__codelineno-21-115" name="__codelineno-21-115" href="#__codelineno-21-115"></a>            <span class="p">}</span>
</span><span id="__span-21-116"><a id="__codelineno-21-116" name="__codelineno-21-116" href="#__codelineno-21-116"></a>
</span><span id="__span-21-117"><a id="__codelineno-21-117" name="__codelineno-21-117" href="#__codelineno-21-117"></a>            <span class="k">yield</span><span class="p">(</span>
</span><span id="__span-21-118"><a id="__codelineno-21-118" name="__codelineno-21-118" href="#__codelineno-21-118"></a>                <span class="nb">id</span><span class="p">,</span>
</span><span id="__span-21-119"><a id="__codelineno-21-119" name="__codelineno-21-119" href="#__codelineno-21-119"></a>                <span class="n">input_label</span><span class="p">,</span>
</span><span id="__span-21-120"><a id="__codelineno-21-120" name="__codelineno-21-120" href="#__codelineno-21-120"></a>                <span class="n">properties</span>
</span><span id="__span-21-121"><a id="__codelineno-21-121" name="__codelineno-21-121" href="#__codelineno-21-121"></a>            <span class="p">)</span>
</span><span id="__span-21-122"><a id="__codelineno-21-122" name="__codelineno-21-122" href="#__codelineno-21-122"></a>
</span><span id="__span-21-123"><a id="__codelineno-21-123" name="__codelineno-21-123" href="#__codelineno-21-123"></a>        <span class="c1"># Generator for nodes in the `target` column</span>
</span><span id="__span-21-124"><a id="__codelineno-21-124" name="__codelineno-21-124" href="#__codelineno-21-124"></a>        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-21-125"><a id="__codelineno-21-125" name="__codelineno-21-125" href="#__codelineno-21-125"></a>            <span class="nb">id</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">target</span>
</span><span id="__span-21-126"><a id="__codelineno-21-126" name="__codelineno-21-126" href="#__codelineno-21-126"></a>            <span class="n">input_label</span> <span class="o">=</span> <span class="s2">&quot;uniprot_protein&quot;</span>
</span><span id="__span-21-127"><a id="__codelineno-21-127" name="__codelineno-21-127" href="#__codelineno-21-127"></a>
</span><span id="__span-21-128"><a id="__codelineno-21-128" name="__codelineno-21-128" href="#__codelineno-21-128"></a>            <span class="n">properties</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-21-129"><a id="__codelineno-21-129" name="__codelineno-21-129" href="#__codelineno-21-129"></a>                <span class="s1">&#39;genesymbol&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">target_genesymbol</span><span class="p">,</span>
</span><span id="__span-21-130"><a id="__codelineno-21-130" name="__codelineno-21-130" href="#__codelineno-21-130"></a>                <span class="s1">&#39;ncbi_tax_id&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">ncbi_tax_id_target</span><span class="p">,</span>
</span><span id="__span-21-131"><a id="__codelineno-21-131" name="__codelineno-21-131" href="#__codelineno-21-131"></a>                <span class="s1">&#39;entity_type&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">entity_type_target</span><span class="p">,</span>
</span><span id="__span-21-132"><a id="__codelineno-21-132" name="__codelineno-21-132" href="#__codelineno-21-132"></a>            <span class="p">}</span>
</span><span id="__span-21-133"><a id="__codelineno-21-133" name="__codelineno-21-133" href="#__codelineno-21-133"></a>
</span><span id="__span-21-134"><a id="__codelineno-21-134" name="__codelineno-21-134" href="#__codelineno-21-134"></a>            <span class="k">yield</span><span class="p">(</span>
</span><span id="__span-21-135"><a id="__codelineno-21-135" name="__codelineno-21-135" href="#__codelineno-21-135"></a>                <span class="nb">id</span><span class="p">,</span>
</span><span id="__span-21-136"><a id="__codelineno-21-136" name="__codelineno-21-136" href="#__codelineno-21-136"></a>                <span class="n">input_label</span><span class="p">,</span>
</span><span id="__span-21-137"><a id="__codelineno-21-137" name="__codelineno-21-137" href="#__codelineno-21-137"></a>                <span class="n">properties</span>
</span><span id="__span-21-138"><a id="__codelineno-21-138" name="__codelineno-21-138" href="#__codelineno-21-138"></a>            <span class="p">)</span>
</span><span id="__span-21-139"><a id="__codelineno-21-139" name="__codelineno-21-139" href="#__codelineno-21-139"></a>
</span><span id="__span-21-140"><a id="__codelineno-21-140" name="__codelineno-21-140" href="#__codelineno-21-140"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">get_edges</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;Generator[tuple[str, str, str, str, dict], None, None]&#39;</span><span class="p">:</span>
</span><span id="__span-21-141"><a id="__codelineno-21-141" name="__codelineno-21-141" href="#__codelineno-21-141"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-21-142"><a id="__codelineno-21-142" name="__codelineno-21-142" href="#__codelineno-21-142"></a><span class="sd">        Yields edge tuples for edge types specified in the adapter constructor.</span>
</span><span id="__span-21-143"><a id="__codelineno-21-143" name="__codelineno-21-143" href="#__codelineno-21-143"></a>
</span><span id="__span-21-144"><a id="__codelineno-21-144" name="__codelineno-21-144" href="#__codelineno-21-144"></a><span class="sd">        Returns:</span>
</span><span id="__span-21-145"><a id="__codelineno-21-145" name="__codelineno-21-145" href="#__codelineno-21-145"></a><span class="sd">            Generator[tuple[str, str, str, str, dict], None, None]:</span>
</span><span id="__span-21-146"><a id="__codelineno-21-146" name="__codelineno-21-146" href="#__codelineno-21-146"></a><span class="sd">                Each tuple is (id, source, target, type, properties).</span>
</span><span id="__span-21-147"><a id="__codelineno-21-147" name="__codelineno-21-147" href="#__codelineno-21-147"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-21-148"><a id="__codelineno-21-148" name="__codelineno-21-148" href="#__codelineno-21-148"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Generating edges.&quot;</span><span class="p">)</span>
</span><span id="__span-21-149"><a id="__codelineno-21-149" name="__codelineno-21-149" href="#__codelineno-21-149"></a>        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_tsv</span><span class="p">()</span>
</span><span id="__span-21-150"><a id="__codelineno-21-150" name="__codelineno-21-150" href="#__codelineno-21-150"></a>
</span><span id="__span-21-151"><a id="__codelineno-21-151" name="__codelineno-21-151" href="#__codelineno-21-151"></a>        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-21-152"><a id="__codelineno-21-152" name="__codelineno-21-152" href="#__codelineno-21-152"></a>            <span class="c1"># Concatenate source and target, i.e., &quot;SOD1EGFR&quot;</span>
</span><span id="__span-21-153"><a id="__codelineno-21-153" name="__codelineno-21-153" href="#__codelineno-21-153"></a>            <span class="nb">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">source</span><span class="si">}{</span><span class="n">row</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s2">&quot;</span>
</span><span id="__span-21-154"><a id="__codelineno-21-154" name="__codelineno-21-154" href="#__codelineno-21-154"></a>
</span><span id="__span-21-155"><a id="__codelineno-21-155" name="__codelineno-21-155" href="#__codelineno-21-155"></a>            <span class="n">source</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">source</span>
</span><span id="__span-21-156"><a id="__codelineno-21-156" name="__codelineno-21-156" href="#__codelineno-21-156"></a>
</span><span id="__span-21-157"><a id="__codelineno-21-157" name="__codelineno-21-157" href="#__codelineno-21-157"></a>            <span class="n">target</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">target</span>
</span><span id="__span-21-158"><a id="__codelineno-21-158" name="__codelineno-21-158" href="#__codelineno-21-158"></a>
</span><span id="__span-21-159"><a id="__codelineno-21-159" name="__codelineno-21-159" href="#__codelineno-21-159"></a>            <span class="nb">type</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">type</span>
</span><span id="__span-21-160"><a id="__codelineno-21-160" name="__codelineno-21-160" href="#__codelineno-21-160"></a>
</span><span id="__span-21-161"><a id="__codelineno-21-161" name="__codelineno-21-161" href="#__codelineno-21-161"></a>            <span class="n">properties</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-21-162"><a id="__codelineno-21-162" name="__codelineno-21-162" href="#__codelineno-21-162"></a>                <span class="s1">&#39;is_directed&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">is_directed</span><span class="p">,</span>
</span><span id="__span-21-163"><a id="__codelineno-21-163" name="__codelineno-21-163" href="#__codelineno-21-163"></a>                <span class="s1">&#39;is_stimulation&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">is_stimulation</span><span class="p">,</span>
</span><span id="__span-21-164"><a id="__codelineno-21-164" name="__codelineno-21-164" href="#__codelineno-21-164"></a>                <span class="s1">&#39;is_inhibition&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">is_inhibition</span><span class="p">,</span>
</span><span id="__span-21-165"><a id="__codelineno-21-165" name="__codelineno-21-165" href="#__codelineno-21-165"></a>                <span class="s1">&#39;consensus_direction&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">consensus_direction</span><span class="p">,</span>
</span><span id="__span-21-166"><a id="__codelineno-21-166" name="__codelineno-21-166" href="#__codelineno-21-166"></a>                <span class="s1">&#39;consensus_stimulation&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">consensus_stimulation</span><span class="p">,</span>
</span><span id="__span-21-167"><a id="__codelineno-21-167" name="__codelineno-21-167" href="#__codelineno-21-167"></a>                <span class="s1">&#39;consensus_inhibition&#39;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">consensus_inhibition</span>
</span><span id="__span-21-168"><a id="__codelineno-21-168" name="__codelineno-21-168" href="#__codelineno-21-168"></a>            <span class="p">}</span>
</span><span id="__span-21-169"><a id="__codelineno-21-169" name="__codelineno-21-169" href="#__codelineno-21-169"></a>
</span><span id="__span-21-170"><a id="__codelineno-21-170" name="__codelineno-21-170" href="#__codelineno-21-170"></a>            <span class="k">yield</span> <span class="p">(</span>
</span><span id="__span-21-171"><a id="__codelineno-21-171" name="__codelineno-21-171" href="#__codelineno-21-171"></a>                <span class="nb">id</span><span class="p">,</span>
</span><span id="__span-21-172"><a id="__codelineno-21-172" name="__codelineno-21-172" href="#__codelineno-21-172"></a>                <span class="n">source</span><span class="p">,</span>
</span><span id="__span-21-173"><a id="__codelineno-21-173" name="__codelineno-21-173" href="#__codelineno-21-173"></a>                <span class="n">target</span><span class="p">,</span>
</span><span id="__span-21-174"><a id="__codelineno-21-174" name="__codelineno-21-174" href="#__codelineno-21-174"></a>                <span class="nb">type</span><span class="p">,</span>
</span><span id="__span-21-175"><a id="__codelineno-21-175" name="__codelineno-21-175" href="#__codelineno-21-175"></a>                <span class="n">properties</span>
</span><span id="__span-21-176"><a id="__codelineno-21-176" name="__codelineno-21-176" href="#__codelineno-21-176"></a>            <span class="p">)</span>
</span><span id="__span-21-177"><a id="__codelineno-21-177" name="__codelineno-21-177" href="#__codelineno-21-177"></a>
</span><span id="__span-21-178"><a id="__codelineno-21-178" name="__codelineno-21-178" href="#__codelineno-21-178"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">get_node_count</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
</span><span id="__span-21-179"><a id="__codelineno-21-179" name="__codelineno-21-179" href="#__codelineno-21-179"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-21-180"><a id="__codelineno-21-180" name="__codelineno-21-180" href="#__codelineno-21-180"></a><span class="sd">        Returns the number of nodes generated by the adapter.</span>
</span><span id="__span-21-181"><a id="__codelineno-21-181" name="__codelineno-21-181" href="#__codelineno-21-181"></a>
</span><span id="__span-21-182"><a id="__codelineno-21-182" name="__codelineno-21-182" href="#__codelineno-21-182"></a><span class="sd">        Returns:</span>
</span><span id="__span-21-183"><a id="__codelineno-21-183" name="__codelineno-21-183" href="#__codelineno-21-183"></a><span class="sd">            int: Number of nodes generated.</span>
</span><span id="__span-21-184"><a id="__codelineno-21-184" name="__codelineno-21-184" href="#__codelineno-21-184"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-21-185"><a id="__codelineno-21-185" name="__codelineno-21-185" href="#__codelineno-21-185"></a>        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_nodes</span><span class="p">())</span>
</span><span id="__span-21-186"><a id="__codelineno-21-186" name="__codelineno-21-186" href="#__codelineno-21-186"></a>
</span><span id="__span-21-187"><a id="__codelineno-21-187" name="__codelineno-21-187" href="#__codelineno-21-187"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">_set_types_and_fields</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node_types</span><span class="p">,</span> <span class="n">node_fields</span><span class="p">,</span> <span class="n">edge_types</span><span class="p">,</span> <span class="n">edge_fields</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-21-188"><a id="__codelineno-21-188" name="__codelineno-21-188" href="#__codelineno-21-188"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-21-189"><a id="__codelineno-21-189" name="__codelineno-21-189" href="#__codelineno-21-189"></a><span class="sd">        Sets the node and edge types and fields for the adapter.</span>
</span><span id="__span-21-190"><a id="__codelineno-21-190" name="__codelineno-21-190" href="#__codelineno-21-190"></a>
</span><span id="__span-21-191"><a id="__codelineno-21-191" name="__codelineno-21-191" href="#__codelineno-21-191"></a><span class="sd">        Args:</span>
</span><span id="__span-21-192"><a id="__codelineno-21-192" name="__codelineno-21-192" href="#__codelineno-21-192"></a><span class="sd">            node_types (Optional[list]): List of node types.</span>
</span><span id="__span-21-193"><a id="__codelineno-21-193" name="__codelineno-21-193" href="#__codelineno-21-193"></a><span class="sd">            node_fields (Optional[list]): List of node fields.</span>
</span><span id="__span-21-194"><a id="__codelineno-21-194" name="__codelineno-21-194" href="#__codelineno-21-194"></a><span class="sd">            edge_types (Optional[list]): List of edge types.</span>
</span><span id="__span-21-195"><a id="__codelineno-21-195" name="__codelineno-21-195" href="#__codelineno-21-195"></a><span class="sd">            edge_fields (Optional[list]): List of edge fields.</span>
</span><span id="__span-21-196"><a id="__codelineno-21-196" name="__codelineno-21-196" href="#__codelineno-21-196"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-21-197"><a id="__codelineno-21-197" name="__codelineno-21-197" href="#__codelineno-21-197"></a>        <span class="k">if</span> <span class="n">node_types</span><span class="p">:</span>
</span><span id="__span-21-198"><a id="__codelineno-21-198" name="__codelineno-21-198" href="#__codelineno-21-198"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">node_types</span> <span class="o">=</span> <span class="n">node_types</span>
</span><span id="__span-21-199"><a id="__codelineno-21-199" name="__codelineno-21-199" href="#__codelineno-21-199"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-21-200"><a id="__codelineno-21-200" name="__codelineno-21-200" href="#__codelineno-21-200"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">node_types</span> <span class="o">=</span> <span class="p">[</span><span class="nb">type</span> <span class="k">for</span> <span class="nb">type</span> <span class="ow">in</span> <span class="n">AdapterNodeType</span><span class="p">]</span>
</span><span id="__span-21-201"><a id="__codelineno-21-201" name="__codelineno-21-201" href="#__codelineno-21-201"></a>
</span><span id="__span-21-202"><a id="__codelineno-21-202" name="__codelineno-21-202" href="#__codelineno-21-202"></a>        <span class="k">if</span> <span class="n">node_fields</span><span class="p">:</span>
</span><span id="__span-21-203"><a id="__codelineno-21-203" name="__codelineno-21-203" href="#__codelineno-21-203"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">node_fields</span> <span class="o">=</span> <span class="n">node_fields</span>
</span><span id="__span-21-204"><a id="__codelineno-21-204" name="__codelineno-21-204" href="#__codelineno-21-204"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-21-205"><a id="__codelineno-21-205" name="__codelineno-21-205" href="#__codelineno-21-205"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">node_fields</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-21-206"><a id="__codelineno-21-206" name="__codelineno-21-206" href="#__codelineno-21-206"></a>                <span class="n">field</span>
</span><span id="__span-21-207"><a id="__codelineno-21-207" name="__codelineno-21-207" href="#__codelineno-21-207"></a>                <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">(</span>
</span><span id="__span-21-208"><a id="__codelineno-21-208" name="__codelineno-21-208" href="#__codelineno-21-208"></a>                    <span class="n">AdapterProteinField</span><span class="p">,</span>
</span><span id="__span-21-209"><a id="__codelineno-21-209" name="__codelineno-21-209" href="#__codelineno-21-209"></a>                <span class="p">)</span>
</span><span id="__span-21-210"><a id="__codelineno-21-210" name="__codelineno-21-210" href="#__codelineno-21-210"></a>            <span class="p">]</span>
</span><span id="__span-21-211"><a id="__codelineno-21-211" name="__codelineno-21-211" href="#__codelineno-21-211"></a>
</span><span id="__span-21-212"><a id="__codelineno-21-212" name="__codelineno-21-212" href="#__codelineno-21-212"></a>        <span class="k">if</span> <span class="n">edge_types</span><span class="p">:</span>
</span><span id="__span-21-213"><a id="__codelineno-21-213" name="__codelineno-21-213" href="#__codelineno-21-213"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">edge_types</span> <span class="o">=</span> <span class="n">edge_types</span>
</span><span id="__span-21-214"><a id="__codelineno-21-214" name="__codelineno-21-214" href="#__codelineno-21-214"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-21-215"><a id="__codelineno-21-215" name="__codelineno-21-215" href="#__codelineno-21-215"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">edge_types</span> <span class="o">=</span> <span class="p">[</span><span class="nb">type</span> <span class="k">for</span> <span class="nb">type</span> <span class="ow">in</span> <span class="n">AdapterEdgeType</span><span class="p">]</span>
</span><span id="__span-21-216"><a id="__codelineno-21-216" name="__codelineno-21-216" href="#__codelineno-21-216"></a>
</span><span id="__span-21-217"><a id="__codelineno-21-217" name="__codelineno-21-217" href="#__codelineno-21-217"></a>        <span class="k">if</span> <span class="n">edge_fields</span><span class="p">:</span>
</span><span id="__span-21-218"><a id="__codelineno-21-218" name="__codelineno-21-218" href="#__codelineno-21-218"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">edge_fields</span> <span class="o">=</span> <span class="n">edge_fields</span>
</span><span id="__span-21-219"><a id="__codelineno-21-219" name="__codelineno-21-219" href="#__codelineno-21-219"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-21-220"><a id="__codelineno-21-220" name="__codelineno-21-220" href="#__codelineno-21-220"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">edge_fields</span> <span class="o">=</span> <span class="p">[</span><span class="n">field</span> <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">()]</span>
</span></code></pre></div>
</details>
<h3 id="step-3-create-a-knowledge-graph-script">Step 3. Create a knowledge graph script</h3>
<figure>
<img alt="Image title" src="../assets/biocypher_section_script.png" width="1000" />

<figcaption>Figure 12. BioCypher pipeline</figcaption>
</figure>
<p><strong>Rationale:</strong> Integrating all steps—downloading the dataset, loading the data, extracting nodes and edges, and exporting graph assets—into a single script streamlines the entire process. This approach makes it easier to build and manage the knowledge graph pipeline efficiently and reproducibly.</p>
<ol>
<li>
<p>Create a BioCypher object</p>
<details class="example">
<summary><strong>File: <code>create_knowledge_graph.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-22-1"><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a><span class="kn">from</span><span class="w"> </span><span class="nn">biocypher</span><span class="w"> </span><span class="kn">import</span> <span class="n">BioCypher</span><span class="p">,</span> <span class="n">FileDownload</span>
</span><span id="__span-22-2"><a id="__codelineno-22-2" name="__codelineno-22-2" href="#__codelineno-22-2"></a>
</span><span id="__span-22-3"><a id="__codelineno-22-3" name="__codelineno-22-3" href="#__codelineno-22-3"></a><span class="c1"># Create an instance of BioCypher</span>
</span><span id="__span-22-4"><a id="__codelineno-22-4" name="__codelineno-22-4" href="#__codelineno-22-4"></a><span class="n">bc</span> <span class="o">=</span> <span class="n">BioCypher</span><span class="p">()</span>
</span></code></pre></div>
</details>
</li>
<li>
<p>Use BioCypher to automatically download the file with cache capabilities.</p>
<details class="example">
<summary><strong>File: <code>create_knowledge_graph.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-23-1"><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a><span class="c1"># Download the file with cache capabilities</span>
</span><span id="__span-23-2"><a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a><span class="n">url_dataset</span> <span class="o">=</span> <span class="p">(</span>
</span><span id="__span-23-3"><a id="__codelineno-23-3" name="__codelineno-23-3" href="#__codelineno-23-3"></a>    <span class="s2">&quot;https://zenodo.org/records/16902349/files/synthetic_protein_interactions.tsv&quot;</span>
</span><span id="__span-23-4"><a id="__codelineno-23-4" name="__codelineno-23-4" href="#__codelineno-23-4"></a><span class="p">)</span>
</span><span id="__span-23-5"><a id="__codelineno-23-5" name="__codelineno-23-5" href="#__codelineno-23-5"></a>
</span><span id="__span-23-6"><a id="__codelineno-23-6" name="__codelineno-23-6" href="#__codelineno-23-6"></a><span class="n">resource</span> <span class="o">=</span> <span class="n">FileDownload</span><span class="p">(</span>
</span><span id="__span-23-7"><a id="__codelineno-23-7" name="__codelineno-23-7" href="#__codelineno-23-7"></a>    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;protein-protein-interaction-dataset&quot;</span><span class="p">,</span>  <span class="c1"># Name of the resource</span>
</span><span id="__span-23-8"><a id="__codelineno-23-8" name="__codelineno-23-8" href="#__codelineno-23-8"></a>    <span class="n">url_s</span><span class="o">=</span><span class="n">url_dataset</span><span class="p">,</span>  <span class="c1"># URL to the resource(s)</span>
</span><span id="__span-23-9"><a id="__codelineno-23-9" name="__codelineno-23-9" href="#__codelineno-23-9"></a>    <span class="n">lifetime</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>  <span class="c1"># seven days cache lifetime</span>
</span><span id="__span-23-10"><a id="__codelineno-23-10" name="__codelineno-23-10" href="#__codelineno-23-10"></a><span class="p">)</span>
</span><span id="__span-23-11"><a id="__codelineno-23-11" name="__codelineno-23-11" href="#__codelineno-23-11"></a><span class="n">paths</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="n">resource</span><span class="p">)</span>  <span class="c1"># Downloads to &#39;.cache&#39; by default</span>
</span><span id="__span-23-12"><a id="__codelineno-23-12" name="__codelineno-23-12" href="#__codelineno-23-12"></a>
</span><span id="__span-23-13"><a id="__codelineno-23-13" name="__codelineno-23-13" href="#__codelineno-23-13"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Path to the resouce: </span><span class="si">{</span><span class="n">paths</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span></code></pre></div>
</details>
</li>
<li>
<p>Instantiate your adapter from the classes you created previously in your adapter file.</p>
<details class="example">
<summary><strong>File: <code>create_knowledge_graph.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-24-1"><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a><span class="kn">from</span><span class="w"> </span><span class="nn">template_package.adapters.adapter_synthetic_proteins</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
</span><span id="__span-24-2"><a id="__codelineno-24-2" name="__codelineno-24-2" href="#__codelineno-24-2"></a>    <span class="n">AdapterNodeType</span><span class="p">,</span>
</span><span id="__span-24-3"><a id="__codelineno-24-3" name="__codelineno-24-3" href="#__codelineno-24-3"></a>    <span class="n">AdapterProteinField</span><span class="p">,</span>
</span><span id="__span-24-4"><a id="__codelineno-24-4" name="__codelineno-24-4" href="#__codelineno-24-4"></a>    <span class="n">AdapterEdgeType</span><span class="p">,</span>
</span><span id="__span-24-5"><a id="__codelineno-24-5" name="__codelineno-24-5" href="#__codelineno-24-5"></a>    <span class="n">Adapter</span><span class="p">,</span>
</span><span id="__span-24-6"><a id="__codelineno-24-6" name="__codelineno-24-6" href="#__codelineno-24-6"></a><span class="p">)</span>
</span><span id="__span-24-7"><a id="__codelineno-24-7" name="__codelineno-24-7" href="#__codelineno-24-7"></a>
</span><span id="__span-24-8"><a id="__codelineno-24-8" name="__codelineno-24-8" href="#__codelineno-24-8"></a><span class="c1"># Choose the node type you want appear in the Knowledge Graph</span>
</span><span id="__span-24-9"><a id="__codelineno-24-9" name="__codelineno-24-9" href="#__codelineno-24-9"></a><span class="n">node_types</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-24-10"><a id="__codelineno-24-10" name="__codelineno-24-10" href="#__codelineno-24-10"></a>    <span class="n">AdapterNodeType</span><span class="o">.</span><span class="n">PROTEIN</span>
</span><span id="__span-24-11"><a id="__codelineno-24-11" name="__codelineno-24-11" href="#__codelineno-24-11"></a><span class="p">]</span>
</span><span id="__span-24-12"><a id="__codelineno-24-12" name="__codelineno-24-12" href="#__codelineno-24-12"></a>
</span><span id="__span-24-13"><a id="__codelineno-24-13" name="__codelineno-24-13" href="#__codelineno-24-13"></a><span class="c1"># Choose protein adapter fields to include in the knowledge graph.</span>
</span><span id="__span-24-14"><a id="__codelineno-24-14" name="__codelineno-24-14" href="#__codelineno-24-14"></a><span class="n">node_fields</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-24-15"><a id="__codelineno-24-15" name="__codelineno-24-15" href="#__codelineno-24-15"></a>    <span class="n">AdapterProteinField</span><span class="o">.</span><span class="n">ID</span><span class="p">,</span>
</span><span id="__span-24-16"><a id="__codelineno-24-16" name="__codelineno-24-16" href="#__codelineno-24-16"></a>    <span class="n">AdapterProteinField</span><span class="o">.</span><span class="n">PREFERRED_ID</span><span class="p">,</span>
</span><span id="__span-24-17"><a id="__codelineno-24-17" name="__codelineno-24-17" href="#__codelineno-24-17"></a>    <span class="n">AdapterProteinField</span><span class="o">.</span><span class="n">GENE_SYMBOL</span><span class="p">,</span>
</span><span id="__span-24-18"><a id="__codelineno-24-18" name="__codelineno-24-18" href="#__codelineno-24-18"></a>    <span class="n">AdapterProteinField</span><span class="o">.</span><span class="n">NCBI_TAX_ID</span>
</span><span id="__span-24-19"><a id="__codelineno-24-19" name="__codelineno-24-19" href="#__codelineno-24-19"></a><span class="p">]</span>
</span><span id="__span-24-20"><a id="__codelineno-24-20" name="__codelineno-24-20" href="#__codelineno-24-20"></a>
</span><span id="__span-24-21"><a id="__codelineno-24-21" name="__codelineno-24-21" href="#__codelineno-24-21"></a><span class="c1"># Choose the node type you want appear in the Knowledge Graph</span>
</span><span id="__span-24-22"><a id="__codelineno-24-22" name="__codelineno-24-22" href="#__codelineno-24-22"></a><span class="n">edge_types</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-24-23"><a id="__codelineno-24-23" name="__codelineno-24-23" href="#__codelineno-24-23"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">PROTEIN_PROTEIN_INTERACTION</span><span class="p">,</span>
</span><span id="__span-24-24"><a id="__codelineno-24-24" name="__codelineno-24-24" href="#__codelineno-24-24"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">BINDING</span><span class="p">,</span>
</span><span id="__span-24-25"><a id="__codelineno-24-25" name="__codelineno-24-25" href="#__codelineno-24-25"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">ACTIVATION</span><span class="p">,</span>
</span><span id="__span-24-26"><a id="__codelineno-24-26" name="__codelineno-24-26" href="#__codelineno-24-26"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">PHOSPHORYLATION</span><span class="p">,</span>
</span><span id="__span-24-27"><a id="__codelineno-24-27" name="__codelineno-24-27" href="#__codelineno-24-27"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">UBIQUITINATION</span><span class="p">,</span>
</span><span id="__span-24-28"><a id="__codelineno-24-28" name="__codelineno-24-28" href="#__codelineno-24-28"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">INHIBITION</span>
</span><span id="__span-24-29"><a id="__codelineno-24-29" name="__codelineno-24-29" href="#__codelineno-24-29"></a><span class="p">]</span>
</span><span id="__span-24-30"><a id="__codelineno-24-30" name="__codelineno-24-30" href="#__codelineno-24-30"></a>
</span><span id="__span-24-31"><a id="__codelineno-24-31" name="__codelineno-24-31" href="#__codelineno-24-31"></a><span class="c1"># (there is not code here!) Choose interaction adapter fields to include in the knowledge graph.</span>
</span><span id="__span-24-32"><a id="__codelineno-24-32" name="__codelineno-24-32" href="#__codelineno-24-32"></a><span class="c1"># By default, in case of not specifying this, BioCypher will bring all the fields defined in the adapter</span>
</span><span id="__span-24-33"><a id="__codelineno-24-33" name="__codelineno-24-33" href="#__codelineno-24-33"></a>
</span><span id="__span-24-34"><a id="__codelineno-24-34" name="__codelineno-24-34" href="#__codelineno-24-34"></a><span class="c1"># Create an adapter instance</span>
</span><span id="__span-24-35"><a id="__codelineno-24-35" name="__codelineno-24-35" href="#__codelineno-24-35"></a><span class="n">adapter</span> <span class="o">=</span> <span class="n">Adapter</span><span class="p">(</span>
</span><span id="__span-24-36"><a id="__codelineno-24-36" name="__codelineno-24-36" href="#__codelineno-24-36"></a>    <span class="n">tsv_path</span> <span class="o">=</span> <span class="n">paths</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
</span><span id="__span-24-37"><a id="__codelineno-24-37" name="__codelineno-24-37" href="#__codelineno-24-37"></a>    <span class="n">node_types</span><span class="o">=</span><span class="n">node_types</span><span class="p">,</span>
</span><span id="__span-24-38"><a id="__codelineno-24-38" name="__codelineno-24-38" href="#__codelineno-24-38"></a>    <span class="n">node_fields</span><span class="o">=</span><span class="n">node_fields</span><span class="p">,</span>
</span><span id="__span-24-39"><a id="__codelineno-24-39" name="__codelineno-24-39" href="#__codelineno-24-39"></a>    <span class="n">edge_types</span><span class="o">=</span><span class="n">edge_types</span><span class="p">,</span>
</span><span id="__span-24-40"><a id="__codelineno-24-40" name="__codelineno-24-40" href="#__codelineno-24-40"></a><span class="p">)</span>
</span></code></pre></div>
</details>
</li>
<li>
<p>Write data from your adapter to BioCypher</p>
<details class="example">
<summary><strong>File: <code>create_knowledge_graph.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-25-1"><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a><span class="c1"># Create a knowledge graph from the adapter</span>
</span><span id="__span-25-2"><a id="__codelineno-25-2" name="__codelineno-25-2" href="#__codelineno-25-2"></a><span class="n">bc</span><span class="o">.</span><span class="n">write_nodes</span><span class="p">(</span><span class="n">adapter</span><span class="o">.</span><span class="n">get_nodes</span><span class="p">())</span>
</span><span id="__span-25-3"><a id="__codelineno-25-3" name="__codelineno-25-3" href="#__codelineno-25-3"></a><span class="n">bc</span><span class="o">.</span><span class="n">write_edges</span><span class="p">(</span><span class="n">adapter</span><span class="o">.</span><span class="n">get_edges</span><span class="p">())</span>
</span></code></pre></div>
</details>
</li>
<li>
<p>Export your graph to Neo4j (generation of TSV files and import script)</p>
<details class="example">
<summary><strong>File: <code>create_knowledge_graph.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-26-1"><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a><span class="c1"># Generate assets for Neo4j exportation</span>
</span><span id="__span-26-2"><a id="__codelineno-26-2" name="__codelineno-26-2" href="#__codelineno-26-2"></a><span class="n">bc</span><span class="o">.</span><span class="n">write_import_call</span><span class="p">()</span>
</span></code></pre></div>
</details>
</li>
<li>
<p>Print summary</p>
<details class="example">
<summary><strong>File: <code>create_knowledge_graph.py</code></strong></summary>
<div class="language-python highlight"><pre><span></span><code><span id="__span-27-1"><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a><span class="c1"># Print a summary when</span>
</span><span id="__span-27-2"><a id="__codelineno-27-2" name="__codelineno-27-2" href="#__codelineno-27-2"></a><span class="n">bc</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
</span></code></pre></div>
</details>
</li>
</ol>
<blockquote>
<p>📝 <strong>Exercise:</strong>
Integrate the aforementioned snippets in a single file called <code>create_knowledge_graph.py</code> script and <strong>RUN IT</strong>!</p>
</blockquote>
<details class="success">
<summary>Answer:</summary>
<p>See the example below for a completed <code>create_knowledge_graph.yaml</code>.</p>
<p><strong>File: <code>create_knowledge_graph.py</code></strong></p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-28-1"><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a><span class="kn">from</span><span class="w"> </span><span class="nn">biocypher</span><span class="w"> </span><span class="kn">import</span> <span class="n">BioCypher</span><span class="p">,</span> <span class="n">FileDownload</span>
</span><span id="__span-28-2"><a id="__codelineno-28-2" name="__codelineno-28-2" href="#__codelineno-28-2"></a><span class="kn">from</span><span class="w"> </span><span class="nn">template_package.adapters.adapter_synthetic_proteins</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
</span><span id="__span-28-3"><a id="__codelineno-28-3" name="__codelineno-28-3" href="#__codelineno-28-3"></a>    <span class="n">AdapterNodeType</span><span class="p">,</span>
</span><span id="__span-28-4"><a id="__codelineno-28-4" name="__codelineno-28-4" href="#__codelineno-28-4"></a>    <span class="n">AdapterProteinField</span><span class="p">,</span>
</span><span id="__span-28-5"><a id="__codelineno-28-5" name="__codelineno-28-5" href="#__codelineno-28-5"></a>    <span class="n">AdapterEdgeType</span><span class="p">,</span>
</span><span id="__span-28-6"><a id="__codelineno-28-6" name="__codelineno-28-6" href="#__codelineno-28-6"></a>    <span class="n">Adapter</span><span class="p">,</span>
</span><span id="__span-28-7"><a id="__codelineno-28-7" name="__codelineno-28-7" href="#__codelineno-28-7"></a><span class="p">)</span>
</span><span id="__span-28-8"><a id="__codelineno-28-8" name="__codelineno-28-8" href="#__codelineno-28-8"></a>
</span><span id="__span-28-9"><a id="__codelineno-28-9" name="__codelineno-28-9" href="#__codelineno-28-9"></a><span class="c1"># Create an instance of BioCypher</span>
</span><span id="__span-28-10"><a id="__codelineno-28-10" name="__codelineno-28-10" href="#__codelineno-28-10"></a><span class="n">bc</span> <span class="o">=</span> <span class="n">BioCypher</span><span class="p">()</span>
</span><span id="__span-28-11"><a id="__codelineno-28-11" name="__codelineno-28-11" href="#__codelineno-28-11"></a>
</span><span id="__span-28-12"><a id="__codelineno-28-12" name="__codelineno-28-12" href="#__codelineno-28-12"></a><span class="c1"># Download the file with cache capabilities</span>
</span><span id="__span-28-13"><a id="__codelineno-28-13" name="__codelineno-28-13" href="#__codelineno-28-13"></a><span class="n">url_dataset</span> <span class="o">=</span> <span class="p">(</span>
</span><span id="__span-28-14"><a id="__codelineno-28-14" name="__codelineno-28-14" href="#__codelineno-28-14"></a>    <span class="s2">&quot;https://zenodo.org/records/16902349/files/synthetic_protein_interactions.tsv&quot;</span>
</span><span id="__span-28-15"><a id="__codelineno-28-15" name="__codelineno-28-15" href="#__codelineno-28-15"></a><span class="p">)</span>
</span><span id="__span-28-16"><a id="__codelineno-28-16" name="__codelineno-28-16" href="#__codelineno-28-16"></a>
</span><span id="__span-28-17"><a id="__codelineno-28-17" name="__codelineno-28-17" href="#__codelineno-28-17"></a><span class="n">resource</span> <span class="o">=</span> <span class="n">FileDownload</span><span class="p">(</span>
</span><span id="__span-28-18"><a id="__codelineno-28-18" name="__codelineno-28-18" href="#__codelineno-28-18"></a>    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;protein-protein-interaction-dataset&quot;</span><span class="p">,</span>  <span class="c1"># Name of the resource</span>
</span><span id="__span-28-19"><a id="__codelineno-28-19" name="__codelineno-28-19" href="#__codelineno-28-19"></a>    <span class="n">url_s</span><span class="o">=</span><span class="n">url_dataset</span><span class="p">,</span>  <span class="c1"># URL to the resource(s)</span>
</span><span id="__span-28-20"><a id="__codelineno-28-20" name="__codelineno-28-20" href="#__codelineno-28-20"></a>    <span class="n">lifetime</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>  <span class="c1"># seven days cache lifetime</span>
</span><span id="__span-28-21"><a id="__codelineno-28-21" name="__codelineno-28-21" href="#__codelineno-28-21"></a><span class="p">)</span>
</span><span id="__span-28-22"><a id="__codelineno-28-22" name="__codelineno-28-22" href="#__codelineno-28-22"></a><span class="n">paths</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="n">resource</span><span class="p">)</span>  <span class="c1"># Downloads to &#39;.cache&#39; by default</span>
</span><span id="__span-28-23"><a id="__codelineno-28-23" name="__codelineno-28-23" href="#__codelineno-28-23"></a>
</span><span id="__span-28-24"><a id="__codelineno-28-24" name="__codelineno-28-24" href="#__codelineno-28-24"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Path to the resouce: </span><span class="si">{</span><span class="n">paths</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-28-25"><a id="__codelineno-28-25" name="__codelineno-28-25" href="#__codelineno-28-25"></a>
</span><span id="__span-28-26"><a id="__codelineno-28-26" name="__codelineno-28-26" href="#__codelineno-28-26"></a>
</span><span id="__span-28-27"><a id="__codelineno-28-27" name="__codelineno-28-27" href="#__codelineno-28-27"></a><span class="c1"># Choose the node type you want appear in the Knowledge Graph</span>
</span><span id="__span-28-28"><a id="__codelineno-28-28" name="__codelineno-28-28" href="#__codelineno-28-28"></a><span class="n">node_types</span> <span class="o">=</span> <span class="p">[</span><span class="n">AdapterNodeType</span><span class="o">.</span><span class="n">PROTEIN</span><span class="p">]</span>
</span><span id="__span-28-29"><a id="__codelineno-28-29" name="__codelineno-28-29" href="#__codelineno-28-29"></a>
</span><span id="__span-28-30"><a id="__codelineno-28-30" name="__codelineno-28-30" href="#__codelineno-28-30"></a><span class="c1"># Choose protein adapter fields to include in the knowledge graph.</span>
</span><span id="__span-28-31"><a id="__codelineno-28-31" name="__codelineno-28-31" href="#__codelineno-28-31"></a><span class="n">node_fields</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-28-32"><a id="__codelineno-28-32" name="__codelineno-28-32" href="#__codelineno-28-32"></a>    <span class="n">AdapterProteinField</span><span class="o">.</span><span class="n">ID</span><span class="p">,</span>
</span><span id="__span-28-33"><a id="__codelineno-28-33" name="__codelineno-28-33" href="#__codelineno-28-33"></a>    <span class="n">AdapterProteinField</span><span class="o">.</span><span class="n">PREFERRED_ID</span><span class="p">,</span>
</span><span id="__span-28-34"><a id="__codelineno-28-34" name="__codelineno-28-34" href="#__codelineno-28-34"></a>    <span class="n">AdapterProteinField</span><span class="o">.</span><span class="n">GENE_SYMBOL</span><span class="p">,</span>
</span><span id="__span-28-35"><a id="__codelineno-28-35" name="__codelineno-28-35" href="#__codelineno-28-35"></a>    <span class="n">AdapterProteinField</span><span class="o">.</span><span class="n">NCBI_TAX_ID</span><span class="p">,</span>
</span><span id="__span-28-36"><a id="__codelineno-28-36" name="__codelineno-28-36" href="#__codelineno-28-36"></a><span class="p">]</span>
</span><span id="__span-28-37"><a id="__codelineno-28-37" name="__codelineno-28-37" href="#__codelineno-28-37"></a>
</span><span id="__span-28-38"><a id="__codelineno-28-38" name="__codelineno-28-38" href="#__codelineno-28-38"></a><span class="c1"># Choose the node type you want appear in the Knowledge Graph</span>
</span><span id="__span-28-39"><a id="__codelineno-28-39" name="__codelineno-28-39" href="#__codelineno-28-39"></a><span class="n">edge_types</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-28-40"><a id="__codelineno-28-40" name="__codelineno-28-40" href="#__codelineno-28-40"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">PROTEIN_PROTEIN_INTERACTION</span><span class="p">,</span>
</span><span id="__span-28-41"><a id="__codelineno-28-41" name="__codelineno-28-41" href="#__codelineno-28-41"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">BINDING</span><span class="p">,</span>
</span><span id="__span-28-42"><a id="__codelineno-28-42" name="__codelineno-28-42" href="#__codelineno-28-42"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">ACTIVATION</span><span class="p">,</span>
</span><span id="__span-28-43"><a id="__codelineno-28-43" name="__codelineno-28-43" href="#__codelineno-28-43"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">PHOSPHORYLATION</span><span class="p">,</span>
</span><span id="__span-28-44"><a id="__codelineno-28-44" name="__codelineno-28-44" href="#__codelineno-28-44"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">UBIQUITINATION</span><span class="p">,</span>
</span><span id="__span-28-45"><a id="__codelineno-28-45" name="__codelineno-28-45" href="#__codelineno-28-45"></a>    <span class="n">AdapterEdgeType</span><span class="o">.</span><span class="n">INHIBITION</span><span class="p">,</span>
</span><span id="__span-28-46"><a id="__codelineno-28-46" name="__codelineno-28-46" href="#__codelineno-28-46"></a><span class="p">]</span>
</span><span id="__span-28-47"><a id="__codelineno-28-47" name="__codelineno-28-47" href="#__codelineno-28-47"></a>
</span><span id="__span-28-48"><a id="__codelineno-28-48" name="__codelineno-28-48" href="#__codelineno-28-48"></a><span class="c1"># (there is not code here!) Choose interaction adapter fields to include in the knowledge graph.</span>
</span><span id="__span-28-49"><a id="__codelineno-28-49" name="__codelineno-28-49" href="#__codelineno-28-49"></a><span class="c1"># By default, in case of not specifying this, BioCypher will bring all the fields defined in the adapter</span>
</span><span id="__span-28-50"><a id="__codelineno-28-50" name="__codelineno-28-50" href="#__codelineno-28-50"></a>
</span><span id="__span-28-51"><a id="__codelineno-28-51" name="__codelineno-28-51" href="#__codelineno-28-51"></a><span class="c1"># Create an adapter instance</span>
</span><span id="__span-28-52"><a id="__codelineno-28-52" name="__codelineno-28-52" href="#__codelineno-28-52"></a><span class="n">adapter</span> <span class="o">=</span> <span class="n">Adapter</span><span class="p">(</span>
</span><span id="__span-28-53"><a id="__codelineno-28-53" name="__codelineno-28-53" href="#__codelineno-28-53"></a>    <span class="n">tsv_path</span><span class="o">=</span><span class="n">paths</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
</span><span id="__span-28-54"><a id="__codelineno-28-54" name="__codelineno-28-54" href="#__codelineno-28-54"></a>    <span class="n">node_types</span><span class="o">=</span><span class="n">node_types</span><span class="p">,</span>
</span><span id="__span-28-55"><a id="__codelineno-28-55" name="__codelineno-28-55" href="#__codelineno-28-55"></a>    <span class="n">node_fields</span><span class="o">=</span><span class="n">node_fields</span><span class="p">,</span>
</span><span id="__span-28-56"><a id="__codelineno-28-56" name="__codelineno-28-56" href="#__codelineno-28-56"></a>    <span class="n">edge_types</span><span class="o">=</span><span class="n">edge_types</span><span class="p">,</span>
</span><span id="__span-28-57"><a id="__codelineno-28-57" name="__codelineno-28-57" href="#__codelineno-28-57"></a><span class="p">)</span>
</span><span id="__span-28-58"><a id="__codelineno-28-58" name="__codelineno-28-58" href="#__codelineno-28-58"></a>
</span><span id="__span-28-59"><a id="__codelineno-28-59" name="__codelineno-28-59" href="#__codelineno-28-59"></a><span class="c1"># Create a knowledge graph from the adapter</span>
</span><span id="__span-28-60"><a id="__codelineno-28-60" name="__codelineno-28-60" href="#__codelineno-28-60"></a><span class="n">bc</span><span class="o">.</span><span class="n">write_nodes</span><span class="p">(</span><span class="n">adapter</span><span class="o">.</span><span class="n">get_nodes</span><span class="p">())</span>
</span><span id="__span-28-61"><a id="__codelineno-28-61" name="__codelineno-28-61" href="#__codelineno-28-61"></a><span class="n">bc</span><span class="o">.</span><span class="n">write_edges</span><span class="p">(</span><span class="n">adapter</span><span class="o">.</span><span class="n">get_edges</span><span class="p">())</span>
</span><span id="__span-28-62"><a id="__codelineno-28-62" name="__codelineno-28-62" href="#__codelineno-28-62"></a>
</span><span id="__span-28-63"><a id="__codelineno-28-63" name="__codelineno-28-63" href="#__codelineno-28-63"></a><span class="c1"># Generate assets for Neo4j exportation</span>
</span><span id="__span-28-64"><a id="__codelineno-28-64" name="__codelineno-28-64" href="#__codelineno-28-64"></a><span class="n">bc</span><span class="o">.</span><span class="n">write_import_call</span><span class="p">()</span>
</span><span id="__span-28-65"><a id="__codelineno-28-65" name="__codelineno-28-65" href="#__codelineno-28-65"></a>
</span><span id="__span-28-66"><a id="__codelineno-28-66" name="__codelineno-28-66" href="#__codelineno-28-66"></a><span class="c1"># Print a summary when</span>
</span><span id="__span-28-67"><a id="__codelineno-28-67" name="__codelineno-28-67" href="#__codelineno-28-67"></a><span class="n">bc</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
</span></code></pre></div>
</details>
<h4 id="run-the-script">Run the script</h4>
<p>You can execute the entire pipeline that loads, processes, and builds the graph by running the following command from the root folder of your project. The example below uses <code>uv</code>.
<div class="language-bash highlight"><pre><span></span><code><span id="__span-29-1"><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a>uv<span class="w"> </span>run<span class="w"> </span>python<span class="w"> </span>create_knowledge_graph.py
</span></code></pre></div></p>
<blockquote>
<p>🏆 <strong>Note:</strong> Once you complete the process, your terminal output should look similar to the following:</p>
</blockquote>
<details class="info">
<summary>Terminal output:</summary>
<div class="language-markdown highlight"><pre><span></span><code><span id="__span-30-1"><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a>INFO -- This is BioCypher v0.10.1.
</span><span id="__span-30-2"><a id="__codelineno-30-2" name="__codelineno-30-2" href="#__codelineno-30-2"></a>INFO -- Logging into <span class="sb">`biocypher-log/biocypher-20250818-153024.log`</span>.
</span><span id="__span-30-3"><a id="__codelineno-30-3" name="__codelineno-30-3" href="#__codelineno-30-3"></a>INFO -- Running BioCypher with schema configuration from config/schema_config.yaml.
</span><span id="__span-30-4"><a id="__codelineno-30-4" name="__codelineno-30-4" href="#__codelineno-30-4"></a>INFO -- Loading cache file .cache/cache.json.
</span><span id="__span-30-5"><a id="__codelineno-30-5" name="__codelineno-30-5" href="#__codelineno-30-5"></a>INFO -- Use cached version from .cache/protein-protein-interaction-dataset.
</span><span id="__span-30-6"><a id="__codelineno-30-6" name="__codelineno-30-6" href="#__codelineno-30-6"></a>Path to the resouce: [&#39;.cache/protein-protein-interaction-dataset/synthetic_protein_interactions.tsv&#39;]
</span><span id="__span-30-7"><a id="__codelineno-30-7" name="__codelineno-30-7" href="#__codelineno-30-7"></a>INFO -- Loading ontologies...
</span><span id="__span-30-8"><a id="__codelineno-30-8" name="__codelineno-30-8" href="#__codelineno-30-8"></a>INFO -- Instantiating OntologyAdapter class for https://github.com/biolink/biolink-model/raw/v3.2.1/biolink-model.owl.ttl.
</span><span id="__span-30-9"><a id="__codelineno-30-9" name="__codelineno-30-9" href="#__codelineno-30-9"></a>INFO -- Reading nodes.
</span><span id="__span-30-10"><a id="__codelineno-30-10" name="__codelineno-30-10" href="#__codelineno-30-10"></a>INFO -- Creating output directory <span class="sb">`/home/hostname/tutorial-basics-biocypher/biocypher-out/20250818153026`</span>.
</span><span id="__span-30-11"><a id="__codelineno-30-11" name="__codelineno-30-11" href="#__codelineno-30-11"></a>WARNING -- Duplicate node type protein found.
</span><span id="__span-30-12"><a id="__codelineno-30-12" name="__codelineno-30-12" href="#__codelineno-30-12"></a>INFO -- Writing 15 entries to Protein-part000.csv
</span><span id="__span-30-13"><a id="__codelineno-30-13" name="__codelineno-30-13" href="#__codelineno-30-13"></a>INFO -- Generating edges.
</span><span id="__span-30-14"><a id="__codelineno-30-14" name="__codelineno-30-14" href="#__codelineno-30-14"></a>WARNING -- Duplicate edge type ubiquitination found.
</span><span id="__span-30-15"><a id="__codelineno-30-15" name="__codelineno-30-15" href="#__codelineno-30-15"></a>WARNING -- Duplicate edge type phosphorylation found.
</span><span id="__span-30-16"><a id="__codelineno-30-16" name="__codelineno-30-16" href="#__codelineno-30-16"></a>INFO -- Writing 3 entries to Binding-part000.csv
</span><span id="__span-30-17"><a id="__codelineno-30-17" name="__codelineno-30-17" href="#__codelineno-30-17"></a>INFO -- Writing 6 entries to Activation-part000.csv
</span><span id="__span-30-18"><a id="__codelineno-30-18" name="__codelineno-30-18" href="#__codelineno-30-18"></a>INFO -- Writing 3 entries to Phosphorylation-part000.csv
</span><span id="__span-30-19"><a id="__codelineno-30-19" name="__codelineno-30-19" href="#__codelineno-30-19"></a>INFO -- Writing 7 entries to Ubiquitination-part000.csv
</span><span id="__span-30-20"><a id="__codelineno-30-20" name="__codelineno-30-20" href="#__codelineno-30-20"></a>INFO -- Writing 2 entries to Inhibition-part000.csv
</span><span id="__span-30-21"><a id="__codelineno-30-21" name="__codelineno-30-21" href="#__codelineno-30-21"></a>INFO -- Writing neo4j import call to <span class="sb">`/home/hostname/tutorial-basics-biocypher/biocypher-out/20250818153026/neo4j-admin-import-call.sh`</span>.
</span><span id="__span-30-22"><a id="__codelineno-30-22" name="__codelineno-30-22" href="#__codelineno-30-22"></a>INFO -- Showing ontology structure based on https://github.com/biolink/biolink-model/raw/v3.2.1/biolink-model.owl.ttl
</span><span id="__span-30-23"><a id="__codelineno-30-23" name="__codelineno-30-23" href="#__codelineno-30-23"></a>INFO --
</span><span id="__span-30-24"><a id="__codelineno-30-24" name="__codelineno-30-24" href="#__codelineno-30-24"></a>entity
</span><span id="__span-30-25"><a id="__codelineno-30-25" name="__codelineno-30-25" href="#__codelineno-30-25"></a>├── association
</span><span id="__span-30-26"><a id="__codelineno-30-26" name="__codelineno-30-26" href="#__codelineno-30-26"></a>│   └── gene to gene association
</span><span id="__span-30-27"><a id="__codelineno-30-27" name="__codelineno-30-27" href="#__codelineno-30-27"></a>│       └── pairwise gene to gene interaction
</span><span id="__span-30-28"><a id="__codelineno-30-28" name="__codelineno-30-28" href="#__codelineno-30-28"></a>│           └── pairwise molecular interaction
</span><span id="__span-30-29"><a id="__codelineno-30-29" name="__codelineno-30-29" href="#__codelineno-30-29"></a>│               └── protein protein interaction
</span><span id="__span-30-30"><a id="__codelineno-30-30" name="__codelineno-30-30" href="#__codelineno-30-30"></a>│                   ├── activation
</span><span id="__span-30-31"><a id="__codelineno-30-31" name="__codelineno-30-31" href="#__codelineno-30-31"></a>│                   ├── binding
</span><span id="__span-30-32"><a id="__codelineno-30-32" name="__codelineno-30-32" href="#__codelineno-30-32"></a>│                   ├── inhibition
</span><span id="__span-30-33"><a id="__codelineno-30-33" name="__codelineno-30-33" href="#__codelineno-30-33"></a>│                   ├── phosphorylation
</span><span id="__span-30-34"><a id="__codelineno-30-34" name="__codelineno-30-34" href="#__codelineno-30-34"></a>│                   └── ubiquitination
</span><span id="__span-30-35"><a id="__codelineno-30-35" name="__codelineno-30-35" href="#__codelineno-30-35"></a>└── named thing
</span><span id="__span-30-36"><a id="__codelineno-30-36" name="__codelineno-30-36" href="#__codelineno-30-36"></a>    └── biological entity
</span><span id="__span-30-37"><a id="__codelineno-30-37" name="__codelineno-30-37" href="#__codelineno-30-37"></a>        └── polypeptide
</span><span id="__span-30-38"><a id="__codelineno-30-38" name="__codelineno-30-38" href="#__codelineno-30-38"></a>            └── protein
</span><span id="__span-30-39"><a id="__codelineno-30-39" name="__codelineno-30-39" href="#__codelineno-30-39"></a>
</span><span id="__span-30-40"><a id="__codelineno-30-40" name="__codelineno-30-40" href="#__codelineno-30-40"></a>INFO -- Duplicate node types encountered (IDs in log):
</span><span id="__span-30-41"><a id="__codelineno-30-41" name="__codelineno-30-41" href="#__codelineno-30-41"></a>    protein
</span><span id="__span-30-42"><a id="__codelineno-30-42" name="__codelineno-30-42" href="#__codelineno-30-42"></a>
</span><span id="__span-30-43"><a id="__codelineno-30-43" name="__codelineno-30-43" href="#__codelineno-30-43"></a>INFO -- Duplicate edge types encountered (IDs in log):
</span><span id="__span-30-44"><a id="__codelineno-30-44" name="__codelineno-30-44" href="#__codelineno-30-44"></a>    ubiquitination
</span><span id="__span-30-45"><a id="__codelineno-30-45" name="__codelineno-30-45" href="#__codelineno-30-45"></a>    phosphorylation
</span><span id="__span-30-46"><a id="__codelineno-30-46" name="__codelineno-30-46" href="#__codelineno-30-46"></a>
</span><span id="__span-30-47"><a id="__codelineno-30-47" name="__codelineno-30-47" href="#__codelineno-30-47"></a>INFO -- No missing labels in input.
</span></code></pre></div>
</details>
<p>Note that BioCypher creates logging information and output files in a subdirectory relative to where it is executed, <code>biocypher-log/biocyper-&lt;TIMESTAMP&gt;.log</code> and <code>biocypher-out/&lt;TIMESTAMP&gt;</code>. This allows you to look up details from the biocypher run and compare with the generated output.</p>
<h2 id="section-4-interacting-with-your-graph-using-neo4j">Section 4. Interacting with your graph using Neo4j</h2>
<h3 id="load-the-graph-using-an-import-script">Load the graph using an import script</h3>
<p>When you run the <code>create_knowledge_graph.py</code> script and it completes successfully, it generates several CSV files and an import script to load the graph data into Neo4j.</p>
<p>a. Look for a folder whose name starts with <code>biocypher-out</code>. Each time you run the script, a new folder is created inside <code>biocypher-out</code> with a timestamp. Inside this folder, you should see the following:</p>
<p>🟨 CSV files associated to <strong>nodes</strong>.</p>
<p>🟦 CSV files associated to <strong>edges</strong>.</p>
<p>🟥 admin import script</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-31-1"><a id="__codelineno-31-1" name="__codelineno-31-1" href="#__codelineno-31-1"></a>/biocypher-out
</span><span id="__span-31-2"><a id="__codelineno-31-2" name="__codelineno-31-2" href="#__codelineno-31-2"></a>└── 20250818153026
</span><span id="__span-31-3"><a id="__codelineno-31-3" name="__codelineno-31-3" href="#__codelineno-31-3"></a>    ├── 🟦 Activation-header.csv
</span><span id="__span-31-4"><a id="__codelineno-31-4" name="__codelineno-31-4" href="#__codelineno-31-4"></a>    ├── 🟦 Activation-part000.csv
</span><span id="__span-31-5"><a id="__codelineno-31-5" name="__codelineno-31-5" href="#__codelineno-31-5"></a>    ├── 🟦 Binding-header.csv
</span><span id="__span-31-6"><a id="__codelineno-31-6" name="__codelineno-31-6" href="#__codelineno-31-6"></a>    ├── 🟦 Binding-part000.csv
</span><span id="__span-31-7"><a id="__codelineno-31-7" name="__codelineno-31-7" href="#__codelineno-31-7"></a>    ├── 🟦 Inhibition-header.csv
</span><span id="__span-31-8"><a id="__codelineno-31-8" name="__codelineno-31-8" href="#__codelineno-31-8"></a>    ├── 🟦 Inhibition-part000.csv
</span><span id="__span-31-9"><a id="__codelineno-31-9" name="__codelineno-31-9" href="#__codelineno-31-9"></a>    ├── 🟥 neo4j-admin-import-call.sh
</span><span id="__span-31-10"><a id="__codelineno-31-10" name="__codelineno-31-10" href="#__codelineno-31-10"></a>    ├── 🟦 Phosphorylation-header.csv
</span><span id="__span-31-11"><a id="__codelineno-31-11" name="__codelineno-31-11" href="#__codelineno-31-11"></a>    ├── 🟦 Phosphorylation-part000.csv
</span><span id="__span-31-12"><a id="__codelineno-31-12" name="__codelineno-31-12" href="#__codelineno-31-12"></a>    ├── 🟨 Protein-header.csv
</span><span id="__span-31-13"><a id="__codelineno-31-13" name="__codelineno-31-13" href="#__codelineno-31-13"></a>    ├── 🟨 Protein-part000.csv
</span><span id="__span-31-14"><a id="__codelineno-31-14" name="__codelineno-31-14" href="#__codelineno-31-14"></a>    ├── 🟦 Ubiquitination-header.csv
</span></code></pre></div>
<p>b. Stop the neo4j instance. You can do this on the GUI or in terminal. In terminal, you must locate the <code>neo4j</code> executable in the Neo4j instance path.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-32-1"><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a>&lt;path<span class="w"> </span>of<span class="w"> </span>your<span class="w"> </span>Neo4j<span class="w"> </span>instance&gt;/bin/neo4j<span class="w"> </span>stop
</span></code></pre></div>
<p>c. Run the  <code>neo4j-admin-import-call.sh</code> script in your <code>biocypher-output/</code>:
<div class="language-bash highlight"><pre><span></span><code><span id="__span-33-1"><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a>bash<span class="w"> </span>./biocypher-out/20250818153026/neo4j-admin-import-call.sh
</span></code></pre></div></p>
<p>d. If everything has been successfully, you should see in terminal something similar to this:</p>
<details class="info">
<summary>Terminal output:</summary>
<div class="language-text highlight"><pre><span></span><code><span id="__span-34-1"><a id="__codelineno-34-1" name="__codelineno-34-1" href="#__codelineno-34-1"></a>Starting to import, output will be saved to: /home/egcarren/.config/neo4j-desktop/Application/Data/dbmss/dbms-08155706-b96e-4e74-a965-7d6d27b78db8/logs/neo4j-admin-import-2025-08-18.15.54.59.log
</span><span id="__span-34-2"><a id="__codelineno-34-2" name="__codelineno-34-2" href="#__codelineno-34-2"></a>Neo4j version: 2025.07.1
</span><span id="__span-34-3"><a id="__codelineno-34-3" name="__codelineno-34-3" href="#__codelineno-34-3"></a>Importing the contents of these files into /home/egcarren/.config/neo4j-desktop/Application/Data/dbmss/dbms-08155706-b96e-4e74-a965-7d6d27b78db8/data/databases/neo4j:
</span><span id="__span-34-4"><a id="__codelineno-34-4" name="__codelineno-34-4" href="#__codelineno-34-4"></a>Nodes:
</span><span id="__span-34-5"><a id="__codelineno-34-5" name="__codelineno-34-5" href="#__codelineno-34-5"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Protein-header.csv
</span><span id="__span-34-6"><a id="__codelineno-34-6" name="__codelineno-34-6" href="#__codelineno-34-6"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Protein-part000.csv
</span><span id="__span-34-7"><a id="__codelineno-34-7" name="__codelineno-34-7" href="#__codelineno-34-7"></a>
</span><span id="__span-34-8"><a id="__codelineno-34-8" name="__codelineno-34-8" href="#__codelineno-34-8"></a>Relationships:
</span><span id="__span-34-9"><a id="__codelineno-34-9" name="__codelineno-34-9" href="#__codelineno-34-9"></a>null:
</span><span id="__span-34-10"><a id="__codelineno-34-10" name="__codelineno-34-10" href="#__codelineno-34-10"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Phosphorylation-header.csv
</span><span id="__span-34-11"><a id="__codelineno-34-11" name="__codelineno-34-11" href="#__codelineno-34-11"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Phosphorylation-part000.csv
</span><span id="__span-34-12"><a id="__codelineno-34-12" name="__codelineno-34-12" href="#__codelineno-34-12"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Ubiquitination-header.csv
</span><span id="__span-34-13"><a id="__codelineno-34-13" name="__codelineno-34-13" href="#__codelineno-34-13"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Ubiquitination-part000.csv
</span><span id="__span-34-14"><a id="__codelineno-34-14" name="__codelineno-34-14" href="#__codelineno-34-14"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Inhibition-header.csv
</span><span id="__span-34-15"><a id="__codelineno-34-15" name="__codelineno-34-15" href="#__codelineno-34-15"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Inhibition-part000.csv
</span><span id="__span-34-16"><a id="__codelineno-34-16" name="__codelineno-34-16" href="#__codelineno-34-16"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Activation-header.csv
</span><span id="__span-34-17"><a id="__codelineno-34-17" name="__codelineno-34-17" href="#__codelineno-34-17"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Activation-part000.csv
</span><span id="__span-34-18"><a id="__codelineno-34-18" name="__codelineno-34-18" href="#__codelineno-34-18"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Binding-header.csv
</span><span id="__span-34-19"><a id="__codelineno-34-19" name="__codelineno-34-19" href="#__codelineno-34-19"></a>/home/egcarren/Downloads/sandbox_edwin/tutorial-basics-biocypher/biocypher-out/20250818153026/Binding-part000.csv
</span><span id="__span-34-20"><a id="__codelineno-34-20" name="__codelineno-34-20" href="#__codelineno-34-20"></a>
</span><span id="__span-34-21"><a id="__codelineno-34-21" name="__codelineno-34-21" href="#__codelineno-34-21"></a>
</span><span id="__span-34-22"><a id="__codelineno-34-22" name="__codelineno-34-22" href="#__codelineno-34-22"></a>Available resources:
</span><span id="__span-34-23"><a id="__codelineno-34-23" name="__codelineno-34-23" href="#__codelineno-34-23"></a>Total machine memory: 15.27GiB
</span><span id="__span-34-24"><a id="__codelineno-34-24" name="__codelineno-34-24" href="#__codelineno-34-24"></a>Free machine memory: 1.112GiB
</span><span id="__span-34-25"><a id="__codelineno-34-25" name="__codelineno-34-25" href="#__codelineno-34-25"></a>Max heap memory : 910.5MiB
</span><span id="__span-34-26"><a id="__codelineno-34-26" name="__codelineno-34-26" href="#__codelineno-34-26"></a>Max worker threads: 12
</span><span id="__span-34-27"><a id="__codelineno-34-27" name="__codelineno-34-27" href="#__codelineno-34-27"></a>Configured max memory: 541.5MiB
</span><span id="__span-34-28"><a id="__codelineno-34-28" name="__codelineno-34-28" href="#__codelineno-34-28"></a>High parallel IO: true
</span><span id="__span-34-29"><a id="__codelineno-34-29" name="__codelineno-34-29" href="#__codelineno-34-29"></a>
</span><span id="__span-34-30"><a id="__codelineno-34-30" name="__codelineno-34-30" href="#__codelineno-34-30"></a>Import starting
</span><span id="__span-34-31"><a id="__codelineno-34-31" name="__codelineno-34-31" href="#__codelineno-34-31"></a>Page cache size: 539.4MiB
</span><span id="__span-34-32"><a id="__codelineno-34-32" name="__codelineno-34-32" href="#__codelineno-34-32"></a>Number of worker threads: 12
</span><span id="__span-34-33"><a id="__codelineno-34-33" name="__codelineno-34-33" href="#__codelineno-34-33"></a>Estimated number of nodes: 15
</span><span id="__span-34-34"><a id="__codelineno-34-34" name="__codelineno-34-34" href="#__codelineno-34-34"></a>Estimated number of relationships: 21
</span><span id="__span-34-35"><a id="__codelineno-34-35" name="__codelineno-34-35" href="#__codelineno-34-35"></a>
</span><span id="__span-34-36"><a id="__codelineno-34-36" name="__codelineno-34-36" href="#__codelineno-34-36"></a>Importing nodes
</span><span id="__span-34-37"><a id="__codelineno-34-37" name="__codelineno-34-37" href="#__codelineno-34-37"></a>.......... .......... .......... .......... ..........   5% ∆295ms [295ms]
</span><span id="__span-34-38"><a id="__codelineno-34-38" name="__codelineno-34-38" href="#__codelineno-34-38"></a>.......... .......... .......... .......... ..........  10% ∆2ms [298ms]
</span><span id="__span-34-39"><a id="__codelineno-34-39" name="__codelineno-34-39" href="#__codelineno-34-39"></a>.......... .......... .......... .......... ..........  15% ∆0ms [298ms]
</span><span id="__span-34-40"><a id="__codelineno-34-40" name="__codelineno-34-40" href="#__codelineno-34-40"></a>.......... .......... .......... .......... ..........  20% ∆0ms [299ms]
</span><span id="__span-34-41"><a id="__codelineno-34-41" name="__codelineno-34-41" href="#__codelineno-34-41"></a>.......... .......... .......... .......... ..........  25% ∆0ms [299ms]
</span><span id="__span-34-42"><a id="__codelineno-34-42" name="__codelineno-34-42" href="#__codelineno-34-42"></a>.......... .......... .......... .......... ..........  30% ∆0ms [299ms]
</span><span id="__span-34-43"><a id="__codelineno-34-43" name="__codelineno-34-43" href="#__codelineno-34-43"></a>.......... .......... .......... .......... ..........  35% ∆0ms [300ms]
</span><span id="__span-34-44"><a id="__codelineno-34-44" name="__codelineno-34-44" href="#__codelineno-34-44"></a>.......... .......... .......... .......... ..........  40% ∆0ms [300ms]
</span><span id="__span-34-45"><a id="__codelineno-34-45" name="__codelineno-34-45" href="#__codelineno-34-45"></a>.......... .......... .......... .......... ..........  45% ∆0ms [300ms]
</span><span id="__span-34-46"><a id="__codelineno-34-46" name="__codelineno-34-46" href="#__codelineno-34-46"></a>.......... .......... .......... .......... ..........  50% ∆0ms [301ms]
</span><span id="__span-34-47"><a id="__codelineno-34-47" name="__codelineno-34-47" href="#__codelineno-34-47"></a>.......... .......... .......... .......... ..........  55% ∆0ms [301ms]
</span><span id="__span-34-48"><a id="__codelineno-34-48" name="__codelineno-34-48" href="#__codelineno-34-48"></a>.......... .......... .......... .......... ..........  60% ∆0ms [301ms]
</span><span id="__span-34-49"><a id="__codelineno-34-49" name="__codelineno-34-49" href="#__codelineno-34-49"></a>.......... .......... .......... .......... ..........  65% ∆0ms [302ms]
</span><span id="__span-34-50"><a id="__codelineno-34-50" name="__codelineno-34-50" href="#__codelineno-34-50"></a>.......... .......... .......... .......... ..........  70% ∆0ms [302ms]
</span><span id="__span-34-51"><a id="__codelineno-34-51" name="__codelineno-34-51" href="#__codelineno-34-51"></a>.......... .......... .......... .......... ..........  75% ∆0ms [302ms]
</span><span id="__span-34-52"><a id="__codelineno-34-52" name="__codelineno-34-52" href="#__codelineno-34-52"></a>.......... .......... .......... .......... ..........  80% ∆0ms [303ms]
</span><span id="__span-34-53"><a id="__codelineno-34-53" name="__codelineno-34-53" href="#__codelineno-34-53"></a>.......... .......... .......... .......... ..........  85% ∆0ms [303ms]
</span><span id="__span-34-54"><a id="__codelineno-34-54" name="__codelineno-34-54" href="#__codelineno-34-54"></a>.......... .......... .......... .......... ..........  90% ∆0ms [303ms]
</span><span id="__span-34-55"><a id="__codelineno-34-55" name="__codelineno-34-55" href="#__codelineno-34-55"></a>.......... .......... .......... .......... ..........  95% ∆0ms [304ms]
</span><span id="__span-34-56"><a id="__codelineno-34-56" name="__codelineno-34-56" href="#__codelineno-34-56"></a>.......... .......... .......... .......... .......... 100% ∆0ms [304ms]
</span><span id="__span-34-57"><a id="__codelineno-34-57" name="__codelineno-34-57" href="#__codelineno-34-57"></a>Prepare ID mapper
</span><span id="__span-34-58"><a id="__codelineno-34-58" name="__codelineno-34-58" href="#__codelineno-34-58"></a>.......... .......... .......... .......... ..........   5% ∆33ms [33ms]
</span><span id="__span-34-59"><a id="__codelineno-34-59" name="__codelineno-34-59" href="#__codelineno-34-59"></a>.......... .......... .......... .......... ..........  10% ∆0ms [33ms]
</span><span id="__span-34-60"><a id="__codelineno-34-60" name="__codelineno-34-60" href="#__codelineno-34-60"></a>.......... .......... .......... .......... ..........  15% ∆0ms [33ms]
</span><span id="__span-34-61"><a id="__codelineno-34-61" name="__codelineno-34-61" href="#__codelineno-34-61"></a>.......... .......... .......... .......... ..........  20% ∆0ms [34ms]
</span><span id="__span-34-62"><a id="__codelineno-34-62" name="__codelineno-34-62" href="#__codelineno-34-62"></a>.......... .......... .......... .......... ..........  25% ∆0ms [34ms]
</span><span id="__span-34-63"><a id="__codelineno-34-63" name="__codelineno-34-63" href="#__codelineno-34-63"></a>.......... .......... .......... .......... ..........  30% ∆0ms [34ms]
</span><span id="__span-34-64"><a id="__codelineno-34-64" name="__codelineno-34-64" href="#__codelineno-34-64"></a>.......... .......... .......... .......... ..........  35% ∆46ms [81ms]
</span><span id="__span-34-65"><a id="__codelineno-34-65" name="__codelineno-34-65" href="#__codelineno-34-65"></a>.......... .......... .......... .......... ..........  40% ∆0ms [81ms]
</span><span id="__span-34-66"><a id="__codelineno-34-66" name="__codelineno-34-66" href="#__codelineno-34-66"></a>.......... .......... .......... .......... ..........  45% ∆0ms [81ms]
</span><span id="__span-34-67"><a id="__codelineno-34-67" name="__codelineno-34-67" href="#__codelineno-34-67"></a>.......... .......... .......... .......... ..........  50% ∆0ms [82ms]
</span><span id="__span-34-68"><a id="__codelineno-34-68" name="__codelineno-34-68" href="#__codelineno-34-68"></a>.......... .......... .......... .......... ..........  55% ∆0ms [82ms]
</span><span id="__span-34-69"><a id="__codelineno-34-69" name="__codelineno-34-69" href="#__codelineno-34-69"></a>.......... .......... .......... .......... ..........  60% ∆0ms [82ms]
</span><span id="__span-34-70"><a id="__codelineno-34-70" name="__codelineno-34-70" href="#__codelineno-34-70"></a>.......... .......... .......... .......... ..........  65% ∆0ms [82ms]
</span><span id="__span-34-71"><a id="__codelineno-34-71" name="__codelineno-34-71" href="#__codelineno-34-71"></a>.......... .......... .......... .......... ..........  70% ∆0ms [82ms]
</span><span id="__span-34-72"><a id="__codelineno-34-72" name="__codelineno-34-72" href="#__codelineno-34-72"></a>.......... .......... .......... .......... ..........  75% ∆0ms [83ms]
</span><span id="__span-34-73"><a id="__codelineno-34-73" name="__codelineno-34-73" href="#__codelineno-34-73"></a>.......... .......... .......... .......... ..........  80% ∆0ms [83ms]
</span><span id="__span-34-74"><a id="__codelineno-34-74" name="__codelineno-34-74" href="#__codelineno-34-74"></a>.......... .......... .......... .......... ..........  85% ∆0ms [83ms]
</span><span id="__span-34-75"><a id="__codelineno-34-75" name="__codelineno-34-75" href="#__codelineno-34-75"></a>.......... .......... .......... .......... ..........  90% ∆0ms [83ms]
</span><span id="__span-34-76"><a id="__codelineno-34-76" name="__codelineno-34-76" href="#__codelineno-34-76"></a>.......... .......... .......... .......... ..........  95% ∆1ms [85ms]
</span><span id="__span-34-77"><a id="__codelineno-34-77" name="__codelineno-34-77" href="#__codelineno-34-77"></a>.......... .......... .......... .......... .......... 100% ∆0ms [85ms]
</span><span id="__span-34-78"><a id="__codelineno-34-78" name="__codelineno-34-78" href="#__codelineno-34-78"></a>Imported 15 nodes in 458ms
</span><span id="__span-34-79"><a id="__codelineno-34-79" name="__codelineno-34-79" href="#__codelineno-34-79"></a>using configuration:Configuration[numberOfWorkers=12, temporaryPath=/home/egcarren/.config/neo4j-desktop/Application/Data/dbmss/dbms-08155706-b96e-4e74-a965-7d6d27b78db8/data/databases/neo4j/temp, applyBatchSize=64, sorterSizeSwitchFactor=0.3]
</span><span id="__span-34-80"><a id="__codelineno-34-80" name="__codelineno-34-80" href="#__codelineno-34-80"></a>Importing relationships
</span><span id="__span-34-81"><a id="__codelineno-34-81" name="__codelineno-34-81" href="#__codelineno-34-81"></a>.......... .......... .......... .......... ..........   5% ∆35ms [35ms]
</span><span id="__span-34-82"><a id="__codelineno-34-82" name="__codelineno-34-82" href="#__codelineno-34-82"></a>.......... .......... .......... .......... ..........  10% ∆0ms [35ms]
</span><span id="__span-34-83"><a id="__codelineno-34-83" name="__codelineno-34-83" href="#__codelineno-34-83"></a>.......... .......... .......... .......... ..........  15% ∆0ms [35ms]
</span><span id="__span-34-84"><a id="__codelineno-34-84" name="__codelineno-34-84" href="#__codelineno-34-84"></a>.......... .......... .......... .......... ..........  20% ∆0ms [35ms]
</span><span id="__span-34-85"><a id="__codelineno-34-85" name="__codelineno-34-85" href="#__codelineno-34-85"></a>.......... .......... .......... .......... ..........  25% ∆0ms [35ms]
</span><span id="__span-34-86"><a id="__codelineno-34-86" name="__codelineno-34-86" href="#__codelineno-34-86"></a>.......... .......... .......... .......... ..........  30% ∆0ms [36ms]
</span><span id="__span-34-87"><a id="__codelineno-34-87" name="__codelineno-34-87" href="#__codelineno-34-87"></a>.......... .......... .......... .......... ..........  35% ∆28ms [64ms]
</span><span id="__span-34-88"><a id="__codelineno-34-88" name="__codelineno-34-88" href="#__codelineno-34-88"></a>.......... .......... .......... .......... ..........  40% ∆0ms [64ms]
</span><span id="__span-34-89"><a id="__codelineno-34-89" name="__codelineno-34-89" href="#__codelineno-34-89"></a>.......... .......... .......... .......... ..........  45% ∆0ms [64ms]
</span><span id="__span-34-90"><a id="__codelineno-34-90" name="__codelineno-34-90" href="#__codelineno-34-90"></a>.......... .......... .......... .......... ..........  50% ∆0ms [64ms]
</span><span id="__span-34-91"><a id="__codelineno-34-91" name="__codelineno-34-91" href="#__codelineno-34-91"></a>.......... .......... .......... .......... ..........  55% ∆0ms [64ms]
</span><span id="__span-34-92"><a id="__codelineno-34-92" name="__codelineno-34-92" href="#__codelineno-34-92"></a>.......... .......... .......... .......... ..........  60% ∆0ms [64ms]
</span><span id="__span-34-93"><a id="__codelineno-34-93" name="__codelineno-34-93" href="#__codelineno-34-93"></a>.......... .......... .......... .......... ..........  65% ∆0ms [64ms]
</span><span id="__span-34-94"><a id="__codelineno-34-94" name="__codelineno-34-94" href="#__codelineno-34-94"></a>.......... .......... .......... .......... ..........  70% ∆0ms [64ms]
</span><span id="__span-34-95"><a id="__codelineno-34-95" name="__codelineno-34-95" href="#__codelineno-34-95"></a>.......... .......... .......... .......... ..........  75% ∆0ms [64ms]
</span><span id="__span-34-96"><a id="__codelineno-34-96" name="__codelineno-34-96" href="#__codelineno-34-96"></a>.......... .......... .......... .......... ..........  80% ∆0ms [64ms]
</span><span id="__span-34-97"><a id="__codelineno-34-97" name="__codelineno-34-97" href="#__codelineno-34-97"></a>.......... .......... .......... .......... ..........  85% ∆0ms [64ms]
</span><span id="__span-34-98"><a id="__codelineno-34-98" name="__codelineno-34-98" href="#__codelineno-34-98"></a>.......... .......... .......... .......... ..........  90% ∆0ms [64ms]
</span><span id="__span-34-99"><a id="__codelineno-34-99" name="__codelineno-34-99" href="#__codelineno-34-99"></a>.......... .......... .......... .......... ..........  95% ∆0ms [65ms]
</span><span id="__span-34-100"><a id="__codelineno-34-100" name="__codelineno-34-100" href="#__codelineno-34-100"></a>.......... .......... .......... .......... .......... 100% ∆0ms [65ms]
</span><span id="__span-34-101"><a id="__codelineno-34-101" name="__codelineno-34-101" href="#__codelineno-34-101"></a>Imported 21 relationships in 248ms
</span><span id="__span-34-102"><a id="__codelineno-34-102" name="__codelineno-34-102" href="#__codelineno-34-102"></a>Flushing stores
</span><span id="__span-34-103"><a id="__codelineno-34-103" name="__codelineno-34-103" href="#__codelineno-34-103"></a>Flush completed in 18ms
</span><span id="__span-34-104"><a id="__codelineno-34-104" name="__codelineno-34-104" href="#__codelineno-34-104"></a>IMPORT DONE in 1s 488ms.
</span></code></pre></div>
</details>
<h3 id="visualize-the-graph">Visualize the graph</h3>
<p>a. Connect to your instance by running Neo4j desktop again. Select your instance and click on "Connect" - the little arrow on the button allows you to expand a menu. Select the option <em>Query</em>.</p>
<figure>
<img alt="Image title" src="../assets/neo4j_explore_graph.png" width="1000" />

<figcaption>Figure 13. Query and Explore options to run on a Neo4j instance.</figcaption>
</figure>
<p>b. Now, click on the asterisk under the Relationships category. You now should see your graph! Compare to the sketch you did previosly in this tutorial</p>
<figure>
<img alt="Image title" src="../assets/neo4j_final_graph.png" width="1000" />

<figcaption>Figure 14. Neo4j graph based on our data.</figcaption>
</figure>
<h3 id="execute-cypher-queries">Execute cypher queries</h3>
<p>Try the following queries:</p>
<ol>
<li>Find relationships between two nodes
<div class="language-cypher highlight"><pre><span></span><code><span id="__span-35-1"><a id="__codelineno-35-1" name="__codelineno-35-1" href="#__codelineno-35-1"></a><span class="k">MATCH</span><span class="w"> </span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">-[</span><span class="n">r</span><span class="o">]-&gt;</span><span class="p">(</span><span class="n">b</span><span class="p">)</span>
</span><span id="__span-35-2"><a id="__codelineno-35-2" name="__codelineno-35-2" href="#__codelineno-35-2"></a><span class="k">RETURN</span><span class="w"> </span><span class="n">a</span><span class="p">,</span><span class="w"> </span><span class="n">r</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="p">;</span>
</span></code></pre></div>
Result:</li>
</ol>
<figure>
<img alt="Image title" src="../assets/neo4j_query_1.png" width="500" />
</figure>
<ol>
<li>Find all the nodes</li>
</ol>
<p><div class="language-cypher highlight"><pre><span></span><code><span id="__span-36-1"><a id="__codelineno-36-1" name="__codelineno-36-1" href="#__codelineno-36-1"></a><span class="k">MATCH</span><span class="w"> </span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
</span><span id="__span-36-2"><a id="__codelineno-36-2" name="__codelineno-36-2" href="#__codelineno-36-2"></a><span class="k">RETURN</span><span class="w"> </span><span class="n">n</span><span class="p">;</span>
</span></code></pre></div>
Result:</p>
<figure>
<img alt="Image title" src="../assets/neo4j_query_2.png" width="500" />
</figure>
<ol>
<li>Find all nodes of a specific type(e.g. <code>Protein</code> in the following query)</li>
</ol>
<p><div class="language-cypher highlight"><pre><span></span><code><span id="__span-37-1"><a id="__codelineno-37-1" name="__codelineno-37-1" href="#__codelineno-37-1"></a><span class="k">MATCH</span><span class="w"> </span><span class="p">(</span><span class="n">n</span><span class="p">:</span><span class="n">Protein</span><span class="p">)</span>
</span><span id="__span-37-2"><a id="__codelineno-37-2" name="__codelineno-37-2" href="#__codelineno-37-2"></a><span class="k">RETURN</span><span class="w"> </span><span class="n">n</span><span class="p">;</span>
</span></code></pre></div>
Result:</p>
<figure>
<img alt="Image title" src="../assets/neo4j_query_3.png" width="500" />
</figure>
<ol>
<li>Find all relationships of a specific type(e.g. <code>Binding</code> in the following query)</li>
</ol>
<p><div class="language-cypher highlight"><pre><span></span><code><span id="__span-38-1"><a id="__codelineno-38-1" name="__codelineno-38-1" href="#__codelineno-38-1"></a><span class="k">MATCH</span><span class="w"> </span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">-[</span><span class="n">r</span><span class="p">:</span><span class="n">Binding</span><span class="o">]-&gt;</span><span class="p">(</span><span class="n">b</span><span class="p">)</span>
</span><span id="__span-38-2"><a id="__codelineno-38-2" name="__codelineno-38-2" href="#__codelineno-38-2"></a><span class="k">RETURN</span><span class="w"> </span><span class="n">a</span><span class="p">,</span><span class="w"> </span><span class="n">r</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="p">;</span>
</span></code></pre></div>
Result:</p>
<figure>
<img alt="Image title" src="../assets/neo4j_query_4.png" width="500" />
</figure>
<ol>
<li>Count relationships of a given type(e.g. <code>Binding</code> in the following query)</li>
</ol>
<p><div class="language-cypher highlight"><pre><span></span><code><span id="__span-39-1"><a id="__codelineno-39-1" name="__codelineno-39-1" href="#__codelineno-39-1"></a><span class="k">MATCH</span><span class="w"> </span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">-[</span><span class="n">r</span><span class="p">:</span><span class="n">Binding</span><span class="o">]-&gt;</span><span class="p">(</span><span class="n">b</span><span class="p">)</span>
</span><span id="__span-39-2"><a id="__codelineno-39-2" name="__codelineno-39-2" href="#__codelineno-39-2"></a><span class="k">RETURN</span><span class="w"> </span><span class="n">COUNT</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w"> </span><span class="k">AS</span><span class="w"> </span><span class="n">totalBindings</span><span class="p">;</span>
</span></code></pre></div>
Result:</p>
<figure>
<img alt="Image title" src="../assets/neo4j_query_5.png" width="250" />
</figure>
<hr />
<h2 id="next-steps">Next Steps</h2>
<ul>
<li>Explore more advanced queries and graph analytics in Neo4j.</li>
<li>Try integrating additional datasets or expanding your graph model. For guidance, see the <a href="https://biocypher.org/BioCypher/learn/tutorials/tutorial001_basics/">Basics</a> tutorial.</li>
<li>Review the <a href="https://biocypher.org/">BioCypher documentation</a> for deeper insights.</li>
</ul>
<h2 id="feedback-contributions">Feedback &amp; Contributions</h2>
<p>If you found this tutorial helpful or have suggestions for improvement, please <strong>open an issue</strong> or <strong>submit a pull request</strong> in the <a href="https://github.com/biocypher/biocypher/issues/new/choose">BioCypher repository</a>. Specific feedback on examples, clarity of instructions, or missing details is especially appreciated.</p>
<hr />
<table>
<thead>
<tr>
<th style="text-align: center;">Last Update</th>
<th style="text-align: left;">Developed by</th>
<th>Affiliation</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;">2025.20.18</td>
<td style="text-align: left;">Shuangshuang Li <br> Edwin Carreño (GH @ecarrenolozano)</td>
<td><a href="https://www.ssc.uni-heidelberg.de/en">Scientific Software Center</a> <br> <a href="https://saezlab.org/">Saezlab</a> - <a href="https://www.ssc.uni-heidelberg.de/en">Scientific Software Center</a></td>
</tr>
</tbody>
</table>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../../agent_api_guide/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Agent API Guide">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Agent API Guide
              </div>
            </div>
          </a>
        
        
          
          <a href="../../../guides/" class="md-footer__link md-footer__link--next" aria-label="Next: Index">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Index
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      © Copyright 2021-2025, BioCypher developers.
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://biocypher.zulipchat.com/" target="_blank" rel="noopener" title="biocypher.zulipchat.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M22.767 3.589c0 1.209-.543 2.283-1.37 2.934l-8.034 7.174c-.149.128-.343-.078-.235-.25l2.946-5.9c.083-.165-.024-.368-.194-.368H4.452c-1.77 0-3.219-1.615-3.219-3.59C1.233 1.616 2.682 0 4.452 0h15.096c1.77-.001 3.219 1.614 3.219 3.589M4.452 24h15.096c1.77 0 3.219-1.616 3.219-3.59s-1.449-3.59-3.219-3.59H8.12c-.17 0-.277-.202-.194-.367l2.946-5.9c.108-.172-.086-.378-.235-.25l-8.033 7.173c-.828.65-1.37 1.725-1.37 2.934 0 1.974 1.448 3.59 3.218 3.59"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://saezlab.org/" target="_blank" rel="noopener" title="saezlab.org" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 0H4v2h16zM4 24h16v-2H4zM20 4H4a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V6a2 2 0 0 0-2-2m-8 2.75A2.25 2.25 0 0 1 14.25 9 2.25 2.25 0 0 1 12 11.25 2.25 2.25 0 0 1 9.75 9 2.25 2.25 0 0 1 12 6.75M17 17H7v-1.5c0-1.67 3.33-2.5 5-2.5s5 .83 5 2.5z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../../..", "features": ["content.code.copy", "content.action.edit", "navigation.tabs", "navigation.instant", "navigation.footer"], "search": "../../../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.50899def.min.js"></script>
      
    
  </body>
</html>